# Spot-deconvolution of spTotal data using BayesPrism + scMuscle
## David W. McKellar
```{r}
library(Seurat, quietly = T)
library(dplyr, quietly = T)
library(ggplot2, quietly = T)
library(patchwork, quietly = T)

# TED, a.k.a. BayesPrism
## https://github.com/Danko-Lab/TED
library(TED, quietly = T)

source("/home/dwm269/DWM_utils/plotting_utils/scThemes.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seutils.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seuplots.R")
```

## Figure settings, colors, & themes
```{r}
# fonts, sizes, etc.
small.font = 6*2
big.font = 8*2
line.width = 0.5
pt.size=0.01
pt.stroke=0.3
label.size=2

source("/home/dwm269/DWM_utils/plotting_utils/scThemes.R")
scTheme <- scThemes(
  small.font = small.font,
  big.font = big.font,
  line.width = line.width,
  pt.size=pt.size,
  pt.stroke=pt.stroke,
  label.size=label.size
)

print(names(scTheme))

# colors
mckolors <- read.csv("/home/dwm269/DWM_utils/plotting_utils/McKolors_v1.csv") %>% 
  as.list() %>%
  lapply(
    FUN=function(X) X[X!=""]
  )
names(mckolors)
```

# Load and subset preprocessed Visium & spTotal data
```{r}
if(is.null(skm.list)){
  load("/local/workdir/dwm269/totalRNA/spTotal/robjs/skm_list_v2.RData")
}
meta_vis <- read.csv("/workdir/dwm269/totalRNA/spTotal/resources/metadata_sheets/meta_sheet_visium.csv")
meta_skm <- meta_vis[meta_vis$tissue=="muscle"&meta_vis$rnase_inhib%in%c("ctrl", "Protector"),]

meta_skm
```


# Load single-cell reference dataset
```{r}
if(!exists("scMuscle.seurat")){
  load("/local/workdir/dwm269/muscle_data/robjs/scMuscle_mm10_clean_v40.RData")
}
head(scMuscle.seurat)
```

# Function for generating an average expression model for a Seurat object
```{r get_cell_type_model}
get_cell_type_model <-
  function(
    SEU, #seurat object
    assay='RNA',
    slot='data',
    cluster.names, #cell type model; pass in metadata column name for desired cell types
    ignore.clusters=NULL, # vector of chars; cluster IDs to ignore in model generation
    cells.use=NULL, # vector of cell barcodes to use; works in combination with ignore.clusters, doublet.classifications
    genes.ignore=NULL, # genes to remove from model
    doublet.classifications=NULL, # name of doublet classification metadata column; default is to ignore classifications
    # nCores=1, #TODO: parallelize with dofor
    verbose=T
  ){
    require(Seurat)

    #build singlet_cell type_gene expression matrix
    ref.mat <- list() #initialize reference expression matrix
    exp.mat <- GetAssayData(SEU, assay=assay, slot=slot) # Pull out expression data

    # subset based on desired cells
    if(!is.null(cells.use)){
      exp.mat <- exp.mat[,cells.use]
    }

    # remove ignored genes
    if(!is.null(genes.ignore)){
      exp.mat = exp.mat[!rownames(SEU)%in%genes.ignore,]
    }

    #Grab meta.data, filtered for singlets
    if(is.null(cells.use)){ #TODO- subset based on cells.use
      meta <- SEU@meta.data  #ignore sinlget/doublet classifications
    }else{
      meta <- SEU@meta.data[cells.use,]
    }

    cell.types <- sort(unlist(unique(meta[[cluster.names]]))) #Get cell type names
    if(!is.null(ignore.clusters)){
      cell.types <- cell.types[(!cell.types%in%ignore.clusters)]
    }

    # Remove NAs from exp.mat and meta
    exp.mat <- exp.mat[,!is.na(meta[[cluster.names]])]
    meta <- meta[!is.na(meta[[cluster.names]]),]

    # Generate cell type references, one cell type at a time
    if(verbose){cat("Building reference profiles... \n")}
    if(verbose & !is.null(cells.use)){cat("   Using the provided cells... \n")}

    for(i in 1:length(cell.types)){
      tmp.cells = rownames(meta)[ as.vector(meta[[cluster.names]])==cell.types[i] ]

      if(verbose){cat(cell.types[i]," (",i,"/",length(cell.types),")","... ", sep="")}

      if(length(tmp.cells)==0){
        if(verbose){cat(" NO CELLS", sep="")}
      }

      if(length(tmp.cells)==1){
        if(verbose){cat(" Only 1 cell!", sep="")}
        ref.mat[[i]] <- exp.mat[,rownames(meta)%in%tmp.cells]  # grab expression data
      }else{
        if(verbose){cat(length(tmp.cells)," cells.", sep="")}
        ref.mat[[i]] <-
          Matrix::rowMeans( #TODO: should these expression profiles be built on row means?
            exp.mat[,rownames(meta)%in%tmp.cells]  # grab expression data

          )
      }

      if(verbose){cat("\n")}
    }
    ref.mat <- do.call(cbind, ref.mat)
    colnames(ref.mat) <- cell.types

    if(verbose){cat("Done!\n")}

    return(ref.mat)
  }

```

# remove mito and ribo genes; important for deconvolution - 
##     See vignette at https://github.com/Danko-Lab/TED
```{r}
# remove mito and ribo genes
genes.remove <-  unique(c(
  rownames(scMuscle.seurat)[grep("Rps",x = rownames(scMuscle.seurat))], # ribo, short
  rownames(scMuscle.seurat)[grep("Rpl",x = rownames(scMuscle.seurat))], # ribo long 
  rownames(scMuscle.seurat)[grep("mt-",x = rownames(scMuscle.seurat))] # mito
))
print(genes.remove)

#Build cell type gene expression profile - see helper_functions_v1.R for details
celltype.gep <- scMuscle.seurat %>% get_cell_type_model(
  slot="counts",
  cluster.names = "harmony_PHATE_IDs",
  genes.ignore = genes.remove 
)
```

# Deconvolution with differentially expressed 
Load in DEGs for scMuscle, filter genes list, and filter gep
```{r}
all.markers <- read.csv("/workdir/dwm269/muscle_data/scMuscle/supplemental_data/scMuscle_harmonytypes_plus_phatebins_markers_SupFile2.csv")

genes.keep = unique(all.markers$gene[all.markers$pct.1>0.5 & abs(all.markers$avg_logFC)>1])
genes.keep = genes.keep[!genes.keep %in% genes.remove]

celltype.gep <- celltype.gep[rownames(celltype.gep) %in% genes.keep, ]
dim(celltype.gep)

gc()
```

Rename genes that are different between mm10 & mm39 (just Mymk and Mymx)
```{r}
rownames(celltype.gep)[rownames(celltype.gep)=="Tmem8c"] <- "Mymk"
rownames(celltype.gep)[rownames(celltype.gep)=="Gm7325"] <- "Mymx"

print(table(rownames(celltype.gep) %in% Features(skm.list[[1]],assay="STARsolo_collapsed")))
```


Run the deconvolution
```{r}
n.cores=50
celltype.bp <- list()
for(i in 1:length(skm.list)){
  cat("Sample: ", meta_skm$sample[i],"\n")
  celltype.bp[[i]] <-run.Ted( 
    ref.dat = t(celltype.gep), #cell types x genes
    X=t(as.matrix(
      GetAssayData(skm.list[[i]],slot="counts",assay="STARsolo_collapsed")
      )), # spot to deconvolve
    cell.type.labels=colnames(celltype.gep),
    input.type = "GEP", 
    n.cores=n.cores
  )
  # gc()
}
```

Add outputs to Seurat objects
```{r}
for(i in 1:length(skm.list)){ # add theta (composition) values as an Assay
  colnames(celltype.bp[[i]]$res$final.gibbs.theta) <- stringr::str_replace_all(rownames(celltype.bp[[i]]$para$input.phi),pattern = " ", replacement="_")
  skm.list[[i]][["celltype.bp"]] <- CreateAssayObject(data=t(celltype.bp[[i]]$res$final.gibbs.theta))
}
```

## Save skm.list
```{r}
save(
  skm.list,
  file="/workdir/dwm269/totalRNA/spTotal/robjs/skm_vis_BP_scMuscle_v2.RData"
)
```


## Plot predicted cell type distributions
```{r fig.height=10, fig.width=8, message=FALSE, warning=FALSE}
# Plot out BP theta values
tmp.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
  stringr::str_remove_all(pattern = "_TA") %>%
  stringr::str_remove_all(pattern = "-SkM")

suppressMessages(
  visListPlot(
    seu.list = skm.list,
    assay = "celltype.bp",
    sample.titles = tmp.titles,
    pt.size = 0.6,
    nrow = 1,
    min.value = 10^-2,
    colormap.same.scale = T,
    colormap = rev(RColorBrewer::brewer.pal(n=11,name="Spectral")),
    # features=GetAssayData(skm.list[[1]],assay="celltype.bp")%>% rownames()%>%unlist()
    features = c(
      "FAPs-(Pro-remodeling)",
      "Monocyte-(Patrolling)",
      "Quiescent-MuSCs",
      "Activated-MuSCs",
      "Committed-Myoblasts",
      "Fusing-Myocytes",
      "Myonuclei-(Type-IIb)",
      "Myonuclei-(Type-IIx)"
    )
  )&coord_fixed(1/1.6)
)
```

# Merge and cluster based on BP theta values
```{r}
skm.merged <- merge(
  x=skm.list[[1]],
  y=skm.list[2:length(skm.list)],
  add.cell.ids = meta_skm$sample
)
```

preprocess using the BayesPrism outputs
```{r}
skm.merged@active.assay <- "celltype.bp"
skm.merged <- skm.merged %>%
  ScaleData() %>%
  RunPCA(
    assay="celltype.bp",
    features=Features(skm.merged,assay="celltype.bp"),
    reduction.key = "BayesPrism.PC_",
    verbose=F,
    npcs=10
  )
```

Plot clusters on PCA
```{r fig.height=4, fig.width=12}
lapply(
  c("injury.zones","timepoint","polyA"),
  FUN=function(GROUP) DimPlot(
  skm.merged,
  cells=sample(Cells(skm.merged)),
  reduction = "pca",
  group.by=GROUP
)+
  scTheme$umap
)%>% wrap_plots()
```

Save ribj
```{r}
save(
  skm.merged,
  file="/workdir/dwm269/totalRNA/spTotal/robjs/skm_merged_bp_v1.RData"
)
```


# Figure S#
```{r}
# BayesPrism maps
tmp.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
  stringr::str_remove_all(pattern = "_TA") %>%
  stringr::str_remove_all(pattern = "-SkM")

bp.map <- visListPlot(
    seu.list = skm.list,
    assay = "celltype.bp",
    sample.titles = tmp.titles,
    pt.size = 0.6,
    nrow = 1,
    min.value = 10^-2,
    colormap.same.scale = T,
    colormap = rev(RColorBrewer::brewer.pal(n=11,name="Spectral")),
    # features=GetAssayData(skm.list[[1]],assay="celltype.bp")%>% rownames()%>%unlist()
    features = c(
      "FAPs-(Pro-remodeling)",
      "Monocyte-(Patrolling)",
      "Quiescent-MuSCs",
      "Activated-MuSCs",
      "Committed-Myoblasts",
      "Fusing-Myocytes",
      "Myonuclei-(Type-IIb)",
      "Myonuclei-(Type-IIx)"
    )
  )&coord_fixed(1/1.6)
```

PCA plots
```{r fig.height=4, fig.width=17.5}
pca.list <- lapply(
  c("injury.zones","timepoint","polyA"),
  FUN=function(GROUP) DimPlot(
  skm.merged,
  cells=sample(Cells(skm.merged)),
  reduction = "pca",
  group.by=GROUP
)+
  scTheme$umap+
  coord_fixed()
)

pca.list[[1]] <- pca.list[[1]] + scale_color_manual(values=mckolors$primary)
pca.list[[2]] <- pca.list[[2]] + scale_color_manual(values=mckolors$simpsons_ggsci[c(3:6)])
pca.list[[3]] <- pca.list[[3]] + scale_color_manual(values=mckolors$txg[c(1,4)])

wrap_plots(
  pca.list,
  nrow=1
)
```

## Look at average BP theta values for each sample/cell type
```{r}
i = meta_skm$timepoint>0
tmp.df <- lapply(
  skm.list[i],
  FUN = function(SEU){
    bp <- GetAssayData(SEU, assay="celltype.bp")
    
    data.frame(
      celltype = rownames(bp),
      mean = rowMeans(bp),
      median = apply(bp,1,median),
      sd = apply(bp,1,sd),
      injury = rep(SEU$timepoint[1], nrow(bp)),
      method = rep(SEU$polyA[1], nrow(bp))
    ) %>%
      # reshape2::melt(
      #   id.vars=c("celltype", "injury","method")
      # )%>%
      return()
  }
) %>%
  do.call(what=rbind)

bp.df <- list()
for(i in unique(tmp.df$injury)){
  for(j in unique(tmp.df$celltype)){
    bp.df[[paste(j,i,sep="_")]] <- data.frame(
      injury = as.character(i),
      celltype = j,
      mean_Visium = tmp.df$mean[tmp.df$celltype == j & tmp.df$injury == i & tmp.df$method == "ctrl"],
      mean_STRS= tmp.df$mean[tmp.df$celltype == j & tmp.df$injury == i & tmp.df$method =="yPAP"],
      sd_Visium = tmp.df$sd[tmp.df$celltype == j & tmp.df$injury == i & tmp.df$method == "ctrl"],
      sd_STRS= tmp.df$sd[tmp.df$celltype == j & tmp.df$injury == i & tmp.df$method =="yPAP"]
    )
  }
}
bp.df <- do.call(rbind,bp.df)
```

plot
```{r fig.height=12, fig.width=12, message=FALSE, warning=FALSE, paged.print=FALSE}
ggplot(
  bp.df,
  aes(
    x = mean_Visium,
    y = mean_STRS,
    color = celltype,
    fill = injury
  )
) +
  # geom_abline(
  #   color="black"
  # )+
  geom_smooth(
    method="lm",
    color="black"
  )+
  ggpmisc::stat_poly_eq(
    # data = tmp.df[tmp.df$STRS>0,],
    method = "lm",
    aes(
      label = paste(..eq.label.., ..rr.label.., sep = "~~~"),
      group=injury
    ),
    # color="blue",
    # label.y = 30,
    parse = TRUE
  ) +
  geom_point(
    alpha = 0.8,
    shape=21,
    size = 3,
    stroke = 1
  )+
  geom_errorbar(
    aes(
      xmin = sd_Visium,
      ymin = sd_STRS,
      xmax = sd_Visium,
      ymax = sd_STRS
    )
  )+
  # ggrepel::geom_label_repel(
  #   aes(
  #     label=celltype
  #   ),
  #   fill=NA,
  #   color="black"
  # )+
  coord_fixed()+
  scale_color_manual(values=mckolors$polychrome)+
  scale_fill_manual(values=viridis(option = "magma",n = 7)[c(2,5,7)])+
  scTheme$scatter+
  facet_wrap(facets='injury')
```


# Session Info
```{r}
sessionInfo()
```