---
title: "R Notebook"
output: html_notebook
---

## Setup
```{r}
# library(ggbio)
library(dplyr, quietly = T)
library(data.table, quietly = T)
library(GenomicRanges, quietly = T)
library(GenomicFeatures, quietly = T)
library(OrganismDbi, quietly = T)
library(plotgardener, quietly = T)
```


## Read in metadata
```{r}
meta_heart <- read.csv("/workdir/dwm269/totalRNA/spTotal/resources/metadata_sheets/meta_sheet_visium.csv")
meta_heart <- meta_heart[meta_heart$tissue == "heart", ]
meta_heart$sample
meta_heart$Sample = c("Control-Mock", "Control-Infected", "yPAP-Mock", "yPAP-Infected")

```


## Read in mouse gene info
```{r}
gene_info <- read.csv("/workdir/dwm269/totalRNA/spTotal/resources/gene_lists/biomaRt_GRCm39_v1.csv")
gene_info$mgi_symbol <- ifelse((gene_info$mgi_symbol == ""), gene_info$ensembl_gene_id, gene_info$mgi_symbol)
gene_info <- gene_info[!duplicated(gene_info$mgi_symbol), ]
table(gene_info$gene_biotype)
gene_info$biotype_groups = gene_info$gene_biotype
gene_info$biotype_groups[gene_info$gene_biotype %in% grep(pattern = "pseudogene", x = unique(gene_info$biotype_groups), value = TRUE)] <- "pseudogene"
gene_info$biotype_groups[gene_info$gene_biotype %in% grep(pattern = "IG_", x = unique(gene_info$biotype_groups), value = TRUE)] <- "IG_genes"
gene_info$biotype_groups[gene_info$gene_biotype %in% grep(pattern = "TR_", x = unique(gene_info$biotype_groups), value = TRUE)] <- "TCR_genes"
table(gene_info$biotype_groups)
```


```{r}
## Read in mouse gene info
gene_info_gtf <- read.csv("./../resources/gene_lists/GRCm39_GENCODEm28_gene_info_gtf.tsv", sep = "\t")
# gene_info_gtf$GeneSymbol[duplicated(gene_info_gtf$GeneSymbol)]
# gene_info_gtf$mgi_symbol <- ifelse((gene_info_gtf$mgi_symbol == ""), gene_info_gtf$ensembl_gene_id, gene_info_gtf$mgi_symbol)
# gene_info_gtf <- gene_info_gtf[!duplicated(gene_info_gtf$mgi_symbol), ]
table(gene_info_gtf$Biotype)
gene_info_gtf$biotype_groups = gene_info_gtf$Biotype
gene_info_gtf$biotype_groups[gene_info_gtf$biotype_groups %in% grep(pattern = "pseudogene", x = unique(gene_info_gtf$biotype_groups), value = TRUE)] <- "pseudogene"
gene_info_gtf$biotype_groups[gene_info_gtf$biotype_groups %in% grep(pattern = "IG_", x = unique(gene_info_gtf$biotype_groups), value = TRUE)] <- "IG_genes"
gene_info_gtf$biotype_groups[gene_info_gtf$biotype_groups %in% grep(pattern = "TR_", x = unique(gene_info_gtf$biotype_groups), value = TRUE)] <- "TCR_genes"
table(gene_info_gtf$biotype_groups)

```


```{r}
grcm39.chromInfo <- fread("/workdir/dwm269/genomes/mm39_all/STAR_GRCm39_GENCODEM28_REOT1L-as/chrNameLength.txt")
colnames(grcm39.chromInfo) <- c("chrom", "length")
T1L_chrom <- grcm39.chromInfo[grcm39.chromInfo$chrom %in% grep("T1L", x = grcm39.chromInfo$chrom, value = T),]
```


```{r}
grcm39_and_REOT1L.txdb <- makeTxDbFromGFF(
  file="/workdir/dwm269/genomes/mm39_all/STAR_GRCm39_GENCODEM28_REOT1L-as/GRCm39_ReoT1L_merged_genes.gtf",
  format="auto", #c("auto", "gff3", "gtf"),
  dataSource="GENCODE_M28",
  organism="Mus musculus",
  taxonomyId=NA,
  circ_seqs=NULL,
  chrominfo=grcm39.chromInfo,
  # miRBaseBuild=NA,
  # metadata=NULL,
  dbxrefTag="gene_id"
)
columns(grcm39_and_REOT1L.txdb)
keytypes(grcm39_and_REOT1L.txdb)

```

```{r}
grcm39_and_REOT1L.orgdb <- makeOrganismDbFromTxDb(
  txdb = grcm39_and_REOT1L.txdb
)
# grcm39.orgdb <- makeOrganismDbFromBiomart(
#   biomart="ENSEMBL_MART_ENSEMBL",
#   dataset = "mmusculus_gene_ensembl"
# )

```

```{r}
grcm39_and_REOT1L.assembly <- assembly(
  Genome="GRCm39",
  TxDb=grcm39_and_REOT1L.txdb,
  OrgDb=grcm39_and_REOT1L.orgdb
  # gene.id.column="",
  # display.column =""
)
```



# Set plotting params and load bigWigs 
```{r}
# Plotting parameters
gene_info_gtf[gene_info_gtf$Chromosome == "T1LReoS1",]

gene_symbol = "T1LReoS4"
# gene_symbol = "Malat1"
params <- pgParams(
  chrom=gene_info_gtf$Chromosome[gene_info_gtf$GeneSymbol == gene_symbol],
  chromstart=gene_info_gtf$Start[gene_info_gtf$GeneSymbol == gene_symbol],
  chromend=gene_info_gtf$End[gene_info_gtf$GeneSymbol == gene_symbol],
  # chrom="chr8", chromstart=85833066, chromend=85851608,#Zfp791
  # gene="Myod1", geneBuffer = 500,
  assembly = grcm39_and_REOT1L.assembly,
  x = 1.25,
  just = c("left", "top"),
  width = 2, 
  height=16,
  length = 5, 
  default.units = "inches"
)


trackHeight <- 1
small.font<-6
big.font<-8
```

# Load in bigWigs
```{r}

bw.plus.list <- lapply(
  paste0(meta_heart$data.dir.STARsolo.REO, "/Aligned.sortedByCoord.dedup.out_plus.bw"),
  FUN = function(PATH) readBigwig(PATH,params = params)#,strand = "+")
)
bw.minus.list <- lapply(
  paste0(meta_heart$data.dir.STARsolo.REO, "/Aligned.sortedByCoord.dedup.out_minus.bw"),
  FUN = function(PATH) readBigwig(PATH,params = params)#, strand="-")
)

bw.plus.list[[1]]

bw.plus.list[[2]] <- readBigwig("./../data/REO/T1L_D7PI_STARsolo/reovirus.dedup.plus.bw", params = params)
bw.minus.list[[2]] <- readBigwig("./../data/REO/T1L_D7PI_STARsolo/reovirus.dedup.minus.bw", params = params)
bw.minus.list[[2]]$score <- -1*bw.minus.list[[2]]$score

bw.plus.list[[4]] <- readBigwig("./../data/REO/Vis_yPAP_3B_STARsolo/reovirus.dedup.plus.bw", params = params)
bw.minus.list[[4]] <- readBigwig("./../data/REO/Vis_yPAP_3B_STARsolo/reovirus.dedup.minus.bw", params = params)
bw.minus.list[[4]]$score <- -1*bw.minus.list[[4]]$score

```

## Plot xGen data 
```{r}

xgen.bw.plus.list <- bw.plus.list
xgen.bw.minus.list <- bw.minus.list

xgen.bw.plus.list[[2]] <- readBigwig("./../data/REO/T1L_D7PI_xGen_STARsolo/reovirus.dedup.plus.bw", params = params)
xgen.bw.minus.list[[2]] <- readBigwig("./../data/REO/T1L_D7PI_xGen_STARsolo/reovirus.dedup.minus.bw", params = params)

# xgen.bw.plus.list[[2]]$score <- log10(xgen.bw.plus.list[[2]]$score + 1)
# xgen.bw.minus.list[[2]]$score <- log10(xgen.bw.minus.list[[2]]$score + 1)

xgen.bw.minus.list[[2]]$score <- -1*xgen.bw.minus.list[[2]]$score

xgen.bw.plus.list[[4]] <- readBigwig("./../data/REO/Vis_yPAP_3B_xGen_STARsolo/reovirus.dedup.plus.bw", params = params)
xgen.bw.minus.list[[4]] <- readBigwig("./../data/REO/Vis_yPAP_3B_xGen_STARsolo/reovirus.dedup.minus.bw", params = params)

# xgen.bw.plus.list[[4]]$score <- log10(xgen.bw.plus.list[[4]]$score + 1)
# xgen.bw.minus.list[[4]]$score <- log10(xgen.bw.minus.list[[4]]$score + 1)

xgen.bw.minus.list[[4]]$score <- -1*xgen.bw.minus.list[[4]]$score
```



# Plotting starts
`Mir6236`
```{r fig.height=5, fig.width=8, message=FALSE, warning=FALSE}
# pdf(
#   file="/workdir/dwm269/totalRNA/spTotal/figures/plotgardener/pg_Mir6236.pdf",
#   width = 5, 
#   height= 17
# )

pageCreate(
    width = 8,
    height = 5,
    default.units = "inches",
    showGuides = F, 
    xgrid = 0, 
    ygrid = 0,
    params = params
)

plotText(
    label = gene_symbol,
    fonsize = small.font,
    fontcolor = "black",
    x=0.05,
    y = 0.20,
    just = c("left", "center"),
    params = params,
    default.units = "inches"
  )

# single-cell
n_above=0.1 #number of plots above these

pos_max = lapply(bw.plus.list, FUN=function(BW) BW$score) %>% unlist() %>% max()
neg_min = lapply(bw.minus.list, FUN=function(BW) BW$score) %>% unlist() %>% min()
xgen_pos_max = lapply(xgen.bw.plus.list, FUN=function(BW) BW$score) %>% unlist() %>% max()
xgen_neg_min = lapply(xgen.bw.minus.list, FUN=function(BW) BW$score) %>% unlist() %>% min()
limit = max(pos_max, abs(neg_min))
limit_xgen = max(xgen_pos_max, abs(xgen_neg_min))
tmp.range.plus <- c(0, limit) #Range for normalization
tmp.range.minus <- c((-1*limit), 0)
xgen.tmp.range.plus <- c(0, limit_xgen) #Range for normalization
xgen.tmp.range.minus <- c((-1*limit_xgen), 0)

for(i in 1:length(bw.plus.list)){
  bw.plus.path = bw.plus.list[[i]]
  bw.minus.path = bw.minus.list[[i]]
  plotText(
    label = meta_heart$Sample[i],
    fonsize = small.font,
    x=0.05,
    fontcolor = "black",
    y = (i+n_above)*trackHeight,
    just = c("left", "center"),
    params = params,
    default.units = "inches"
  )
  plotSignal(
    data = bw.plus.path,
    params = params,
    fill = "#37a7db",
    linecolor = "#37a7db",
    y = (i+n_above)*trackHeight-(trackHeight/2),
    height = (trackHeight/2),
    negData = TRUE,
    range=tmp.range.plus,
    scale=T
  )
  plotSignal(
    data = bw.minus.path,
    params = params, 
    fill = "#fc0362",
    linecolor = "#fc0362",
    y = (i+n_above)*trackHeight,
    height = (trackHeight/2),
    negData = TRUE,
    range=tmp.range.minus,
    scale=T
  )
  
  # xGen ploting starts here
  bw.plus.path = xgen.bw.plus.list[[i]]
  bw.minus.path = xgen.bw.minus.list[[i]]
  
  plotSignal(
    data = bw.plus.path,
    params = params,
    fill = "#37a7db",
    linecolor = "#37a7db",
    x = 3.4,
    y = (i+n_above)*trackHeight-(trackHeight/2),
    height = (trackHeight/2),
    negData = TRUE,
    range=xgen.tmp.range.plus,
    scale=T
  )
  plotSignal(
    data = bw.minus.path,
    params = params, 
    fill = "#fc0362",
    linecolor = "#fc0362",
    x = 3.4,
    y = (i+n_above)*trackHeight,
    height = (trackHeight/2),
    negData = TRUE,
    range=xgen.tmp.range.minus,
    scale=T
  )
}
# genesPlot <- plotGenes(
#     params = params,
#     # geneHighlights = data.frame(
#     #   "gene" = c("Mir6236"),
#     #   "color" = c("#225EA8")
#     # ),
#     #   geneBackground = "grey",
#     # fill="black",
#     y = 5
# )

## Plot genome label
# plotGenomeLabel(
#   # plot = genesPlot,
#   params = params,
#   fontsize = 8,
#   y = 6, 
#   scale = "Kb"
# )

```



```{r}

print(gene_info_gtf[gene_info_gtf == "Gapdh",])

columns(grcm39_and_REOT1L.txdb)
keytypes(grcm39_and_REOT1L.txdb)

# keys <- paste("T1LReovirus",  sprintf("%02d", 1:10), sep = "")
genePos_reovirus <- select(grcm39_and_REOT1L.txdb,
                  keys = keys,
                  keytype = "GENEID",
                  columns = columns(grcm39_and_REOT1L.txdb)
)

keys <- c("ENSMUSG00000057666.19")
genePos_host <- select(grcm39_and_REOT1L.txdb,
                  keys = keys,
                  keytype = "GENEID",
                  columns = columns(grcm39_and_REOT1L.txdb)
)

geneRanges <- GRanges(genePos$EXONCHROM, 
                      ranges = IRanges(genePos$EXONSTART, genePos$EXONEND), 
                      GENEID = genePos$GENEID, strand = genePos$EXONSTRAND)

exo <- exonsBy(grcm39_and_REOT1L.txdb, "gene") 
generegion <- exo[keys]
```


```{r}
library(GenomicAlignments)

my.reads <- readGAlignments(file="/workdir/dwm269/totalRNA/data/STARsolo/GRCm39_GENCODEM27_Visium/REO/Vis_yPAP_3B_STARsolo/Aligned.sortedByCoord.out.bam", param=ScanBamParam(which=unlist(generegion), what=c("seq", "mapq","flag")), use.names = T)
my.reads

bam = readGappedReads(file="/workdir/dwm269/totalRNA/data/STARsolo/GRCm39_GENCODEM27_Visium/REO/Vis_yPAP_3B_STARsolo/Aligned.sortedByCoord.out.bam",
                      param=ScanBamParam(which=unlist(generegion)), use.names = TRUE)
bam
```

```{r}
reads.Mod <- autoplot(my.reads, geom = "rect") 
Coverage.Mod <- autoplot(my.reads, geom = "line", stat = "coverage", facets = seqnames)
geneMod <- autoplot(grcm39_and_REOT1L.txdb, which = generegion$T1LReovirus01, names.expr = "gene_name")
tracks(Reads=geneMod, Coverage=Coverage.Mod, Gene=reads.Mod, heights = c(0.1, 0.5, 0.5))

bamfile <- system.file("extdata", "/workdir/dwm269/totalRNA/data/STARsolo/GRCm39_GENCODEM27_Visium/REO/Vis_yPAP_3B_STARsolo/Aligned.sortedByCoord.out.bam",    package="biovizBase")
## need to set use.names = TRUE
ga <- readGAlignments(bamfile,
                      param = ScanBamParam(which = unlist(generegion)),
                      use.names = TRUE)
autoplot(bam, stat = "coverage") 
```

