# Comparison of standard visium to polyA visium in Reovirus-infected heart
This analysis uses a reference genome comprised of GRCm39/GENCODE_M27 and the Reovirus genome

# Session setup
## Libs, setwd
```{r message=FALSE, warning=FALSE}
library(Matrix)
library(dplyr)
library(Seurat)
library(future)
library(parallel)
library(ggplot2)
library(patchwork)
library(pals)
library(viridis)
library(data.table)
library(shades)
source("./../DWM_utils/plotting_utils/scThemes.R")
source("./../DWM_utils/sc_utils/seurat_helpers/seutils.R")
source("./../DWM_utils/sc_utils/seurat_helpers/seuplots.R")
```

## Figure settings & colors
```{r}
# fonts, sizes, etc.
small.font = 6*2
big.font = 8*2
line.width = 0.5
pt.size=0.01
pt.stroke=0.3
label.size=2

# colors
colPals <- list()
# colPals[["sample"]] <- ggsci::pal_rickandmorty()(meta_vis$sample %>% unique() %>% length())
colPals[["celltypes"]] <- as.vector(pals::polychrome())[3:32]
```

## Plot themes 
```{r}
scTheme <- scThemes(
  small.font = small.font,
  big.font = big.font,
  line.width = line.width,
  pt.size=pt.size,
  pt.stroke=pt.stroke,
  label.size=label.size
)

print(names(scTheme))

```

```{r}
mckolors <- read.csv("/home/dwm269/DWM_utils/plotting_utils/McKolors_v1.csv") %>% 
  as.list() %>%
  lapply(
    FUN=function(X) X[X!=""]
  )
mckolors
```
# Load in data
## Read in metadata
```{r}
meta_heart <- read.csv("/workdir/dwm269/totalRNA/spTotal/resources/meta_sheet.csv")
meta_heart <- meta_heart[meta_heart$tissue=="heart",]

print(meta_heart)
```

## Read in count mats (spaceranger) and initialize Seurat objects
```{r}
heart.list <- list()
for(i in 1:nrow(meta_heart)){ 
  if(file.exists(paste0(meta_heart$data.dir.spaceranger.REO[i], '/outs/filtered_feature_bc_matrix'))){ 
    cat("Reading #",i, ": ", meta_heart$data.dir.spaceranger.REO[i], '...\n')
    heart.list[[i]] <- Seurat::Load10X_Spatial(
      data.dir = paste0(meta_heart$data.dir.spaceranger.REO[i], '/outs'),
      filter.matrix = T
    )
    
    # Strip "-1" suffix from cell barcode (colnames)
    if(stringr::str_detect(Cells(heart.list[[i]]),pattern = "-1")){
      cat("     Stripping suffix from spot barcodes...\n")
      heart.list[[i]] <- RenameCells(
        object = heart.list[[i]],
        new.names = stringr::str_remove_all(Cells(heart.list[[i]]),pattern="-1")
      )
    }
    
    cat("    Done!\n")
  }else{
    cat("Data not found for # ", i, " (", meta_heart$data.dir.spaceranger.REO[i], ")", "\n")
  }
}

```
## Add metadata 
```{r}
for(i in 1:nrow(meta_heart)){
  for(j in colnames(meta_heart)){
    heart.list[[i]]@meta.data[[j]] <- meta_heart[i,j]
  }
}
```

## Read in matrices - STARsolo 
```{r}
mat.list <- list()
for(i in 1:nrow(meta_heart)){ 
  if(file.exists(paste0(meta_heart$data.dir.STARsolo.REO[i], "/Solo.out/GeneFull/raw/"))){ 
    cat("Reading #",i, ": ", meta_heart$data.dir.STARsolo.REO[i], ' \n')
    mat.list[[i]] <- Read10X(
      data.dir = paste0(meta_heart$data.dir.STARsolo.REO[i], "/Solo.out/GeneFull/raw/")
    )
  }else{
    cat("Data not found for # ", i, " (", meta_heart$data.dir.STARsolo.REO[i], ")", "\n")
  }
}

cat(sum(unlist(lapply(mat.list, ncol))),"spots (total) loaded...\n")
```

## Add STARsolo counts as new assay
```{r}
for(i in 1:length(heart.list)){
  cat(paste0("Adding STARsolo for ", meta_heart$sample[i]),"\n")
  tmp <- CreateAssayObject(
    counts=mat.list[[i]],
    min.cells = 3
  )
  print(dim(tmp))
  
  heart.list[[i]][["STARsolo"]] <- subset(
    tmp,
    cells = Cells(heart.list[[i]])
  )
  
  heart.list[[i]]@active.assay <- "STARsolo"
}
rm(mat.list)
```


## Collapse multimappers, preprocess collapsed counts
```{r message=FALSE, warning=FALSE}
heart.list <- lapply(
  heart.list,
  FUN = function(SEU) collapseMultimappers(
    SEU,
    assay="STARsolo",
    new.assay.name = "STARsolo_collapsed",
    verbose=T
    )
) %>% lapply(
  FUN = function(SEU) seuPreProcess(
    SEU, 
    assay="STARsolo_collapsed",
    res = 0.6,
    nFeatures = 2000,
    verbose=F
  )
)
```
# biomaRt

```{r}
library(biomaRt)
if(!exists("mouse.info")){
  mouse = biomaRt::useMart(
    # biomart = "ENSEMBL_MART_MOUSE",
   biomart = "ENSEMBL_MART_ENSEMBL",
    dataset = "mmusculus_gene_ensembl"
    # host = "https://apr2020.archive.ensembl.org" ##for mm10 info
  )
  mouse.info <- biomaRt::getBM(
    attributes = c(
      "mgi_symbol", "gene_biotype","transcript_biotype", 
      # "description", "name_1006","definition_1006","external_gene_name",
      "ensembl_gene_id","ensembl_transcript_id","ensembl_peptide_id",
      "chromosome_name","transcript_length", "strand"
    ),
    uniqueRows = T,
    # host = "useast.ensembl.org",
    verbose = T,
    mart = mouse
  )
}
cat(
  paste("Using", length(unique(mouse.info$mgi_symbol)), "genes, and ",length(unique(mouse.info$gene_biotype)),"biotypes...\n")
)
```

## Add kallisto counts
```{r}
library(SeuratDisk)
# convert h5ad to h5seurat
for(i in 1:nrow(meta_heart)){
  tmp.h5 <- paste0(meta_heart$data.dir.kallisto.REO[i],"/kb_out/counts_unfiltered/adata")
  if(!file.exists(paste0(tmp.h5,".h5seurat"))){
    message(paste0("Converting h5seurat for",meta_heart$capture.area[i],"...\n"))
    Convert(
      paste0(tmp.h5,".h5ad"),
      dest = "h5seurat",
      overwrite = TRUE
    )
  }

  message(paste0("Loading & cleaning kallisto output for ",meta_heart$sample[i],"...\n"))
  
  # Load h5seurat and subset out counts for sample
  tmp.seu <- LoadH5Seurat(
    paste0(tmp.h5,".h5seurat"),
  ) %>% subset( # select cells
    cells=Cells(heart.list[[i]])
  )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  # collapse multimappers
  tmp.seu <- collapseMultimappers(
    tmp.seu,
    assay="RNA",
    new.assay.name = "RNA",
    verbose=F
  )
  
  # convert ensembl IDs to MGI gene IDs
  tmp.mat <- GetAssayData(tmp.seu,assay="RNA")
  rownames(tmp.mat) <- ens2gene(
    ens=rownames(tmp.mat),
    biomart.info = mouse.info,
    ens.colname = "ensembl_gene_id",
    gene.colname = "mgi_symbol",
    force.unique = T,
    ncores=24,
    verbose=F
  )
  
  # add kallisto counts to vis_list
  heart.list[[i]][["kallisto"]] <- CreateAssayObject(
    counts = tmp.mat,
    min.cells = 1
  )
  
  rm(tmp.seu)
  rm(tmp.mat)
}
gc()
```


### Collapse multimappers and preprocess collapsed counts (kallisto)
```{r warning=FALSE}
# collapse multimappers
heart.list <- lapply(
  heart.list,
  FUN = function(SEU) collapseMultimappers(
    SEU,
    assay="kallisto",
    new.assay.name = "kallisto_collapsed",
    verbose=T
    )
)

# preprocess
heart.list <- lapply(
  heart.list,
  FUN = function(SEU) seuPreProcess(
    SEU, 
    assay="kallisto_collapsed",
    nfeatures = 10000,
    res=0.6,
    verbose=F
  )
)
```

# Save list of seurat objects, and each object as an adata individually
```{r}
save(
  heart.list,
  file = "/workdir/mm2937/polyAtotal/robjs/heart_list_v1.RData"
)
cat("Done.\n\n")
```


# Session Info
```{r}
sessionInfo()
```