# Comparison of standard visium to polyA visium in Reovirus-infected heart
This analysis uses a reference genome comprised of GRCm39/GENCODE_M27 and the Reovirus genome

# Session setup
## Libs, setwd
```{r message=FALSE, warning=FALSE}
library(Matrix)
library(dplyr)
library(Seurat)
library(future)

library(ggplot2)
library(patchwork)
library(pals)
library(viridis)
library(data.table)
library(shades)
source("/home/dwm269/DWM_utils/plotting_utils/scThemes.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seutils.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seuplots.R")
```

## Figure settings & colors
```{r}
# fonts, sizes, etc.
small.font = 6*2
big.font = 8*2
line.width = 0.5
pt.size=0.01
pt.stroke=0.3
label.size=2
```

## Plot themes 
```{r}
scTheme <- scThemes(
  small.font = small.font,
  big.font = big.font,
  line.width = line.width,
  pt.size=pt.size,
  pt.stroke=pt.stroke,
  label.size=label.size
)

print(names(scTheme))
```

```{r}
mckolors <- read.csv("/home/dwm269/DWM_utils/plotting_utils/McKolors_v1.csv") %>% 
  as.list() %>%
  lapply(
    FUN=function(X) X[X!=""]
  )
mckolors
```
# Load in data
## Read in metadata
```{r}
meta_heart <- read.csv("/workdir/dwm269/totalRNA/spTotal/resources/meta_sheet_visium.csv")
meta_heart <- meta_heart[meta_heart$tissue=="heart",]

# meta_heart <- meta_heart[c(1,3,4),]

print(meta_heart)
```

## Get gene biotype info from the GENCODE annotations
`gtf.info` is just a reformatted .tsv of the GENCODE M28 annotations .gtf
```{r}
gtf.info <- read.csv(
  file="/workdir/dwm269/totalRNA/spTotal/resources/gene_lists/GRCm39_GENCODEm28_gene_info_gtf.tsv",
  sep = "\t"
)

# Remove period/decimal suffixes from ensembl IDs
gtf.info$GeneID <- strsplit(
  x=gtf.info$GENEID,
  split = "\\."
)%>%
  lapply(
    FUN=function(X) X[[1]]
  )%>%
  unlist()

# Add ".1" to suffix of Reovirus transcripts for copatibility with `ens2gene()`
gtf.info$GENEID[grep(gtf.info$GeneSymbol,pattern="Reo")] <- paste0(
  gtf.info$GENEID[grep(gtf.info$GeneSymbol,pattern="Reo")],
  ".1"
)

cat(
  paste("Using", length(unique(gtf.info$GeneSymbol)), "genes, and ",length(unique(gtf.info$Biotype)),"biotypes...\n")
)

tail(gtf.info)
```

## Read in count mats (spaceranger) and initialize Seurat objects
```{r}
heart.list <- list()
for(i in 1:nrow(meta_heart)){ 
  if(file.exists(paste0(meta_heart$data.dir.spaceranger.REO[i], '/outs/filtered_feature_bc_matrix'))){ 
    cat("Reading #",i, ": ", meta_heart$data.dir.spaceranger.REO[i], '...\n')
    heart.list[[i]] <- Seurat::Load10X_Spatial(
      data.dir = paste0(meta_heart$data.dir.spaceranger.REO[i], '/outs'),
      filter.matrix = T
    )
    
    # Strip "-1" suffix from cell barcode (colnames)
    if(stringr::str_detect(Cells(heart.list[[i]]),pattern = "-1")){
      cat("     Stripping suffix from spot barcodes...\n")
      heart.list[[i]] <- RenameCells(
        object = heart.list[[i]],
        new.names = stringr::str_remove_all(Cells(heart.list[[i]]),pattern="-1")
      )
    }
    
    cat("    Done!\n")
  }else{
    cat("Data not found for # ", i, " (", meta_heart$data.dir.spaceranger.REO[i], ")", "\n")
  }
}

```
## Add metadata 
```{r}
for(i in 1:nrow(meta_heart)){
  for(j in colnames(meta_heart)){
    heart.list[[i]]@meta.data[[j]] <- meta_heart[i,j]
  }
}
```

## Read in matrices - STARsolo 
```{r}
mat.list <- list()
for(i in 1:nrow(meta_heart)){ 
  if(file.exists(paste0(meta_heart$data.dir.STARsolo.REO[i], "/Solo.out/GeneFull/raw/"))){ 
    cat("Reading #",i, ": ", meta_heart$data.dir.STARsolo.REO[i], ' \n')
    mat.list[[i]] <- Read10X(
      data.dir = paste0(meta_heart$data.dir.STARsolo.REO[i], "/Solo.out/GeneFull/raw/")
    )
  }else{
    cat("Data not found for # ", i, " (", meta_heart$data.dir.STARsolo.REO[i], ")", "\n")
  }
}

cat(sum(unlist(lapply(mat.list, ncol))),"spots (total) loaded...\n")
```

## Add STARsolo counts as new assay
```{r}
for(i in 1:length(heart.list)){
  cat(paste0("Adding STARsolo for ", meta_heart$sample[i]),"\n")
  tmp <- CreateAssayObject(
    counts=mat.list[[i]],
    min.cells = 3
  )
  print(dim(tmp))
  
  heart.list[[i]][["STARsolo"]] <- subset(
    tmp,
    cells = Cells(heart.list[[i]])
  )
  
  heart.list[[i]]@active.assay <- "STARsolo"
}
rm(mat.list)
```

## Collapse multimappers, preprocess collapsed counts
```{r message=FALSE, warning=FALSE}
heart.list <- lapply(
  heart.list,
  FUN = function(SEU) collapseMultimappers(
    SEU,
    assay="STARsolo",
    new.assay.name = "STARsolo_collapsed",
    verbose=T
    )
) %>% lapply(
  FUN = function(SEU) seuPreProcess(
    SEU, 
    assay="STARsolo_collapsed",
    res = 0.6,
    verbose=F
  )
)
```
## Add kallisto counts
```{r message=TRUE, warning=FALSE}
for(i in 1:nrow(meta_heart)){
  message(paste0("Loading & cleaning kallisto output for ",meta_heart$sample[i],"..."))
  
  tmp.seu <- ReadMtx(
    mtx=paste0(meta_heart$data.dir.kallisto.REO[i],"/kb_standard/counts_unfiltered/output.mtx"),
    cells=paste0(meta_heart$data.dir.kallisto.REO[i],"/kb_standard/counts_unfiltered/output.barcodes.txt"),
    features=paste0(meta_heart$data.dir.kallisto.REO[i],"/kb_standard/counts_unfiltered/output.genes.txt"),
    feature.column=1,
    mtx.transpose = T
  ) %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(heart.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )

  # convert ensembl IDs to MGI gene IDs
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  rownames(tmp.mat) <- ens2gene(
    ens=rownames(tmp.mat),
    biomart.info = gtf.info,
    ens.colname = "GENEID",
    # ens.colname = "GeneID", # ensembl IDs, no periods
    gene.colname = "GeneSymbol",
    force.unique = T,
    ncores=24,
    verbose=T
  )
  
  # add kallisto counts to vis_list
  heart.list[[i]][["kallisto"]] <- CreateAssayObject(
    counts = tmp.mat,
    min.cells = 1
  )
  
  cat("\n")
  
  rm(tmp.seu)
  rm(tmp.mat)
}
gc()
```

### Collapse multimappers and preprocess collapsed counts (kallisto)
```{r warning=FALSE}
# collapse multimappers
heart.list <- lapply(
  heart.list,
  FUN = function(SEU) collapseMultimappers(
    SEU,
    assay="kallisto",
    new.assay.name = "kallisto_collapsed",
    verbose=T
    )
)

# preprocess
heart.list <- lapply(
  heart.list,
  FUN = function(SEU) seuPreProcess(
    SEU, 
    assay="kallisto_collapsed",
    verbose=F
  )
)
```

# Add spliced/unspliced counts as new assays
Add spliced & unspliced count matrices (from kallisto la_manno workflow) to the Seurat objects as new assays
```{r}
require("R.utils")
NCORES=12
for(i in 1:length(heart.list)){
  
  # Spliced ----
  tmp.seu <- ReadMtx(
    mtx=paste0(meta_heart$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/spliced.mtx"),
    cells=paste0(meta_heart$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/spliced.barcodes.txt"),
    features=paste0(meta_heart$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/spliced.genes.txt"),
    feature.column=1,
    mtx.transpose = T
  ) %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(heart.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  print(i)
  # convert ensembl IDs to MGI gene IDs
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  
  
  # Add zeroes for spots missing reads
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  zero.cells <- Cells(heart.list[[i]])[!Cells(heart.list[[i]])%in%Cells(tmp.seu)]
  if(length(zero.cells)>0){
    zero.mat <- matrix(
      data=0,
      ncol = length(zero.cells),
      nrow = nrow(tmp.mat)
    )
    colnames(zero.mat) <- zero.cells
    rownames(zero.mat) <- rownames(zero.mat)
    
    tmp.mat <- cbind(tmp.mat,zero.mat)
    
    tmp.mat <- tmp.mat[,Cells(heart.list[[i]])] # make sure cells/spots are in the right order
    
  }
  
  
  rownames(tmp.mat) <- ens2gene(
    ens=rownames(tmp.mat),
    biomart.info = gtf.info,
    ens.colname = "GENEID", # Ensembl ID, yes periods
    # ens.colname = "GeneID", # ensembl IDs, no periods
    gene.colname = "GeneSymbol",
    force.unique = T,
    ncores=NCORES,
    verbose=T
  )
  
  # add kallisto counts to vis_list
  heart.list[[i]][["spliced"]] <- CreateAssayObject(
    counts = tmp.mat,
    min.cells = 1
  )
  
  
  # Unspliced ----
  tmp.seu <- ReadMtx(
    mtx=paste0(meta_heart$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/unspliced.mtx"),
    cells=paste0(meta_heart$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/unspliced.barcodes.txt"),
    features=paste0(meta_heart$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/unspliced.genes.txt"),
    feature.column=1,
    mtx.transpose = T
  ) %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(heart.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  # convert ensembl IDs to MGI gene IDs
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  
  # Add zeroes for spots missing reads
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  zero.cells <- Cells(heart.list[[i]])[!Cells(heart.list[[i]])%in%Cells(tmp.seu)]
  if(length(zero.cells)>0){
    zero.mat <- matrix(
      data=0,
      ncol = length(zero.cells),
      nrow = nrow(tmp.mat)
    )
    colnames(zero.mat) <- zero.cells
    rownames(zero.mat) <- rownames(zero.mat)
    
    tmp.mat <- cbind(tmp.mat,zero.mat)
    
    
    tmp.mat <- tmp.mat[,Cells(heart.list[[i]])] # make sure cells/spots are in the right order
    
  }
  
  rownames(tmp.mat) <- ens2gene(
    ens=rownames(tmp.mat),
    biomart.info = gtf.info,
    ens.colname = "GENEID", # Ensembl ID, yes periods
    # ens.colname = "GeneID", # ensembl IDs, no periods
    gene.colname = "GeneSymbol",
    force.unique = T,
    ncores=NCORES,
    verbose=T
  )
  print(i)
  
  # add kallisto counts to vis_list
  heart.list[[i]][["unspliced"]] <- CreateAssayObject(
    counts = tmp.mat,
    min.cells = 1
  )
}

rm(tmp.mat)
rm(tmp.seu)
gc()
```
Normalize the spliced/unspliced 
```{r}
heart.list <- lapply(
  heart.list,
  FUN=function(SEU){
    SEU <- NormalizeData(
      SEU,
      assay="spliced"
    )
    
    SEU <- NormalizeData(
      SEU,
      assay="unspliced"
    )
    
    return(SEU)
  }
)
message("Done.")
```

# Read in kallisto/miRbase miRNA outputs
```{r}
require("R.utils")
NCORES=20
for(i in 1:length(heart.list)){
  
  tmp.seu <- ReadMtx(
    mtx=paste0(meta_heart$data.dir.kallisto[i],"/kb_miRNA_k11/counts_unfiltered/matrix.mtx"),
    cells=paste0(meta_heart$data.dir.kallisto[i],"/kb_miRNA_k11/counts_unfiltered/barcodes.tsv"),
    features=paste0(meta_heart$data.dir.kallisto[i],"/kb_miRNA_k11/counts_unfiltered/genes.tsv"),
    feature.column=1,
    mtx.transpose = F
  ) %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(heart.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  # Add zeroes for spots missing miRNA reads
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  zero.cells <- Cells(heart.list[[i]])[!Cells(heart.list[[i]])%in%Cells(tmp.seu)]
  if(length(zero.cells)>0){
    zero.mat <- matrix(
      data=0,
      ncol = length(zero.cells),
      nrow = nrow(tmp.mat)
    )
    colnames(zero.mat) <- zero.cells
    rownames(zero.mat) <- rownames(zero.mat)
    
    tmp.mat <- cbind(tmp.mat,zero.mat)
    
    
    tmp.mat <- tmp.mat[,Cells(heart.list[[i]])] # make sure cells/spots are in the right order
    
  }
  
  # add kallisto counts to vis_list
  heart.list[[i]][["miRNA"]] <- CreateAssayObject(
    counts = tmp.mat
    # min.cells = 1
  )
}

rm(tmp.mat)
rm(tmp.seu)
gc()
```
Normalize miRNA counts
```{r}
heart.list <- lapply(
  heart.list,
  FUN=function(SEU) NormalizeData(
    SEU,
    assay = "miRNA"
  ) 
)
```

# Read in kallisto/miRbase miRNA outputs
```{r}
require("R.utils")
NCORES=8
for(i in 1:length(heart.list)){
  
  # tmp.seu <- ReadMtx(
  #   mtx=paste0(meta_heart$data.dir.kallisto[i],"/mirge3_matrix/matrix.mtx.gz"),
  #   cells=paste0(meta_heart$data.dir.kallisto[i],"/mirge3_matrix/barcodes.tsv.gz"),
  #   features=paste0(meta_heart$data.dir.kallisto[i],"/mirge3_matrix/features.tsv.gz"),
  #   feature.column=1,
  #   mtx.transpose = F
  # ) %>%
  #   CreateSeuratObject() %>%
  #   subset( # select cells
  #     cells=Cells(heart.list[[i]])
  #   )
  
  tmp.seu <- mirge.list[[i]] %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(heart.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  # Add zeroes for spots missing miRNA reads
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  zero.cells <- Cells(heart.list[[i]])[!Cells(heart.list[[i]])%in%Cells(tmp.seu)]
  if(length(zero.cells)>0){
    zero.mat <- matrix(
      data=0,
      ncol = length(zero.cells),
      nrow = nrow(tmp.mat)
    )
    colnames(zero.mat) <- zero.cells
    rownames(zero.mat) <- rownames(zero.mat)
    
    tmp.mat <- cbind(tmp.mat,zero.mat)
    
    
    tmp.mat <- tmp.mat[,Cells(heart.list[[i]])] # make sure cells/spots are in the right order
    
  }
  
  # add kallisto counts to vis_list
  heart.list[[i]][["mirge3"]] <- CreateAssayObject(
    counts = tmp.mat
    # min.cells = 1
  )
}

rm(tmp.mat)
rm(tmp.seu)
gc()
```
Normalize miRge counts
```{r}
heart.list[3:4] <- lapply(
  heart.list[3:4],
  FUN=function(SEU) NormalizeData(
    SEU,
    assay = "mirge3"
  ) 
)
```

# Add spatial locations as a reduction (for easy plotting with DimPlot)
```{r warning=FALSE}

meta_heart$h_flip<- c(
  1,-1,
  -1,-1
)
meta_heart$v_flip<- c(
  1,1,
  -1,-1
)

heart.list <- mapply(
  FUN = function(SEU, h_flip, v_flip){
    
    #Note- flipped axes, so that sections are in the correct orientation
    tmp <- as.matrix(cbind(
      SEU@images$slice1@coordinates$col * h_flip,
      SEU@images$slice1@coordinates$row * v_flip
    ))
      
    colnames(tmp) <- paste0("space_", 1:2)
    rownames(tmp) <- colnames(SEU)
    
    SEU[["space"]] <- CreateDimReducObject(
      embeddings=as.matrix(tmp),
      assay="kallisto_collapsed",
      key = "space_"
    )
    
    return(SEU)
  },
  heart.list,
  meta_heart$h_flip,
  meta_heart$v_flip
)
```

# Add Reovirus counts
## Get list of detected Reovirus gene segments
```{r}
reo.genes <- lapply(
  heart.list, 
  FUN=function(VIS) grepGenes(
    SEU=VIS,
    pattern = "T1L",
    assay="kallisto_collapsed",
    # sort.by = "ABC",
    verbose=F
    )
  ) %>% 
  unlist() %>%
  unique() %>%
  sort()
print(reo.genes)
```

## Count total number of Reovirus transcripts in each spot
```{r}
heart.list <- lapply(
  heart.list,
  FUN = function(SEU){
    if(reo.genes %in% Features(SEU,assay = "kallisto_collapsed") %>% any()){
      tmp.counts = GetAssayData(
        SEU,
        assay="kallisto_collapsed",
        slot="counts"
      )[reo.genes,]
      tmp.counts = colSums(tmp.counts)
      SEU$reo.counts <- tmp.counts
      SEU$reo.log2p1 <- log2(tmp.counts+1)
    }else{
      SEU$reo.counts <- 0
      SEU$reo.log2p1 <- 0
    }
    
    return(SEU)
  }
)
cat("Done!")
```

# Load in xGen numbers
```{r}
xGen.list <- list()
xGen.samples <- paste0(
  "/workdir/dwm269/totalRNA/data/kallisto/GRCm39_GENCODEM28_Visium/REO/",
  c(
    "T1L_D7PI_xGen", 
    # "Vis_yPAP_2C_xGen",
    "Vis_yPAP_3B_xGen"
  )
)

for(i in 1:length(xGen.samples)){
  
  tmp.seu <- ReadMtx(
    mtx=paste0(xGen.samples[i],"/kb_standard/counts_unfiltered/output.mtx"),
    cells=paste0(xGen.samples[i],"/kb_standard/counts_unfiltered/output.barcodes.txt"),
    features=paste0(xGen.samples[i],"/kb_standard/counts_unfiltered/output.genes.txt"),
    feature.column=1,
    mtx.transpose = T
  ) %>%
    CreateSeuratObject() 
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )

  # convert ensembl IDs to MGI gene IDs
  tmp.mat <- GetAssayData(tmp.seu,assay="RNA")
  rownames(tmp.mat) <- ens2gene(
    ens=rownames(tmp.mat),
    biomart.info = gtf.info,
    ens.colname = "GENEID",
    gene.colname = "GeneSymbol",
    force.unique = T,
    ncores=12,
    verbose=T
  )
  
  # add kallisto counts to xGen_list
  xGen.list[[i]] <- CreateSeuratObject(
    counts = tmp.mat,
    min.cells = 1
  )
  rm(tmp.seu)
  rm(tmp.mat)
}
gc()
cat("Done!")
```

```{r}
# Add xGen numbers 
## Control Visium + xGen ----
tmp.mat <- GetAssayData(xGen.list[[1]])
tmp.feat <- rownames(tmp.mat)[grep(rownames(tmp.mat),pattern="T1LReo")] # Find detected Reo genes
tmp.mat <- tmp.mat[tmp.feat, colnames(tmp.mat) %in% Cells(heart.list[[2]])] #only keep Reo genes and spots that pass QC

zero.spots = Cells(heart.list[[2]])[!Cells(heart.list[[2]])%in%colnames(tmp.mat)] # barcodes of *spots not detected* in xGen assay
tmp.mat <- cbind( # add zeroes for spots not detected by xGen 
  tmp.mat, 
  matrix(
    data = 0,
    nrow=nrow(tmp.mat),
    ncol=length(zero.spots),
    dimnames=list(
      rownames(tmp.mat),
      zero.spots
    )
  )%>% as.sparse()
)
tmp.mat <- tmp.mat[, Cells(heart.list[[2]])] #make sure everything is in the right order

heart.list[[2]][["xGen.kallisto"]] <- CreateAssayObject(counts = tmp.mat)
heart.list[[2]]$nCount_xGen.kallisto.log2p1 <- log2(heart.list[[2]]$nCount_xGen.kallisto+1)

# Sense transcript info
tmp.feat <- rownames(tmp.mat)[grep(rownames(tmp.mat),pattern="T1LReo")]
heart.list[[2]]$nCount_xGen.kallisto.sense <- tmp.mat[tmp.feat[!grepl(tmp.feat,pattern="-as")],] %>% colSums()
heart.list[[2]]$nCount_xGen.kallisto.sense.log2p1 <- log2(heart.list[[2]]$nCount_xGen.kallisto.sense+1)

# Antisense transcript info
tmp.feat <- tmp.feat[grep(tmp.feat,pattern="-as")]
if(length(tmp.feat)>1){
  heart.list[[2]]$nCount_xGen.kallisto.as <- tmp.mat[tmp.feat,] %>% colSums()
}else if(length(tmp.feat)==1){
  heart.list[[2]]$nCount_xGen.kallisto.as <- tmp.mat[tmp.feat,]  
}else{
  heart.list[[2]]$nCount_xGen.kallisto.as <- rep(0,ncol(heart.list[[2]]))
}
heart.list[[2]]$nCount_xGen.kallisto.as.log2p1 <- log2(heart.list[[2]]$nCount_xGen.kallisto.as+1)

## Protector + spTotal ----
tmp.mat <- GetAssayData(xGen.list[[2]])
tmp.feat <- rownames(tmp.mat)[grep(rownames(tmp.mat),pattern="T1LReo")] # Find detected Reo genes
tmp.mat <- tmp.mat[tmp.feat, colnames(tmp.mat) %in% Cells(heart.list[[4]])] #only keep Reo genes and spots that pass QC

zero.spots = Cells(heart.list[[4]])[!Cells(heart.list[[4]])%in%colnames(tmp.mat)] # barcodes of *spots not detected* in xGen assay
tmp.mat <- cbind( # add zeroes for spots not detected by xGen 
  tmp.mat, 
  matrix(
    data = 0,
    nrow=nrow(tmp.mat),
    ncol=length(zero.spots),
    dimnames=list(
      rownames(tmp.mat),
      zero.spots
    )
  )%>% as.sparse()
)
tmp.mat <- tmp.mat[, Cells(heart.list[[4]])] #make sure everything is in the right order

heart.list[[4]][["xGen.kallisto"]] <- CreateAssayObject(counts = tmp.mat)
heart.list[[4]]$nCount_xGen.kallisto.log2p1 <- log2(heart.list[[4]]$nCount_xGen.kallisto+1)

# Sense transcript info
tmp.feat <- rownames(tmp.mat)[grep(rownames(tmp.mat),pattern="T1LReo")]
heart.list[[4]]$nCount_xGen.kallisto.sense <- tmp.mat[tmp.feat[!grepl(tmp.feat,pattern="-as")],] %>% colSums()
heart.list[[4]]$nCount_xGen.kallisto.sense.log2p1 <- log2(heart.list[[4]]$nCount_xGen.kallisto.sense+1)

# Antisense transcript info
tmp.feat <- rownames(tmp.mat)[grep(rownames(tmp.mat),pattern="T1LReo")]
heart.list[[4]]$nCount_xGen.kallisto.as <- tmp.mat[tmp.feat[grep(tmp.feat,pattern="-as")],] %>% colSums()
heart.list[[4]]$nCount_xGen.kallisto.as.log2p1 <- log2(heart.list[[4]]$nCount_xGen.kallisto.as+1)
```

# Spot clustering
```{r}
#TODO
```

# Save list of seurat objects, and each object as an adata individually
```{r}
save(
  heart.list,
  file = "/workdir/dwm269/totalRNA/spTotal/robjs/heart_list_v3.RData" #v3 - M28 updates & cleaned up Reo counts (antisense)
)
cat("Done.\n\n")
```
# Merge Seurat objects for some plots
```{r}
if(!exists("heart.merged")){
  heart.merged <- merge(
    x=heart.list[[1]],
    y=heart.list[c(2:length(heart.list))],
    add.cell.ids = meta_heart$sample
  )
}
```


# Reovirus plots
Total counts, with and without xGen/polyA
```{r}
visListPlot(
  heart.list[c(2,4)],
  sample.titles = meta_heart$sample[c(2,4)],
  reduction = "space",
  pt.size = 0.5,
  features=c(
    "reo.log2p1",
    "nCount_xGen.kallisto.log2p1",
    "nCount_xGen.kallisto.sense.log2p1",
    "nCount_xGen.kallisto.as.log2p1"
    ),
  alt.titles = c(
    "log2(Std. Reovirus UMIs+1)",
    "log2(xGen Reovirus UMIs+1)", 
    "log2(xGen Reo-sense UMIs+1)",
    "log2(xGen Reo-antisense UMIs+1)"
    ),
  axis.title.angle.y = 0,
  combine=T,nrow = 1,
  colormap = "magma",
  colormap.direction = -1,
  colormap.same.scale = T
)&theme(
    legend.position="right"
  )&coord_fixed(
    ratio = 1.6
  )
```
Sense versus antisense for each spot
```{r}
ggplot(
  heart.merged@meta.data[sample(Cells(heart.merged)),],
  aes(
    x=nCount_xGen.kallisto.sense.log2p1,
    y=nCount_xGen.kallisto.as.log2p1,
    color=polyA
  )
)+
  geom_abline()+
  geom_point(
    alpha=0.7
  )+
  scTheme$scatter+
  scale_color_manual(
    values=mckolors$txg[c(1,4)]
  )+
  labs(
    x="Sense strand, all Reovirus segments\nlog2(UMIs+1)",
    y="Antisense strand, all Reovirus segments\nlog2(UMIs+1)"
  )
```
```{r}
ggplot(
  heart.merged@meta.data[sample(Cells(heart.merged)),],
  aes(
    x=reo.log2p1,
    y=nCount_xGen.kallisto.log2p1,
    # x=reo.counts,
    # y=nCount_xGen.kallisto,
    color=polyA
  )
)+
  geom_abline()+
  geom_point(
    alpha=0.7
  )+
  scTheme$scatter+
  scale_color_manual(
    values=mckolors$txg[c(1,4)]
  )+
  labs(
    x="Standard library prep\nlog2(Reovirus UMIs+1)",
    y="xGen Enrichment\nlog2(Reovirus UMIs+1)"
  )
```
Co-expression map of sense/antisense transcripts
```{r}
suppressMessages(
  visCoMap(
    heart.list[c(2,4)],
    sample.titles = meta_heart$sample[c(2,4)],
    reduction = "space",
    assay="xGen.kallisto",
    slot="counts",
    pt.size = 0.1,
    features=,
    axis.title.angle.y = 0,
    legend.position = "right",
    combine=T,ncol = 4,
    colormap = "magma",
    colormap.direction = -1,
    colormap.same.scale = F
  )%>%
    wrap_plots(
      guides="collect"
    )&coord_fixed(
      ratio = 1.6
    )
)

```


Each Reovirus segment, plus its antisense counterpart
```{r}
suppressMessages(
  visListPlot(
    heart.list[c(2,4)],
    sample.titles = meta_heart$sample[c(2,4)],
    reduction = "space",
    assay="xGen.kallisto",
    slot="counts",
    pt.size = 0.1,
    features=reo.genes,
    # features=Features(heart.list[[4]],assay="kallisto")%>%tail(),
    # alt.titles = c("log2(Reovirus UMIs+1)","log2(xGen Reovirus UMIs+1)"),
    axis.title.angle.y = 0,
    legend.position = "right",
    combine=T,ncol = 4,
    colormap = "magma",
    colormap.direction = -1,
    colormap.same.scale = F
  )%>%
    wrap_plots(
      guides="collect"
    )&coord_fixed(
      ratio = 1.6
    )
)
```

```{r}
suppressMessages(
  visListPlot(
    heart.list[c(2,4)],
    sample.titles = meta_heart$sample[c(2,4)],
    reduction = "space",
    assay="xGen.kallisto",
    slot="counts",
    pt.size = 0.1,
    features=reo.genes[9:10],
    # features=Features(heart.list[[4]],assay="kallisto")%>%tail(),
    # alt.titles = c("log2(Reovirus UMIs+1)","log2(xGen Reovirus UMIs+1)"),
    axis.title.angle.y = 0,
    legend.position = "right",
    combine=T,ncol = 4,
    colormap = "magma",
    colormap.direction = -1,
    colormap.same.scale = F
  )&coord_fixed(
      ratio = 1.6
    )
)
```
# Spatially-dependent gene expression patterns
## Clustering results
```{r message=FALSE, warning=FALSE}
suppressMessages(
  visListPlot(
    heart.list,
    sample.titles = stringr::str_remove_all(meta_heart$sample,pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="Heart-"),
    assay="kallisto_collapsed",
    reduction="space",
    slot = 'counts',
    pt.size=0.8,
    legend.position = "bottom",
    font.size = 8,
    axis.title.angle.y=0,
    nrow = 1,
    colormap = "magma", colormap.direction = -1,
    # combine = F,
    verbose=F,
    features = c("reo.counts.kallisto",reo.genes[1:6])
  )
)
```


# Plotting
##
```{r fig.height=4, fig.width=5, message=FALSE, warning=FALSE}
GENES <- c(
  'Rsad2',
  'Irf7',
  'Ifit3', 'Ubd'
  )

tmp.df <- data_frame(
  nCounts = heart.vis$nCount_Spatial,
  Reo_Load = GetAssayData(heart.vis,assay="xGen")%>%colSums(),
  Immune_Genes = GetAssayData(heart.vis,assay="Spatial",slot = "counts")[GENES,]%>%colSums(),
  Rsad2 = GetAssayData(heart.vis,assay="Spatial",slot = "data")["Rsad2",],
  Ifit3 = GetAssayData(heart.vis,assay="Spatial",slot = "data")["Ifit3",]
)

ggplot(
  tmp.df,
  aes(
    x=Rsad2,
    y=Ifit3,
    color=nCounts
  )
)+
  geom_point()+
  theme_minimal()+
  scale_color_viridis()

```

```{r}
tmp.feat=c(
  "mmu-let-7a-5p",
  # "mmu-let-7c-5p",
  # "mmu-let-7d-5p",
  "mmu-miR-133a-3p/133b-3p", 
  "mmu-miR-142a-3p",
  # "mmu-miR-145a-5p",
  "mmu-miR-1a-3p" 
)%>%sort()
suppressMessages(
  visListPlot(
    heart.list[c(3,4)],
    sample.titles = meta_heart$sample[c(3,4)],
    reduction = "space",
    assay="mirge3",
    slot="counts",
    pt.size = 0.1,
    features="nCount_mirge3",
    # features=Features(heart.list[[4]],assay="kallisto")%>%tail(),
    # alt.titles = c("log2(Reovirus UMIs+1)","log2(xGen Reovirus UMIs+1)"),
    axis.title.angle.y = 0,
    legend.position = "right",
    combine=T,ncol = 4,
    colormap = "plasma",
    colormap.direction = -1,
    colormap.same.scale = F
  )&coord_fixed(
      ratio = 1.6
    )
)
```


# Session Info
```{r}
sessionInfo()
```