# Comparison of standard visium to polyA visium in Reovirus-infected heart
This analysis uses a reference genome comprised of GRCm39/GENCODE_M27 and the Reovirus genome

# Session setup
## Libs, setwd
```{r message=FALSE, warning=FALSE}
library(Matrix)
library(dplyr)
library(Seurat)
library(future)

library(ggplot2)
library(patchwork)
library(pals)
library(viridis)
library(data.table)
library(shades)
source("/home/dwm269/DWM_utils/plotting_utils/scThemes.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seutils.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seuplots.R")
```

## Figure settings & colors
```{r}
# fonts, sizes, etc.
small.font = 6*2
big.font = 8*2
line.width = 0.5
pt.size=0.01
pt.stroke=0.3
label.size=2

# colors
colPals <- list()
# colPals[["sample"]] <- ggsci::pal_rickandmorty()(meta_vis$sample %>% unique() %>% length())
colPals[["celltypes"]] <- as.vector(pals::polychrome())[3:32]
```

## Plot themes 
```{r}
scTheme <- scThemes(
  small.font = small.font,
  big.font = big.font,
  line.width = line.width,
  pt.size=pt.size,
  pt.stroke=pt.stroke,
  label.size=label.size
)

print(names(scTheme))

```

```{r}
mckolors <- read.csv("/home/dwm269/DWM_utils/plotting_utils/McKolors_v1.csv") %>% 
  as.list() %>%
  lapply(
    FUN=function(X) X[X!=""]
  )
mckolors
```
# Load in data
## Read in metadata
```{r}
meta_heart <- read.csv("/workdir/dwm269/totalRNA/spTotal/resources/meta_sheet.csv")
meta_heart <- meta_heart[meta_heart$tissue=="heart",]

print(meta_heart)
```

## Read in count mats (spaceranger) and initialize Seurat objects
```{r}
heart.list <- list()
for(i in 1:nrow(meta_heart)){ 
  if(file.exists(paste0(meta_heart$data.dir.spaceranger.REO[i], '/outs/filtered_feature_bc_matrix'))){ 
    cat("Reading #",i, ": ", meta_heart$data.dir.spaceranger.REO[i], '...\n')
    heart.list[[i]] <- Seurat::Load10X_Spatial(
      data.dir = paste0(meta_heart$data.dir.spaceranger.REO[i], '/outs'),
      filter.matrix = T
    )
    
    # Strip "-1" suffix from cell barcode (colnames)
    if(stringr::str_detect(Cells(heart.list[[i]]),pattern = "-1")){
      cat("     Stripping suffix from spot barcodes...\n")
      heart.list[[i]] <- RenameCells(
        object = heart.list[[i]],
        new.names = stringr::str_remove_all(Cells(heart.list[[i]]),pattern="-1")
      )
    }
    
    cat("    Done!\n")
  }else{
    cat("Data not found for # ", i, " (", meta_heart$data.dir.spaceranger.REO[i], ")", "\n")
  }
}

```
## Add metadata 
```{r}
for(i in 1:nrow(meta_heart)){
  for(j in colnames(meta_heart)){
    heart.list[[i]]@meta.data[[j]] <- meta_heart[i,j]
  }
}
```

## Read in matrices - STARsolo 
```{r}
mat.list <- list()
for(i in 1:nrow(meta_heart)){ 
  if(file.exists(paste0(meta_heart$data.dir.STARsolo.REO[i], "/Solo.out/GeneFull/raw/"))){ 
    cat("Reading #",i, ": ", meta_heart$data.dir.STARsolo.REO[i], ' \n')
    mat.list[[i]] <- Read10X(
      data.dir = paste0(meta_heart$data.dir.STARsolo.REO[i], "/Solo.out/GeneFull/raw/")
    )
  }else{
    cat("Data not found for # ", i, " (", meta_heart$data.dir.STARsolo.REO[i], ")", "\n")
  }
}

cat(sum(unlist(lapply(mat.list, ncol))),"spots (total) loaded...\n")
```

## Add STARsolo counts as new assay
```{r}
for(i in 1:length(heart.list)){
  cat(paste0("Adding STARsolo for ", meta_heart$sample[i]),"\n")
  tmp <- CreateAssayObject(
    counts=mat.list[[i]],
    min.cells = 3
  )
  print(dim(tmp))
  
  heart.list[[i]][["STARsolo"]] <- subset(
    tmp,
    cells = Cells(heart.list[[i]])
  )
  
  heart.list[[i]]@active.assay <- "STARsolo"
}
rm(mat.list)
```

## Collapse multimappers, preprocess collapsed counts
```{r message=FALSE, warning=FALSE}
heart.list <- lapply(
  heart.list,
  FUN = function(SEU) collapseMultimappers(
    SEU,
    assay="STARsolo",
    new.assay.name = "STARsolo_collapsed",
    verbose=T
    )
) %>% lapply(
  FUN = function(SEU) seuPreProcess(
    SEU, 
    assay="STARsolo_collapsed",
    res = 0.6,
    verbose=F
  )
)
```
## Add kallisto counts
```{r message=TRUE, warning=FALSE}
require(SeuratDisk)
for(i in 1:nrow(meta_heart)){
  message(paste0("Loading & cleaning kallisto output for ",meta_heart$sample[i],"...\n"))
  
  # convert h5ad to h5seurat
  # tmp.h5 <- paste0(meta_heart$data.dir.kallisto.REO[i],"/kb_standard/counts_unfiltered/adata")
  # if(!file.exists(paste0(tmp.h5,".h5seurat"))){
  #   message(paste0("Converting h5seurat for",meta_heart$capture.area[i],"...\n"))
  #   Convert(
  #     paste0(tmp.h5,".h5ad"),
  #     dest = "h5seurat",
  #     overwrite = TRUE
  #   )
  # }
  # # Load h5seurat and subset out counts for sample
  # tmp.seu <- LoadH5Seurat(
  #   paste0(tmp.h5,".h5seurat"),
  # ) %>% subset( # select cells
  #   cells=Cells(heart.list[[i]])
  # )
  
  tmp.seu <- ReadMtx(
    mtx=paste0(meta_heart$data.dir.kallisto.REO[i],"/kb_standard/counts_unfiltered/cells_x_genes.mtx"),
    cells=paste0(meta_heart$data.dir.kallisto.REO[i],"/kb_standard/counts_unfiltered/cells_x_genes.barcodes.txt"),
    features=paste0(meta_heart$data.dir.kallisto.REO[i],"/kb_standard/counts_unfiltered/cells_x_genes.genes.txt"),
    feature.column=1,
    mtx.transpose = T
  ) %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(heart.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  # collapse multimappers
  tmp.seu <- collapseMultimappers(
    tmp.seu,
    assay="RNA",
    new.assay.name = "RNA",
    verbose=F
  )
  
  # convert ensembl IDs to MGI gene IDs
  tmp.mat <- GetAssayData(tmp.seu,assay="RNA")
  rownames(tmp.mat) <- ens2gene(
    ens=rownames(tmp.mat),
    biomart.info = mouse.info,
    ens.colname = "ensembl_gene_id",
    gene.colname = "mgi_symbol",
    force.unique = T,
    ncores=24,
    verbose=F
  )
  
  # add kallisto counts to vis_list
  heart.list[[i]][["kallisto"]] <- CreateAssayObject(
    counts = tmp.mat,
    min.cells = 1
  )
  
  rm(tmp.seu)
  rm(tmp.mat)
}
gc()
```

## Add Reovirus counts based on kallisto output
```{r}
heart.list <- lapply(
  heart.list,
  FUN = function(SEU){
    
    if(reo.genes %in% Features(SEU,assay = "kallisto_collapsed") %>% any()){
      tmp.counts = GetAssayData(
        SEU,
        assay="kallisto_collapsed",
        slot="counts"
      )[reo.genes,]
      tmp.counts = colSums(tmp.counts)
      SEU$reo.counts.kallisto <- tmp.counts
    }else{
      SEU$reo.counts.kallisto <- 0
    }
    
    return(SEU)
  }
)
```


### Collapse multimappers and preprocess collapsed counts (kallisto)
```{r warning=FALSE}
# collapse multimappers
heart.list <- lapply(
  heart.list,
  FUN = function(SEU) collapseMultimappers(
    SEU,
    assay="kallisto",
    new.assay.name = "kallisto_collapsed",
    verbose=T
    )
)

# preprocess
heart.list <- lapply(
  heart.list,
  FUN = function(SEU) seuPreProcess(
    SEU, 
    assay="kallisto_collapsed",
    verbose=F
  )
)
```

# Add spatial locations as a reduction (for easy plotting with DimPlot)
```{r warning=FALSE}

meta_heart$h_flip<- c(
  1,-1,1,-1
)
meta_heart$v_flip<- c(
  1,1,1,-1
)

#TODO- Temporarily flipped the coordinates, so that sections are in the correct orientation
heart.list <- mapply(
  FUN = function(SEU, h_flip, v_flip){
    
    #Original code (see above #TODO)
    # tmp <- as.matrix(cbind(
    #   SEU@images$slice1@coordinates$row * h_flip,
    #   SEU@images$slice1@coordinates$col * v_flip
    # ))
    
    tmp <- as.matrix(cbind(
      SEU@images$slice1@coordinates$col * h_flip,
      SEU@images$slice1@coordinates$row * v_flip
    ))
      
    colnames(tmp) <- paste0("space_", 1:2)
    rownames(tmp) <- colnames(SEU)
    
    SEU[["space"]] <- CreateDimReducObject(
      embeddings=as.matrix(tmp),
      assay="STARsolo_collapsed",
      key = "space_"
    )
    
    return(SEU)
  },
  heart.list,
  meta_heart$h_flip,
  meta_heart$v_flip
)
```

# Save list of seurat objects, and each object as an adata individually
```{r}
save(
  heart.list,
  file = "/workdir/dwm269/totalRNA/spTotal/robjs/heart_list_v2.RData"
)
cat("Done.\n\n")
```


## Save h5Seurat objects
```{r}
# require(SeuratDisk)
# for(i in 1:length(heart.list)){
#   tmp.filename = paste0("/workdir/dwm269/totalRNA/spTotal/pyobjs/",meta_heart$sample[i],"_v1",".h5Seurat")
#   cat(paste0("Saving ",tmp.filename,"...\n"))
#   SaveH5Seurat(
#     heart.list[[i]], 
#     filename = tmp.filename,
#     overwrite = T
#   )
#   cat("Done.\n\n")
#   
#   cat(paste0("Converting ",tmp.filename," to h5ad...\n"))
#   Convert(
#     tmp.filename, 
#     dest = "h5ad"
#   )
#   cat("Done.\n\n")
# }
```

# Spatially-dependent gene expression patterns
## Clustering results
```{r message=FALSE, warning=FALSE}
suppressMessages(
  visListPlot(
    heart.list,
    sample.titles = stringr::str_remove_all(meta_heart$sample,pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="Heart-"),
    assay="kallisto_collapsed",
    reduction="space",
    slot = 'counts',
    pt.size=0.8,
    legend.position = "bottom",
    font.size = 8,
    axis.title.angle.y=0,
    nrow = 1,
    colormap = "magma", colormap.direction = -1,
    # combine = F,
    verbose=F,
    features = c("reo.counts.kallisto",top.reo.genes[1:6])
  )
)
```


```{r}
lapply(
  heart.list,
  FUN=function(VIS) DimPlot(
    VIS,
    reduction = "space",
    group.by = c(
      "kallisto_collapsed_snn_res.0.8"
    )
  )
) %>% 
  lapply(
    FUN=function(VIS) VIS +
      scTheme$umap +
      theme(
        axis.title = element_blank(),
        axis.line =  element_blank(),
        plot.title=element_blank()
      )
  ) %>% wrap_plots()
```


## Differential gene expression
### Across clusters, within each sample
```{r}
dge.list <- lapply(
  heart.list,
  FUN = function(VIS){ 
    Idents(VIS) <- "STARsolo_collapsed_snn_res.0.6"
    
    FindAllMarkers(
      VIS,
      assay = "STARsolo_collapsed"
    ) %>% 
      return()
  }
)

head(dge.list[[4]])
```
### yPAP vs. Ctrl (infected)
```{r}
t1l.vis <- merge(
  heart.list[[2]],
  heart.list[[4]]
)

Idents(t1l.vis) <- "polyA"

pa.t1l.dge <- FindAllMarkers(
  t1l.vis
)

head(pa.t1l.dge,20)
```
### Plot interesting yPAP-specific genes
```{r}
tmp.features <- c(
  # "Gm11335",
  "5-8S-rRNA", #intronic feature
  "Vault",
  "Rny1",
  "7SK",
  "Mettl26", # much higher 5' coverage - isoform?
  "U2.1",
  "Snord118"
)

# Top miRs
tmp.features = grepGenes(
  heart.list[[4]],
  pattern = "Mir",
  filter.pattern = "hg",
  sort.by = "expression",
  assay = "STARsolo_collapsed"
  )[1:10]

tmp.features = c(
  "Meg3",
  ""
)

# Top Mito tRNAs
tmp.features = grepGenes(heart.list[[4]],"mt-T")[1:10]

# Top scaRNAs
tmp.features = grepGenes(heart.list[[4]],"Scarna")#[1:10]

# Top histone RNAs
## http://www.informatics.jax.org/mgihome/nomen/gene_name_initiative.shtml
tmp.features = grepGenes(heart.list[[4]],pattern=c("H1f","Macroh","H2ab","H2ac","H2ax","H2b","H3c","H4c"))[1:10]

visListPlot(
  heart.list,
  sample.titles = stringr::str_remove_all(meta_heart$sample,pattern = "Vis_")%>%
    stringr::str_remove_all(pattern="_Heart"),
  assay="STARsolo_collapsed",
  reduction="space",
  # slot = 'counts',
  pt.size=0.6,
  legend.position = "bottom",
  font.size = 12,
  axis.title.angle.y=0,
  nrow = 1,
  # combine = F,
  verbose=F,
  features = tmp.features
) 
```

## Genes w/ intronic features
```{r}
c(
  "Zfp791",
  "Cdk8",
  "Lrrc15", 
  "Zc3h7a",
  "Gphn",
  "Hexb"
)

```


# Add Reovirus counts
```{r}
reo.genes <- lapply(
  heart.list, 
  FUN=function(VIS) grepGenes(
    SEU=VIS,
    pattern = "T1L",
    assay="kallisto",
    sort.by = T,
    verbose=F
    )
  ) %>% 
  unlist() %>%
  unique() %>%
  sort()
print(reo.genes)
```

## Count total number of Reovirus transcripts in each spot
```{r}
heart.list <- lapply(
  heart.list,
  FUN = function(SEU){
    if(reo.genes %in% Features(SEU,assay = "kallisto") %>% any()){
      tmp.counts = GetAssayData(
        SEU,
        assay="kallisto",
        slot="counts"
      )[reo.genes,]
      tmp.counts = colSums(tmp.counts)
      SEU$reo.counts <- tmp.counts
      SEU$reo.log2p1 <- log2(tmp.counts+1)
    }else{
      SEU$reo.counts <- 0
      SEU$reo.log2p1 <- 0
    }
    
    return(SEU)
  }
)
```

# Load in xGen numbers
```{r}
library(SeuratDisk)
xGen.list <- list()
xGen.samples <- paste0(
  "/workdir/dwm269/totalRNA/data/kallisto/GRCm39_REOT1L_Visium/",
  c(
    "T1L_D7PI_xGen",
    "Vis_yPAP_2C_xGen",
    "Vis_yPAP_3B_xGen"
  ),
  "/kb_standard/counts_unfiltered/adata"
)

# convert h5ad to h5seurat
for(i in 1:length(xGen.samples)){
  tmp.h5 <- xGen.samples[i]
  print(tmp.h5)
  if(!file.exists(paste0(tmp.h5,".h5seurat"))){
    
    Convert(
      paste0(tmp.h5,".h5ad"),
      dest = "h5seurat",
      overwrite = TRUE
    )
  }

  # Load h5seurat and subset out counts for sample
  tmp.seu <- LoadH5Seurat(
    paste0(tmp.h5,".h5seurat"),
  ) 
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  # convert ensembl IDs to MGI gene IDs
  # tmp.mat <- GetAssayData(tmp.seu,assay="RNA")
  # rownames(tmp.mat) <- ens2gene(
  #   ens=rownames(tmp.mat),
  #   biomart.info = mouse.info,
  #   ens.colname = "ensembl_gene_id",
  #   gene.colname = "mgi_symbol",
  #   force.unique = T,
  #   ncores=24,
  #   verbose=F
  # )
  
  # add kallisto counts to vis_list
  # xGen.list[[i]][["kallisto"]] <- CreateAssayObject(
  #   counts = tmp.mat,
  #   min.cells = 1
  # )
  xGen.list[[i]] <- tmp.seu
  rm(tmp.seu)
  # rm(tmp.mat)
}
gc()
```

```{r}
reo.genes = grepGenes(
  xGen.list[[2]],
  pattern = "Reo",
  sort.by = "ABC"
)

# Add xGen numbers 
## Control Visium
tmp.mat <- GetAssayData(xGen.list[[1]])
tmp.mat <- tmp.mat[reo.genes, colnames(tmp.mat) %in% Cells(heart.list[[2]])] #only keep Reo genes and spots that pass QC

zero.spots = Cells(heart.list[[2]])[!Cells(heart.list[[2]])%in%colnames(tmp.mat)] # barcodes of *spots not detected* in xGen assay
tmp.mat <- cbind( # add zeroes for spots not detected by xGen 
  tmp.mat, 
  matrix(
    data = 0,
    nrow=nrow(tmp.mat),
    ncol=length(zero.spots),
    dimnames=list(
      rownames(tmp.mat),
      zero.spots
    )
  )%>% as.sparse()
)
tmp.mat <- tmp.mat[reo.genes, Cells(heart.list[[2]])] #make sure everything is in the right order

heart.list[[2]][["xGen.kallisto"]] <- CreateAssayObject(counts = tmp.mat)
#meta_heart$sample=="yPAP-Pro_Heart-D7T1L"
heart.list[[2]]$nCount_xGen.kallisto.log2p1 <- log2(heart.list[[2]]$nCount_xGen.kallisto+1)


## SUPERase + spTotal
#TODO


## Protector + spTotal
tmp.mat <- GetAssayData(xGen.list[[3]])
tmp.mat <- tmp.mat[reo.genes, colnames(tmp.mat) %in% Cells(heart.list[[4]])] #only keep Reo genes and spots that pass QC

heart.list[[4]][["xGen.kallisto"]] <- CreateAssayObject(counts = tmp.mat)
#meta_heart$sample=="yPAP-Pro_Heart-D7T1L"
heart.list[[4]]$nCount_xGen.kallisto.log2p1 <- log2(heart.list[[4]]$nCount_xGen.kallisto+1)

# Plot total counts
visListPlot(
  heart.list[c(2,4)],
  sample.titles = meta_heart$sample[c(2,4)],
  reduction = "space",
  pt.size = 2,
  features=c("reo.log2p1","nCount_xGen.kallisto.log2p1"),
  alt.titles = c("log2(Reovirus UMIs+1)","log2(xGen Reovirus UMIs+1)"),
  colormap = "magma",
  colormap.direction = -1,
  colormap.same.scale = T
)&theme(
  legend.position="right"
)&
  coord_fixed(
    ratio = 1.6
  )
```



# Plotting
## Feature plots

```{r}

visListPlot(
  heart.list,#[1:2],
  sample.titles = stringr::str_remove_all(meta_heart$sample,pattern = "Vis_") %>%
    # stringr::str_remove_all(pattern ="yPAP-")%>%
    # stringr::str_remove_all(pattern ="ctrl-")%>%
    stringr::str_remove_all(pattern ="Heart-"),
  assay="STARsolo_collapsed",
  reduction="space",
  # slot = 'counts',
  pt.size=0.8,
  legend.position = "bottom",
  font.size = 12,
  axis.title.angle.y=0,
  nrow = 1,
  # combine = F,
  verbose=F,
  features = c(
    "reo.counts"
    # "Pecam1", "Ctss", "Gzma", "Ankrd1", "Timp1"
  )
    # reo.genes

) 
```
# Heart genes
```{r fig.height=8, fig.width=8, message=FALSE, warning=FALSE}
visListPlot(
  list(heart.vis),
  sample.titles = meta_vis$sample,
  slot = 'counts',
  pt.size=1.7,
  legend.position = "right",
  font.size = 12,
  # features = ribo.genes[ribo.genes %in% rownames(vis.list[[3]])]
  features= c('Rsad2', 'Irf7','Ifit3', 'Ubd'),
  combine=F
) %>% wrap_plots()&theme(plot.margin = unit(rep(0,4),"in"))

```

##
```{r fig.height=4, fig.width=8, message=FALSE, warning=FALSE}
heart.vis@active.assay <- "xGen"

visListPlot(
  list(heart.vis),
  sample.titles = "Heart_T1L_D7PI",
  slot = 'counts',
  pt.size=1.7,
  legend.position = "right",
  font.size = 12,
  features=c(
    # "ReoT1L-T1LReoS1", "ReoT1L-T1LReoS2", "ReoT1L-T1LReoS3", 
    "ReoT1L-T1LReoS4",
    # "ReoT1L-T1LReoM1", "ReoT1L-T1LReoM2", 
    "ReoT1L-T1LReoM3"
    # "ReoT1L-T1LReoL1","ReoT1L-T1LReoL2", "ReoT1L-T1LReoL3"
    ),
  combine=F
) %>% wrap_plots()&scale_color_viridis_c(
  na.value = gray(0.69),
  values = c(0.001,1),
  option="inferno",
  direction=-1
  )

```
##
```{r fig.height=4, fig.width=5, message=FALSE, warning=FALSE}
GENES <- c(
  'Rsad2',
  'Irf7',
  'Ifit3', 'Ubd'
  )

tmp.df <- data_frame(
  nCounts = heart.vis$nCount_Spatial,
  Reo_Load = GetAssayData(heart.vis,assay="xGen")%>%colSums(),
  Immune_Genes = GetAssayData(heart.vis,assay="Spatial",slot = "counts")[GENES,]%>%colSums(),
  Rsad2 = GetAssayData(heart.vis,assay="Spatial",slot = "data")["Rsad2",],
  Ifit3 = GetAssayData(heart.vis,assay="Spatial",slot = "data")["Ifit3",]
)

ggplot(
  tmp.df,
  aes(
    x=Rsad2,
    y=Ifit3,
    color=nCounts
  )
)+
  geom_point()+
  theme_minimal()+
  scale_color_viridis()

```


# Session Info
```{r}
sessionInfo()
```