#  Small RNA pseudobulk analysis of spTotal datasets
## David W. McKellar

# Session setup
## Libs, setwd
```{r message=FALSE, warning=FALSE}
# library(Matrix)
library(dplyr)

library(ggplot2)
library(patchwork)
library(viridis)
library(data.table)
library(shades)

source("/home/dwm269/DWM_utils/plotting_utils/scThemes.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seutils.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seuplots.R")
```
## Figure settings
```{r}
small.font = 6*2
big.font = 8*2
line.width = 0.5
pt.size=0.01
pt.stroke=0.3
label.size=2
```
## Plot themes 
```{r}
scTheme <- scThemes(
  small.font = small.font,
  big.font = big.font,
  line.width = line.width,
  pt.size=pt.size,
  pt.stroke=pt.stroke,
  label.size=label.size
)

print(names(scTheme))
```
## Personal color palettes
```{r}
mckolors <- read.csv("/home/dwm269/DWM_utils/plotting_utils/McKolors_v1.csv") %>% 
  as.list() %>%
  lapply(
    FUN=function(X) X[X!=""]
  )
names(mckolors)
```

## Read in metadata, `meta_smRNA`
```{r}
meta_smRNA <- read.csv(
  file="/workdir/dwm269/totalRNA/spTotal/resources/metadata_sheets/meta_sheet_smRNA.csv"
)
meta_smRNA <- meta_smRNA[with(meta_smRNA, order(meta_smRNA$tissue, meta_smRNA$chemistry)),] # sort samples by tissue & chemistry
# meta_smRNA <- meta_smRNA[meta_smRNA$rnase_inhib!="Protector",]
meta_smRNA <- meta_smRNA[!meta_smRNA$tissue%in% c("cells","nuclei"),] # filter out cell line samples
meta_smRNA$sample <- factor(meta_smRNA$sample,levels=meta_smRNA$sample)
rownames(meta_smRNA) <- meta_smRNA$sample
print(meta_smRNA)
```

# Alignment annotations
```{r}
df.anno <- lapply(
  meta_smRNA$data.dir.smRNA,
  FUN=function(datadir) read.csv(paste0(datadir,"/miRge/annotation.report.csv"))
  ) %>%
  do.call(what = rbind)
df.anno <- cbind(meta_smRNA, df.anno)
df.anno
```

```{r}
ggplot(
  df.anno,
  aes(
    x=All.miRNA.Reads/Total.Input.Reads,
    y=Remaining.Reads/Total.Input.Reads,
    size=Total.Input.Reads,
    color=rnase_inhib,
    shape=tissue
  )
)+
  geom_point(
    alpha=0.8
  )+
  scTheme$fig1bcd+
  theme(
    legend.position="right"
  )+
  scale_x_log10()+
  scale_color_manual(
    values=mckolors$polychrome[6:32]
  )

```
## Biotype distribution
```{r fig.height=6, fig.width=10, warning=FALSE}
tmp.vars = c(
  "Filtered.miRNA.Reads", "Hairpin.miRNAs",
  "mature.tRNA.Reads","primary.tRNA.Reads",
  "snoRNA.Reads"
  # "rRNA.Reads",
  # "ncRNA.others"
  # "mRNA.Reads",
  # "Remaining.Reads"     
)

tmp <- reshape2::melt(
  df.anno,
  id.vars=c("sample","chemistry"), 
  measure.vars=tmp.vars,
  variable.name="biotype"
  )
tmp$sample <- factor(
  tmp$sample,
  levels=levels(meta_smRNA$sample)
)

ggplot(
  tmp,
  aes(
    x=sample,
    y=value,
    fill=biotype
  )
)+
  geom_col(
    position="fill"
  )+
  scale_fill_manual(values=mckolors$polychrome[c(6:12,14:32)])+
  scale_y_continuous(labels = scales::percent)+
  labs(
    color="non-mRNA biotypes"
  )+
  scTheme$bar+
  theme(
    axis.text.x=element_text(angle=45, hjust=1, vjust=1),
    axis.text.y = element_text(angle=0),
    legend.position="right",
    legend.title = element_text(face="bold"),
    axis.title = element_blank()
  )

```

# Pseudobulk expression comparison
```{r}
mir.rpm <- lapply(
  meta_smRNA$data.dir.smRNA,
  FUN=function(datadir){
    X = read.csv(paste0(datadir,"/miRge/miR.RPM.csv"),row.names = "miRNA") %>% 
      as.matrix()  
    return(
      log2(X+1)
    )
  }
) %>%
  do.call(what = cbind)
colnames(mir.rpm) <- meta_smRNA$sample

head(mir.rpm)
```
# Heart samples
```{r}
tmp.feat = c(
  "mmu-let-7a-5p",
  "mmu-let-7b-5p",
  "mmu-let-7c-5p",
  "mmu-let-7d-5p",
  "mmu-let-7f-5p",
  "mmu-let-7g-5p",
  "mmu-let-7i-5p",
  "mmu-miR-1a-3p",
  "mmu-miR-21a-5p",
  "mmu-miR-22-3p",
  "mmu-miR-26a-5p",
  "mmu-miR-30a-5p",
  "mmu-miR-30e-5p",
  "mmu-miR-133a-3p/133b-3p",
  "mmu-miR-142a-3p" ,
  "mmu-miR-145a-5p",
  "mmu-miR-322-5p",
  # "mmu-miR-706",
  "mmu-miR-3473b/3473e"
)
i=1:21
rownames(mir.rpm)[grep(rownames(mir.rpm), pattern="3473")]

tmp.df = data.frame(meta_smRNA[i,c("chemistry","polyA")])
rownames(tmp.df)<-meta_smRNA$sample[i]

pheatmap::pheatmap(
  mir.rpm[tmp.feat, i],
  annotation_col = tmp.df,
  border_color = "black",
  cluster_rows = F,
  cluster_cols = F
)
```
# Muscle samples
```{r fig.height=8, fig.width=8}
tmp.feat = c(
  "mmu-let-7a-5p",
  "mmu-let-7b-5p",
  "mmu-let-7c-5p",
  "mmu-let-7d-5p",
  "mmu-let-7f-5p",
  "mmu-let-7g-5p",
  "mmu-let-7i-5p",
  "mmu-miR-1a-3p",
  "mmu-miR-1a-1-5p",
  "mmu-miR-1a-2-5p",
  "mmu-miR-10a-5p",
  "mmu-miR-10b-5p",
  "mmu-miR-16-5p",
  "mmu-miR-21a-5p",
  "mmu-miR-22-3p",
  "mmu-miR-26a-5p",
  "mmu-miR-30a-5p",
  "mmu-miR-30e-5p",
  "mmu-miR-125b-1-3p", "mmu-miR-125b-2-3p", "mmu-miR-125b-5p",
  "mmu-miR-126a-5p",
  "mmu-miR-133a-3p/133b-3p",
  "mmu-miR-142a-3p" ,
  "mmu-miR-143-3p",
  "mmu-miR-190a-3p",  "mmu-miR-190a-5p", 
  "mmu-miR-206-3p",
  "mmu-miR-451a",
  "mmu-miR-486a-5p/486b-5p"
  # "mmu-miR-706",
)

tmp.feat.b = c(
  "mmu-miR-7064-5p",
  "mmu-miR-378a-5p",
  "mmu-miR-574-5p",
  "mmu-miR-216c-5p",
  "mmu-miR-378a-5p",
  "mmu-miR-344h-3p",
  "mmu-miR-196b-5p",
  "mmu-miR-34c-5p"
)
rownames(mir.rpm)[grep(rownames(mir.rpm), pattern="1a")]

tmp.col.nums <- c(22:56)
# tmp.col.nums <- c(
#   18:31,
#   42:44,
#   # 33,35,36,34,37,
#   38:41
# )

tmp.df = data.frame(meta_smRNA[tmp.col.nums,c("chemistry","polyA")])
rownames(tmp.df)<-meta_smRNA$sample[tmp.col.nums]

pheatmap::pheatmap(
  mir.rpm[tmp.feat, tmp.col.nums],
  annotation_col = tmp.df,
  border_color = "black",
  cluster_rows = F,
  cluster_cols = F
)
```


# Comparison between spTotal & smRNAseq
```{r}
tmp <- reshape2::melt(
  mir.rpm,
  id.vars=colnames(tmp)
)
colnames(tmp) <- c("gene","sample","expression")
tmp$chemistry <- meta_smRNA[tmp$sample,"chemistry"]
tmp$polyA <- meta_smRNA[tmp$sample,"polyA"]
tmp$tissue <- meta_smRNA[tmp$sample,"tissue"]

tmp <- reshape2::dcast(
  tmp[tmp$polyA != "ctrl",],
  gene+tissue ~ chemistry,
  value.var = "expression",
  fun.aggregate = mean,fill = 0
)
tmp %>% tail()

ggplot(
  tmp,
  aes(
    x=smRNAseq,
    y=spTotal,
    color=tissue
  )
)+
  geom_point(
    alpha=0.6
  )+
  geom_smooth(
    method="lm",
    color="black"
  )+
  # ggrepel::geom_text_repel(
  #   data=tmp[tmp$Visium/tmp$smRNAseq>0,],
  #   aes(label=gene)
  # )+
  scale_color_manual(
    values=mckolors$cubs
  )+
  scTheme$scatter+
  facet_wrap(facets = "tissue")+
  ggpubr::stat_regline_equation(
    label.x=0, 
    label.y=18,
    size=small.font/2,
    ) +
  ggpubr::stat_cor(
    aes(label=..rr.label..), 
    label.x=0, 
    label.y=17,
    size=small.font/2
    )+
  theme(
    strip.text = element_text(
      size=big.font,
      face="bold"
    ),
    legend.position="none"
  )+
  labs(
    x="smRNAseq - Mean log2(RPM+1)",
    y="spTotal - Mean log2(RPM+1)"
  )

```

# Comparison to kallisto+miRBase
```{r}
mir.kallisto <- lapply(
  skm.list,
  FUN=function(SEU){
    counts = rowSums(GetAssayData(SEU,assay="miRNA",slot="counts"))
    cpm = counts*10^6/sum(counts)
    
    return(
      log2(cpm+1)
      )
  }
)

mir.kallisto <- do.call(cbind,mir.kallisto)

colnames(mir.kallisto) <- paste0(meta_skm$sample,"_kal")
head(mir.kallisto)
```
```{r}
tmp.feat = c(
  "mmu-let-7a-5p",
  "mmu-let-7b-5p",
  "mmu-let-7c-5p",
  "mmu-let-7d-5p",
  "mmu-let-7f-5p",
  "mmu-let-7g-5p",
  "mmu-let-7i-5p",
  "mmu-miR-1a-3p",
  "mmu-miR-1a-1-5p",
  "mmu-miR-1a-2-5p",
  "mmu-miR-10a-5p",
  "mmu-miR-10b-5p",
  "mmu-miR-16-5p",
  "mmu-miR-21a-5p",
  "mmu-miR-22-3p",
  "mmu-miR-26a-5p",
  "mmu-miR-30a-5p",
  "mmu-miR-30e-5p",
  "mmu-miR-125b-1-3p", "mmu-miR-125b-2-3p", "mmu-miR-125b-5p",
  "mmu-miR-126a-5p",
  "mmu-miR-133a-3p/133b-3p",
  "mmu-miR-142a-3p" ,
  "mmu-miR-143-3p",
  "mmu-miR-190a-3p",  "mmu-miR-190a-5p", 
  "mmu-miR-206-3p",
  "mmu-miR-451a",
  "mmu-miR-486a-5p/486b-5p"
  # "mmu-miR-706",
)
tmp.feat = tmp.feat[tmp.feat%in%rownames(mir.rpm) & tmp.feat%in%rownames(mir.kallisto)]


pheatmap::pheatmap(
  cbind(
    mir.rpm[tmp.feat, tmp.col.nums],
    mir.kallisto[tmp.feat,]
  ),
  # annotation_col = tmp.df,
  border_color = "black",
  cluster_rows = F,
  cluster_cols = F
)
```

