#  Comparison of standard visium to polyA visium in notexin-injured tibialis anterior muscles
## David W. McKellar

# Session setup
## Libs, setwd
```{r message=FALSE, warning=FALSE}
library(Matrix)
library(dplyr)
library(Seurat)
library(future)

library(ggplot2)
library(patchwork)
library(viridis)
library(data.table)
library(shades)

source("/home/dwm269/DWM_utils/plotting_utils/scThemes.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seutils.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seuplots.R")

setwd("/local/workdir/dwm269/totalRNA/spTotal/")
```

## Figure settings & colors
```{r}
# fonts, sizes, etc.
small.font = 6*2
big.font = 8*2
line.width = 0.5
pt.size=0.01
pt.stroke=0.3
label.size=2
```

## Plot themes 
```{r}
scTheme <- scThemes(
  small.font = small.font,
  big.font = big.font,
  line.width = line.width,
  pt.size=pt.size,
  pt.stroke=pt.stroke,
  label.size=label.size
)

print(names(scTheme))
```
```{r}
mckolors <- read.csv("/home/dwm269/DWM_utils/plotting_utils/McKolors_v1.csv") %>% 
  as.list() %>%
  lapply(
    FUN=function(X) X[X!=""]
  )
names(mckolors)
```

# Load data & metadata
```{r}
meta_vis <- read.csv("/workdir/dwm269/totalRNA/spTotal/resources/metadata_sheets/meta_sheet_visium.csv")
meta_skm <- meta_vis[meta_vis$tissue=="muscle"&meta_vis$rnase_inhib%in%c("ctrl", "Protector"),] #subset metadata

# load("/workdir/dwm269/totalRNA/spTotal/robjs/skm_list_v2.RData")
meta_skm
```

# STARsolo vs. kallisto/BUStools
```{r warning=FALSE}
wrap_plots(
  visListPlot(
    skm.list,
    sample.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="_SkM"),
    reduction="space",
    pt.size=0.4,
    legend.position = "bottom",
    font.size = 6,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    colormap = mckolors$Spectral %>% rev(),
    colormap.same.scale = T,
    features = c(
      "nFeature_STARsolo_collapsed","nFeature_kallisto_collapsed"
    )
  )&theme(
    legend.text = element_text(size=6)
  )&coord_fixed(ratio=1/1.6),
  
  visListPlot(
    skm.list,
    sample.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="_SkM"),
    reduction="space",
    pt.size=0.4,
    legend.position = "bottom",
    font.size = 6,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    colormap = mckolors$Spectral %>% rev(),
    colormap.same.scale = T,
    features = c(
      "nCount_STARsolo_collapsed","nCount_kallisto_collapsed"
    )
  )&theme(
    legend.text = element_text(size=6)
  )&coord_fixed(ratio=1/1.6),
  nrow=1
)
```

Scatter plot for each spot, colored by preproceessing, to look at # Features and # UMIs detected by STARsolo vs. kallisto
```{r fig.height=6, fig.width=12}
if(!exists("skm.merged")){
  skm.merged <- merge(
    skm.list[[1]],
    skm.list[2:length(skm.list)],
    add.cell.ids = meta_skm$sample
  )
}
wrap_plots(
  ggplot(
    skm.merged@meta.data[sample(rownames(skm.merged@meta.data)),],
    aes(
      x=nFeature_STARsolo,
      y=nFeature_kallisto,
      color=polyA
    )
  )+
    geom_point(alpha=0.5)+
    geom_abline()+
    scTheme$fig1bcd+
    scale_color_manual(values=mckolors$txg)+
    guides(color=guide_legend(override.aes = list(size=4,alpha=1,fill=NA))),
  
  ggplot(
    skm.merged@meta.data[sample(rownames(skm.merged@meta.data)),],
    aes(
      x=nCount_STARsolo,
      y=nCount_kallisto,
      color=polyA
    )
  )+
    geom_point(alpha=0.5)+
    geom_abline()+
    scTheme$fig1bcd+
    scale_color_manual(values=mckolors$txg)+
    guides(color=guide_legend(override.aes = list(size=4,alpha=1,fill=NA))),
  guides = "collect"
)&scTheme$scatter&theme(
  legend.position="right"
)
```
# Knee plots by assay
```{r}
lapply(
  skm.list,
  FUN=function(VIS){
    qplot(
      1:length(VIS$nCount_kallisto),
      sort(VIS$nCount_kallisto, decreasing=T)
    )+
      scTheme$scatter+
      theme(axis.title.x = element_blank())+
      labs(
        y="nCount_kallisto",
        title=VIS$sample[1]
      )+
      scale_y_continuous(limits = c(0,75000))
  }
)%>%
  wrap_plots()&scTheme$scatter

```

# Get gene biotype info from the GENCODE annotations
`gtf.info` is just a reformatted .tsv of the GENCODE M28 annotations .gtf
```{r}
gtf.info <- read.csv(
  file="/workdir/dwm269/totalRNA/spTotal/resources/gene_lists/GRCm39_GENCODEm28_gene_info_gtf.tsv",
  sep = "\t"
)

# Remove period/decimal suffixes from ensembl IDs
gtf.info$GeneID <- strsplit(
  x=gtf.info$GENEID,
  split = "\\."
)%>%
  lapply(
    FUN=function(X) X[[1]]
  )%>%
  unlist()

cat(
  paste("Using", length(unique(gtf.info$GeneSymbol)), "genes, and ",length(unique(gtf.info$Biotype)),"biotypes...\n")
)

head(gtf.info)
```

## gene lists by biotype
```{r}
genes = list()
for(biotype in unique(gtf.info$Biotype)){
  genes[[biotype]] <- gtf.info$GENEID[gtf.info$Biotype==biotype] %>%
    ens2gene(
      biomart.info = gtf.info,
      ens.colname = "GENEID",
      gene.colname = "GeneSymbol"
      ) %>%
    unique()
}

names(genes)
```

## Gene biotype spatial plots
```{r warning=FALSE}
visListPlot(
  skm.list,
  sample.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
        stringr::str_remove_all(pattern ="yPAP_")%>%
        stringr::str_remove_all(pattern ="ctrl_")%>%
        stringr::str_remove_all(pattern ="_SkM"),
  # assay="STARsolo_collapsed",
  reduction="space",
  # slot = 'counts',
  pt.size=0.6,
  legend.position = "bottom",
  font.size = 12,
  axis.title.angle.y=0,
  nrow = 1,
  combine = T,
  verbose=F,
  # colormap="inferno",colormap.direction = -1,
  colormap = mckolors$RdYlBu%>%rev(),
  features = c(
    # "nFeature_STARsolo","nCount_STARsolo","percent.mt",
    "pct.protein_coding",
    # "pct.lncRNA",
    "pct.rRNA",
    # "pct.Mt_rRNA", 
    "pct.Mt_tRNA",
    # "pct.ribo.pp",
    "pct.snoRNA",
    "pct.miRNA",
    "pct.snRNA"
  ) 
)%>% suppressMessages()
```


# Gene spatial feature plots
## Canonical myogenesis markers
```{r fig.height=8, fig.width=8}
suppressMessages(
  visListPlot(
    skm.list,
    sample.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="_SkM"),
    reduction="space",
    assay="kallisto_collapsed", slot = 'data',
    pt.size=0.4,
    legend.position = "bottom",
    font.size = 6,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    colormap = "plasma",
    # colormap = mckolors$Spectral%>%rev(),
    features = c(
      # "Dicer1"
      "Myf5","Myod1","Myog","Myh1","Myh4"
    )
  )&theme(
    legend.text = element_text(size=6)
  )&coord_fixed(
    ratio=1/1.6
  )
)
```

## MyomiRs
### Spatial maps of canonical myomiRs
```{r fig.height=8, fig.width=8, message=FALSE, warning=FALSE}
tmp.feat = c(
  "mmu-miR-1a-3p",
  "mmu-miR-206-3p",
  "mmu-miR-133a-3p/133b-3p"
)

suppressMessages(
  visListPlot(
    skm.list,
    sample.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="_SkM"),
    reduction="space",
    assay="mirge3",
    slot = 'data',
    pt.size=0.4,
    legend.position = "bottom",
    font.size = 8,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    colormap = "plasma",
    # colormap = mckolors$Spectral%>%rev(),
    features=tmp.feat
    # features = c("nFeature_mirge3", "nCount_mirge3")
  )&theme(
    legend.text = element_text(size=6)
  )&coord_fixed(
    ratio=1/1.6
  )
)
```

### Co-expression map of miRs and their targets, through injury response
Read in the lists of miR targets, downloaded from [TargetScan8](http://www.targetscan.org/mmu_80/)
    [Info on TargetScan metrics](http://www.targetscan.org/faqs.Release_7.html)
```{r}
ts8.path <- "/workdir/dwm269/totalRNA/spTotal/resources/TargetScan8/"
ts8_targets <- list(
  # let7_5p = read.csv(paste0(ts8.path,"TargetScan8.0__let-7-5p_miR-98-5p.predicted_targets.txt"),sep = "\t",header = T),
  mir1_3p = read.csv(paste0(ts8.path,"TargetScan8.0__miR-1-3p_206-3p.predicted_targets.txt"),sep = "\t",header = T)
  # mir133_3p = read.csv(paste0(ts8.path,"TargetScan8.0__miR-133-3p.predicted_targets.txt"),sep = "\t",header = T),
  # mir142b = read.csv(paste0(ts8.path,"TargetScan8.0__miR-142b.predicted_targets.txt"),sep = "\t",header = T),
  # mir1a_5p = read.csv(paste0(ts8.path,"TargetScan8.0__miR-1a-5p.predicted_targets.txt"),sep = "\t",header = T),
  # mir21_5p = read.csv(paste0(ts8.path,"TargetScan8.0__miR-21-5p.predicted_targets.txt"),sep = "\t",header = T)
  # mir196_5p = read.csv(paste0(ts8.path,"TargetScan8.0__miR-196-5p.predicted_targets.txt"),sep = "\t",header = T),
  # mir378_5p = read.csv(paste0(ts8.path,"TargetScan8.0__miR-378a-5p.predicted_targets.txt"),sep = "\t",header = T),
  # mir574_5p = read.csv(paste0(ts8.path,"TargetScan8.0__miR-574-5p.predicted_targets.txt"),sep = "\t",header = T)
  # mir486 = read.csv(paste0(ts8.path,"TargetScan8.0__miR-326-5p_486-3p_6954-5p.predicted_targets.txt"),sep = "\t",header = T)
) %>%
  lapply( # Filter target lists based on number of binding sites
    FUN=function(X)
      X[X[["Aggregate.PCT"]]>0.75 & X[["Total.context...score"]] < -0.5,]
      # X[X[["Total.context...score"]] < -0.5,]
      # X[order(X[["Total.context...score"]])[1:100],] # top 100 genes
  )

# Separate mir1 and mir206 targets
ts8_targets$mir206_3p <- ts8_targets$mir1_3p[ts8_targets$mir1_3p$Representative.miRNA=="mmu-miR-206-3p",]
ts8_targets$mir1_3p <- ts8_targets$mir1_3p[ts8_targets$mir1_3p$Representative.miRNA=="mmu-miR-1a-3p",]

print(names(ts8_targets))

qplot(
  as.numeric(ts8_targets$mir1_3p$Total.context...score), 
  as.numeric(ts8_targets$mir1_3p$Aggregate.PCT)
  )+
  scTheme$scatter+
  labs(x="context++ Score",y="Prob. of conserved targeting (P_ct)")
```

Manually annotate lists of miRs (miRbase_IDs -> GENCODE)
```{r}
#TODO: load in miRbase data (somehow?) to translate gene names
ts8_miRs <- list(
  let7_5p = c(
    "Mirlet7a-1","Mirlet7a-2","Mirlet7b","Mirlet7c-1","Mirlet7c-2","Mirlet7d","Mirlet7e",
    "Mirlet7f-1","Mirlet7f-2","mmu-let-7f-1", "Mirlet7g","Mirlet7i","Mir98" 
 ),
  mir1_3p = c(
    "Mir1a-1", "Mir1a-2","Mir206"
  ),
  mir133_3p = c(
    "Mir133a-1", "Mir133a-2", "Mir133b", "Mir133c" 
  ),
 mir142b = c(
   	"Mir142b"
 ),
 mir1a_5p = c(
   	"Mir1a-1"
 ),
 mir21_5p = c(
   	"Mir21a"
 )
 # mir486 = c(
 #   	mmu-miR-326-5p    mmu-miR-486a-3p    mmu-miR-486b-3p    mmu-miR-6954-5p 
 # )
)
print(names(ts8_miRs))
```

Compute module scores for each list of miRs 
```{r}
for(miR in names(ts8_miRs)){
  cat(paste0("Scoring expression of miRs in the '", miR, "' set...\n"))
  
  skm.list <- lapply(
    skm.list,
    FUN=function(SEU){ 
      print(sum(ts8_miRs[[miR]] %in% Features(SEU, assay="kallisto_collapsed")))
      if(any(ts8_miRs[[miR]] %in% Features(SEU, assay="kallisto_collapsed"))){
        tmp.feats = list(ts8_miRs[[miR]])
        SEU <- AddModuleScore(
          SEU,
          features = tmp.feats,
          assay = "kallisto_collapsed",
          name = paste0(miR,"_miRs")
        )
      }
      return(SEU)
    }
  )
  rm(tmp.feats)
}
```

Compute module scores for each list of target genes
```{r}
for(miR in names(ts8_targets)){
  cat(paste0("Scoring expression of targets in the '", miR, "' set...\n"))
  
  # tmp.assay="spliced"
  # tmp.assay="unspliced"
  tmp.assay="kallisto_collapsed"
  
  skm.list <- lapply(
    skm.list,
    FUN=function(SEU){ 
      cat(sum(ts8_targets[[miR]]$Target.gene %in% Features(SEU, assay=tmp.assay)))
      cat(" ")
      if(sum(ts8_targets[[miR]]$Target.gene %in% Features(SEU, assay=tmp.assay))>1){
        
        tmp.feats = list(
          ts8_targets[[miR]]$Target.gene[ts8_targets[[miR]]$Target.gene %in% Features(SEU, assay=tmp.assay)]
        )
        
        SEU <- AddModuleScore(
          SEU,
          features = tmp.feats,
          assay = tmp.assay,
          name = paste0(miR,"_targets")
          # name = paste0(miR,"_targets_spliced")
          # name = paste0(miR,"_targets_unspliced")
        )
      }
      return(SEU)
    }
  )
  cat("\n")
}
```

Plot target scores
```{r fig.height=10, fig.width=8}
suppressMessages(
  visListPlot(
    skm.list,
    sample.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="_SkM"),
    reduction="space",
    assay="kallisto_collapsed", 
    slot = 'data',
    pt.size=0.4,
    legend.position = "bottom",
    font.size = 6,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    # colormap = "plasma",
    colormap = mckolors$Spectral%>%rev(),
    features = c(
      # "let7_5p_targets1",
      # "mir133_3p_targets1"
      # "mir21_5p_targets1",
      # "mir142b_targets1",
      
      "mir1_3p_targets1",
      "mir206_3p_targets1",
      
      "mir1_3p_targets_spliced1",
      "mir206_3p_targets_spliced1"
    ),
    # alt.titles = c(
    #   "Mir1/Mir206 Target Score",
    #   "Mir133a Target Score"
    # )
  )&theme(
    legend.text = element_text(size=6)
  )&coord_fixed(
    ratio=1/1.6
  )
)
```

Co-expression scatter plots
```{r}
if(!exists("skm.merged")){
  skm.merged <- merge(
    skm.list[[1]],
    skm.list[2:length(skm.list)],
    add.cell.ids = meta_skm$sample
  )
}

# Mir1 ---
plot.list <- lapply(
  skm.list,
  FUN=function(SEU) FeatureScatter(
    SEU,
    feature1 = "mmu-miR-1a-3p",
    slot="counts",
    feature2 = "mir1_3p_targets1",
    shuffle = T,
    cols = viridis(8)[SEU$timepoint[1]+1],
    plot.cor = T,
    pt.size = 0.1,
    jitter = F,
    group.by = "timepoint"
  )+
    scTheme$scatter+
    xlim(c(0,40))+
    ylim(c(-0.1,0.15))+
    theme(
      plot.title = element_text(hjust=0.5),
      axis.title.x = element_blank(),
      axis.text.x = element_blank()
    )+
    scale_alpha_manual(values = 0.5)
) 
plot.list[[length(plot.list)]] <- plot.list[[length(plot.list)]] +
  theme(
    axis.title.x = element_text(color="black"),
    axis.text.x = element_text(color="black")
  )
wrap_plots(
  plot.list,
  ncol=1,
  guides="collect"
)

# Ctss ----
plot.list <- lapply(
  skm.list,
  FUN=function(SEU) FeatureScatter(
    SEU,
    feature1 = "Ctss",
    slot="data",
    feature2 = "mir1_3p_targets1",
    shuffle = T,
    cols = viridis(8)[SEU$timepoint[1]+1],
    plot.cor = T,
    pt.size = 0.1,
    jitter = F,
    group.by = "timepoint"
  )+
    scTheme$scatter+
    xlim(c(0,3.5))+
    ylim(c(-0.1,0.15))+
    theme(
      plot.title = element_text(hjust=0.5),
      axis.title.y = element_blank(),
      axis.title.x = element_blank(),
      axis.text.x = element_blank()
    )+
    scale_alpha_manual(values = 0.5)
) 
plot.list[[length(plot.list)]] <- plot.list[[length(plot.list)]] +
  theme(
    axis.title.y = element_text(color="black",vjust=1,angle = 90),
    axis.title.x = element_text(color="black",face="bold.italic"),
    axis.text.x = element_text(color="black")
  )
cts.plots <- wrap_plots(
  plot.list,
  ncol=1,
  guides="collect"
)
```

Plot co-expression maps
```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE}
suppressMessages(
  visListPlot(
    skm.list,
    sample.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="_SkM"),
    reduction="space",
    assay="kallisto_collapsed", 
    slot = 'counts',
    pt.size=0.4,
    legend.position = "bottom",
    font.size = 6,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    colormap = "plasma",
    # colormap = mckolors$Spectral%>%rev(),
    features = c(
      "let7_5p_targets1","let7_5p_miRs1",
      "mir1_3p_targets1","mir1_3p_miRs1",
      "mir133_3p_targets1","mir133_3p_miRs1" 
    )
  )&theme(
    legend.text = element_text(size=6)
  )&coord_fixed(
    ratio=1/1.6
  )
)
```

# Clustering analysis
Plot...
```{r}
lapply(
  skm.list[c(1:3,5:7)],
  FUN=function(VIS) DimPlot(
    VIS,
    reduction = "space",
    pt.size=0.4,
    cols = list(
      injury_zone=mckolors$primary[1],
      injury_border=mckolors$primary[2],
      myofiber=mckolors$primary[3]
    ),
    group.by = c(
      "injury.zones"
      # "kallisto_collapsed_snn_res.1"
    )
  )+
    ggtitle(VIS$sample[1])+
    theme(
      plot.title=element_text(color="black"),
      plot.margin = unit(rep(0,4), units="in")
      )
) %>% 
  lapply(
    FUN=function(VIS) VIS +
      scTheme$umap +
      theme(
        axis.title = element_blank(),
        axis.line =  element_blank(),
        plot.title=element_blank()
      )
  ) %>% 
  wrap_plots(
    guides="collect"
  )&coord_fixed(ratio=1/1.6)
```
Differential gene expression across clusters
```{r}
Idents(skm.merged)<- "injury.zones"
injury.markers <- subset(
  skm.merged,
  cells= Cells(skm.merged)[skm.merged$polyA=="yPAP"]
) %>% FindAllMarkers(
  # skm.merged,
  assay = "kallisto_collapsed",
  densify = T
)

# add biotype info to output
##TODO: not sure why this isn't working; gives wrong biotype...
# injury.markers$biotype <- lapply(
#   injury.markers$gene,
#   FUN=function(X) gtf.info$Biotype[gtf.info$GeneSymbol==X]
# )%>%unlist()

# tmp <- list()
for(GENE in rownames(injury.markers)){
  tmp.gene = injury.markers[GENE,'gene']
  tmp <- gtf.info$Biotype[gtf.info$GeneSymbol==tmp.gene]%>%head(n=1)
  if(length(tmp)==1){
    injury.markers[GENE,'biotype'] <- tmp
  }else{
    injury.markers[GENE,'biotype'] <- "unknown"
  }
}
# injury.markers$biotype <- unlist(tmp)
write.csv(
  injury.markers,
  file="/workdir/dwm269/totalRNA/spTotal/resources/injury_markers.csv"
)
cat("Done.")
```

Differential gene expression across chemistries
```{r}
Idents(skm.merged)<- "polyA"
pa.markers <-  FindAllMarkers(
  skm.merged,
  assay = "kallisto_collapsed",
  densify = T
)

# add biotype info to output
for(GENE in rownames(pa.markers)){
  tmp.gene = pa.markers[GENE,'gene']
  tmp <- gtf.info$Biotype[gtf.info$GeneSymbol==tmp.gene]%>%head(n=1)
  if(length(tmp)==1){
    pa.markers[GENE,'biotype'] <- tmp
  }else{
    pa.markers[GENE,'biotype'] <- "unknown"
  }
}

write.csv(
  pa.markers,
  file="/workdir/dwm269/totalRNA/spTotal/resources/polyA_markers.csv"
)
cat("Done.")
```

# Spliced vs. unspliced analysis
Add spliced & unspliced count matrices (from kallisto la_manno workflow) to the Seurat objects as new assays
```{r}
require("R.utils")
NCORES=12
for(i in 1:length(skm.list)){
  
  # Spliced ----
  tmp.seu <- ReadMtx(
    mtx=paste0(meta_skm$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/spliced.mtx"),
    cells=paste0(meta_skm$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/spliced.barcodes.txt"),
    features=paste0(meta_skm$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/spliced.genes.txt"),
    feature.column=1,
    mtx.transpose = T
  ) %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(skm.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  # convert ensembl IDs to MGI gene IDs
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  rownames(tmp.mat) <- ens2gene(
    ens=rownames(tmp.mat),
    biomart.info = gtf.info,
    ens.colname = "GENEID", # Ensembl ID, yes periods
    # ens.colname = "GeneID", # ensembl IDs, no periods
    gene.colname = "GeneSymbol",
    force.unique = T,
    ncores=NCORES,
    verbose=T
  )
  
  # add kallisto counts to vis_list
  skm.list[[i]][["spliced"]] <- CreateAssayObject(
    counts = tmp.mat,
    min.cells = 1
  )
  
  
  # Unspliced ----
  tmp.seu <- ReadMtx(
    mtx=paste0(meta_skm$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/unspliced.mtx"),
    cells=paste0(meta_skm$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/unspliced.barcodes.txt"),
    features=paste0(meta_skm$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/unspliced.genes.txt"),
    feature.column=1,
    mtx.transpose = T
  ) %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(skm.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  # convert ensembl IDs to MGI gene IDs
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  rownames(tmp.mat) <- ens2gene(
    ens=rownames(tmp.mat),
    biomart.info = gtf.info,
    ens.colname = "GENEID", # Ensembl ID, yes periods
    # ens.colname = "GeneID", # ensembl IDs, no periods
    gene.colname = "GeneSymbol",
    force.unique = T,
    ncores=NCORES,
    verbose=T
  )
  
  # add kallisto counts to vis_list
  skm.list[[i]][["unspliced"]] <- CreateAssayObject(
    counts = tmp.mat,
    min.cells = 1
  )
}

rm(tmp.mat)
rm(tmp.seu)
gc()
```
Normalize the spliced/unspliced read counts for visualization
```{r}
skm.list <- lapply(
  skm.list,
  FUN=function(SEU){
    SEU <- NormalizeData(
      SEU,
      assay="spliced"
    )
    
    SEU <- NormalizeData(
      SEU,
      assay="unspliced"
    )
    
    return(SEU)
  }
)
message("Done.")
```
Find top unspliced genes
```{r}
lapply(
  skm.list[c(4:7)],
  FUN=function(SEU) grepGenes(
    SEU,
    assay="unspliced",
    pattern=""
  )%>% head(n=10)
)

```

### Look at unspliced biotype scores 
```{r paged.print=TRUE}
skm.list <- lapply(
  skm.list,
  FUN = function(SEU) seu_biotypes(
    SEU=SEU,
    biomart=gtf.info,
    gene.colname = "GeneSymbol",
    biotype.colname = "Biotype",
    prefix = "pct.unspliced.",
    assay = "unspliced"
  )
)%>%suppressMessages()
```

Plot maps for unspliced gene biotype
```{r}
suppressMessages(
  visListPlot(
    skm.list,
    sample.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="_SkM"),
    reduction="space",
    assay="kallisto_unspliced", 
    slot = 'data',
    pt.size=0.4,
    legend.position = "bottom",
    font.size = 6,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    # colormap = "plasma",
    colormap = mckolors$Spectral%>%rev(),
    features = c(
      "pct.unspliced.protein_coding", "pct.unspliced.lncRNA", "pct.snRNA"
    )
  )&theme(
    legend.text = element_text(size=6)
  )&coord_fixed(
    ratio=1/1.6
  )
)
```

### Visualize gene splicing ratios with visCoMap
```{r fig.height=8, fig.width=6}
tmp.feat = "Hexb"
suppressMessages(
  visCoMap(
    skm.list,
    sample.titles = meta_skm$sample,
    reduction="space",
    slot = c('data','data'),
    pt.size=0.6,
    legend.position = "bottom",
    font.size = 8,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    # colormap = mckolors$Spectral,
    # colormap.direction = 1,
    comap.fxn = function(X) X[1]/X[2],
    coex.name = "spliced/unspliced", 
    alt.titles = paste(tmp.feat,c("spliced", "unspliced")),
    features = c(tmp.feat, tmp.feat),assay=c("spliced","unspliced"),
  )&coord_fixed(ratio=1/1.6)
)
```
# Spatially variable features
Find spatially variable features
```{r}
library(parallel)
num.cores=9
cl <- makeCluster(num.cores)

skm.list <- mclapply(
  skm.list,
  FUN = function(SEU) FindSpatiallyVariableFeatures(
      SEU,
      assay = "kallisto_collapsed", 
      # features = VariableFeatures(SEU)[1:1000],
      selection.method = "markvariogram"
  ),
  mc.cores=num.cores
)

```

Pull out top features
```{r}
top.spatial <- lapply(
  skm.list,
  FUN=function(VIS) SpatiallyVariableFeatures(
    VIS, selection.method = "markvariogram"
  ) %>%
    head(n=50)
) 

tmp.spatial <- top.spatial %>%
  unlist() %>%
  unique()
# tmp.spatial = tmp.spatial[tmp.spati]
tmp.spatial
```

Plot some of the top features
```{r fig.height=6, fig.width=16, message=FALSE, warning=FALSE}
suppressMessages(
  visListPlot(
    skm.list,
    sample.titles = stringr::str_remove_all(meta_skm$sample, pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="_SkM"),
    reduction="space",
    slot = 'data',
    pt.size=0.25,
    legend.position = "bottom",
    font.size = 8,
    axis.title.angle.y=0,
    colormap = mckolors$cubs,
    nrow = 1,
    combine = T,
    verbose=F,
    features = c("Zc3h7a","ENSMUSG00002076173"),assay=c("unspliced","kallisto"),
    
    # features = tmp.spatial[tmp.spatial %in% genes.ncrna] %>%rev()%>% head(n=12)
    # features=grepGenes(skm.list[[7]],assay="kallisto_collapsed",pattern = "Mir",filter.pattern = "hg")[1:12]
    # features = grepGenes(skm.list[[6]],assay="kallisto_collapsed",pattern = genes$miRNA,filter.pattern = "ENS")[1:12]
    # features=grepGenes(skm.list[[6]],assay="kallisto_collapsed",pattern = c("Sno","SNO"))[1:12]
  )&theme(
    legend.text = element_text(size=6)
  )
)
```

# DGEA across injury timepoints (in spTotal samples)
```{r}
tmp.seu <- merge(
  x = skm.list[[4]],
  y = skm.list[5:7],
  add.cell.ids = meta_skm$sample[4:7]
)

Idents(tmp.seu)<- "sample"
  
skm.injury.markers <- FindAllMarkers(
  tmp.seu,
  assay = "kallisto_collapsed"
)
```

DGEA across polyA, timepoint, and injury zone
```{r}
if(!exists("skm.merged")){
  skm.merged <- merge(
    skm.list[[1]],
    skm.list[2:length(skm.list)],
    add.cell.ids = meta_skm$sample
  )
}
pa.injury.markers <- list()
Idents(skm.merged)<-"polyA"
for(iz in c("myofiber","injury_border","injury_zone")){
  for(tp in c(2,5,7)){
    pa.injury.markers[[paste0(iz,"_",tp)]] <- subset(
      skm.merged,
      cells=Cells(skm.merged)[skm.merged$timepoint==tp & skm.merged$injury.zones==iz]
    )%>% FindMarkers(
      # skm.merged,
      ident.1="ctrl",
      ident.2="yPAP",
      # cells.1 = Cells(skm.merged)[skm.merged$polyA=="ctrl" & skm.merged$timepoint==timepoint],
      # cells.2 = Cells(skm.merged)[skm.merged$polyA=="yPAP" & skm.merged$timepoint==timepoint],
      assay = "kallisto_collapsed"
    )
  }
}
```

Browse output...
```{r}
i=2
ggVolcano_v2(
  pa.injury.markers[[i]][pa.injury.markers[[i]]$pct.2<0.3,],repel = F
)

```
Filter...
```{r}
tmp.feat <- skm.injury.markers$gene[
  skm.injury.markers$cluster=="yPAP-Pro_SkM-D2"&
    skm.injury.markers$pct.2<0.3&
    skm.injury.markers$avg_log2FC>0.5
    # skm.injury.markers$p_val<10^-10
  ]

tmp.biotypes <- lapply(
  tmp.feat,
  FUN=function(X) gtf.info$Biotype[gtf.info$GeneSymbol==X]
)

tmp.feat[tmp.biotypes!="protein_coding"]
```

```{r}
pa.injury.markers[[5]][pa.injury.markers[[5]]$pct.1<0.1,]

tmp.feat=rownames(pa.injury.markers[['myofiber_7']])[pa.injury.markers[['myofiber_7']]$pct.1<0.1]
pa.injury.markers[['injury_zone_5']][!rownames(pa.injury.markers[['injury_zone_5']])%in%tmp.feat,]
```

Plot!
```{r fig.height=8, fig.width=8}
suppressMessages(
  visListPlot(
    skm.list,
    sample.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="_SkM"),
    reduction="space",
    assay="kallisto_collapsed", 
    slot = 'data',
    pt.size=0.4,
    legend.position = "bottom",
    font.size = 6,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    colormap = "plasma",
    # colormap = mckolors$Spectral%>%rev(),
    features = c(
      # "Gm25939",
      # "ENSMUSG00002075551",
      # "Ttyh1", #?
      # "Rpph1",
      # "Rmrp",
      # "H4c4",
      # "mt-Th",
      # "7SK.1",
      # "Rny1",
      # "Rny3"
      "Atp5o"
      # "Notch1",
      # "Gm13568"
      # "Vaultrc5",
      # "Gm43846"
      
    )
  )&theme(
    legend.text = element_text(size=6)
  )&coord_fixed(
    ratio=1/1.6
  )
)

# "Gm42899" "Gm43972" "Tmem38b",
```

# cis-NAT analysis
Load and filter table of cis-NATS
```{r}
cisnat.info <- read.csv(
  file="/local/workdir/dwm269/totalRNA/spTotal/resources/gene_lists/overlap_M28.tsv",
  # file=paste0(getwd(),"/resources/gene_lists/overlap_M28.tsv"),
  sep="\t"
)

# Get list of all genes
tmp.feat.ctrl <- lapply(
  skm.list[1:3],
  FUN = function(SEU) Features(SEU, assay="kallisto_collapsed")
)%>%
  unlist()%>%
  unique()

tmp.feat.pa <- lapply(
  skm.list[4:7],
  FUN = function(SEU) Features(SEU, assay="kallisto_collapsed")
)%>%
  unlist()%>%
  unique()

# cisnat.info <- cisnat.info[cisnat.info$GeneSymbol_Plus %in% tmp.feat.pa[!tmp.feat.pa%in%tmp.feat.ctrl],]

# Keep pairs found in the polyA samples
cisnat.info <- cisnat.info[cisnat.info$GeneSymbol_Plus %in% tmp.feat.pa & cisnat.info$GeneSymbol_Minus %in% tmp.feat.pa,]
# Remove piars found in the CTRL datasets (to focus on pairs only visible in polyA)
cisnat.info <- cisnat.info[!cisnat.info$GeneSymbol_Plus %in% tmp.feat.ctrl & cisnat.info$GeneSymbol_Minus %in% tmp.feat.ctrl,]

cisnat.info$pair <- paste0(cisnat.info$GeneSymbol_Plus,"_",cisnat.info$GeneSymbol_Minus)

rownames(cisnat.info)<-1:nrow(cisnat.info)
cisnat.info
```

Compute pseudobulk log2-fold expression ratios for each sample
```{r message=FALSE, warning=FALSE}
ratio.list <- list()
for(i in 1:nrow(cisnat.info)){
  tmp.feat = c(cisnat.info$GeneSymbol_Plus[i],cisnat.info$GeneSymbol_Minus[i])
  ratio.list[[i]] <- lapply(
    skm.list[4:7],
    FUN = function(SEU){
      if(all(tmp.feat%in% Features(SEU, assay="kallisto_collapsed"))){
        tmp = GetAssayData(
          SEU,
          assay="kallisto_collapsed",
          slot = "data"
        )[tmp.feat,]%>%rowMeans()

        out = as.numeric(log2(tmp[1]/tmp[2]))
      }else{
        out=NA
      }
      return(out)
    }
  )
  ratio.list[[i]] <- do.call(cbind,ratio.list[[i]])
  colnames(ratio.list[[i]]) <- meta_skm$sample[4:7]
}

tmp.df <- do.call(rbind, ratio.list)
cisnat.info <- cbind(cisnat.info, tmp.df)
rm(tmp.df)
```

```{r}
unique(cisnat.info$GeneSymbol_Minus)%>%sort()
```


Plot co-expression
```{r fig.height=8, fig.width=10, message=FALSE, warning=FALSE}
tmp.feat = c("Mybpc1","Gm48752")
tmp.feat = c("Notch1","Gm13568")
tmp.feat = c("Myh1","Myhas")
tmp.feat = c("Hexb","Gfm2")
tmp.feat = c("Acvr2a","Orc4")
tmp.feat = c("Mrpl58","Ict1os")
tmp.feat = c("Bmp4","Gm15222")
tmp.feat = c("Map2k1","Uchl4")
tmp.feat = c("Pard6a","Enkd1") # no overlap
tmp.feat = c("Mylpf","Septin1") # no overlap
tmp.feat = c("Igfbp7","Polr2b")
tmp.feat = c("Ctsb","Fdft1")
tmp.feat = c("Adgrg6","Gm10335")
tmp.feat = c("Mcc","A930012L18Rik")
tmp.feat = c("Trim14","Gm22247")
tmp.feat = c("Pds5a","Gm42727")
tmp.feat = c("1600010M07Rik","Gm25636")
tmp.feat = c("1600010M07Rik","Snora23a")
tmp.feat = c("Gm38384","Gm5914")
tmp.feat = c("Glg1","ENSMUSG00002075523")
tmp.feat = c("Nr6a1","Nr6a1os")
tmp.feat = c("H3c13","Gm20628")
tmp.feat = c("Myh2","9430073C21Rik")
tmp.feat = c("Myh1","Myhas")
tmp.feat = c("Rn18s-rs5","AY036118")
  
 
suppressMessages(
  visCoMap(
    skm.list[4:7],
    sample.titles = meta_skm$sample,
    reduction="space",
    slot = c('data','data'),
    pt.size=0.3,
    legend.position = "bottom",
    font.size = 6,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    # colormap = mckolors$Spectral,
    # colormap.direction = 1,
    # comap.fxn = function(X) X[1]/X[2],
    coex.name = paste(tmp.feat[1], tmp.feat[2],sep = " x "), alt.titles = tmp.feat,
    features = tmp.feat,
    assay=c("kallisto_collapsed","kallisto_collapsed"),
    # scatter.group.by = "injury.zones",
    scatter.theme = scTheme$scatter,
    include.scatter = T
  )&coord_fixed(ratio=1/1.6)
)
```


# snoRNAs
Get top snoRNAs
```{r}
top.snornas <- lapply(
  skm.list[4:7],
  FUN=function(SEU) grepGenes(
    SEU = SEU,
    pattern= c("Sno","SNO"),
    assay="kallisto_collapsed"
  )%>% head(n=10)
)

print(unlist(top.snornas)%>%unique())
```

Plot them
```{r fig.height=10, fig.width=8}
tmp.feat = c("Mir133a-1hg","Mir1a-1","Mir206","Mir133b")
tmp.assay=c("STARsolo_collapsed","STARsolo_collapsed","STARsolo_collapsed","STARsolo_collapsed")

suppressMessages(
  visListPlot(
    skm.list,
    sample.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="_SkM"),
    reduction="space",
    assay=tmp.assay, 
    slot = 'data',
    pt.size=0.4,
    legend.position = "bottom",
    font.size = 6,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    colormap = "plasma",
    # colormap = mckolors$Spectral%>%rev(),
    # features = unlist(top.snornas)%>%unique()%>% head(n=5)
    features=tmp.feat
  )&theme(
    legend.text = element_text(size=6)
  )&coord_fixed(
    ratio=1/1.6
  )
)
```


```{r}
tmp.feat=c(
  # "mmu-let-7a-5p",
  # "mmu-let-7c-5p",
  # "mmu-let-7d-5p",
  # "mmu-miR-206-3p",
  # "mmu-miR-133a-3p/133b-3p", 
  # "mmu-miR-142a-3p",
  # "mmu-miR-145a-5p",
  "mmu-miR-1a-3p" 
)%>%sort()

suppressMessages(
  visListPlot(
    skm.list[c(6,7)],
    sample.titles = meta_skm$sample[c(6,7)],
    alt.titles = tmp.feat,
    reduction = "space",
    assay="mirge3",
    slot="counts",
    pt.size = 0.1,
    # features="nCount_mirge3",
    features=tmp.feat,
    axis.title.angle.y = 0,
    legend.position = "right",
    combine=T,ncol = 4,
    colormap = "plasma",
    colormap.direction = -1,
    colormap.same.scale = F
  )&coord_fixed(
      ratio = 1/1.6
    )
)
```
# Session Info
```{r}
sessionInfo()
```