#  Additional processing for the skeletal muscle samples (standard visium & polyA visium in notexin-injured tibialis anterior muscles)
## David W. McKellar

# Session setup
## Libs, setwd
```{r message=FALSE, warning=FALSE}
library(Matrix)
library(dplyr)
library(Seurat)
library(future)

library(ggplot2)
library(patchwork)
library(viridis)
library(data.table)
library(shades)

source("/home/dwm269/DWM_utils/plotting_utils/scThemes.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seutils.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seuplots.R")

setwd("/local/workdir/dwm269/totalRNA/spTotal/")
```

## Figure settings & colors
```{r}
# fonts, sizes, etc.
small.font = 6*2
big.font = 8*2
line.width = 0.5
pt.size=0.01
pt.stroke=0.3
label.size=2
```

## Plot themes 
```{r}
scTheme <- scThemes(
  small.font = small.font,
  big.font = big.font,
  line.width = line.width,
  pt.size=pt.size,
  pt.stroke=pt.stroke,
  label.size=label.size
)

print(names(scTheme))
```
```{r}
mckolors <- read.csv("/home/dwm269/DWM_utils/plotting_utils/McKolors_v1.csv") %>% 
  as.list() %>%
  lapply(
    FUN=function(X) X[X!=""]
  )
names(mckolors)
```

# Load and subset preprocessed Visium +/- spTotal data
```{r}
if(!exists("vis.list")){
  load("/local/workdir/dwm269/totalRNA/spTotal/robjs/vis_list_v5.RData")
}
cat("Done.")
```

# Load metadata
```{r}
meta_vis <- read.csv("/workdir/dwm269/totalRNA/spTotal/resources/metadata_sheets/meta_sheet_visium.csv")
meta_skm <- meta_vis[meta_vis$tissue=="muscle"&meta_vis$rnase_inhib%in%c("ctrl", "Protector"),] #subset metadata

skm.list <- vis.list[meta_vis$tissue=="muscle"&meta_vis$rnase_inhib%in%c("ctrl", "Protector")] # subset out Seurat objects
skm.list <- skm.list[with(meta_skm, order(meta_skm$polyA,meta_skm$timepoint))] # reorder list for visualization

meta_skm <- meta_skm[with(meta_skm, order(meta_skm$polyA,meta_skm$timepoint)),] #sort metadata (by preprocessing, then by injury timepoint)

meta_skm
```

# Knee plots by assay
```{r}
lapply(
  skm.list,
  FUN=function(VIS){
    qplot(
      1:length(VIS$nCount_kallisto),
      sort(VIS$nCount_kallisto, decreasing=T)
    )+
      scTheme$scatter+
      theme(axis.title.x = element_blank())+
      labs(
        y="nCount_kallisto",
        title=VIS$sample[1]
      )+
      scale_y_continuous(limits = c(0,75000))
  }
)%>%
  wrap_plots()&scTheme$scatter

```

# Get gene biotype info from the GENCODE annotations
`gtf.info` is just a reformatted .tsv of the GENCODE M28 annotations .gtf
```{r}
gtf.info <- read.csv(
  file="/workdir/dwm269/totalRNA/spTotal/resources/gene_lists/GRCm39_GENCODEm28_gene_info_gtf.tsv",
  sep = "\t"
)

# Remove period/decimal suffixes from ensembl IDs
gtf.info$GeneID <- strsplit(
  x=gtf.info$GENEID,
  split = "\\."
)%>%
  lapply(
    FUN=function(X) X[[1]]
  )%>%
  unlist()

cat(
  paste("Using", length(unique(gtf.info$GeneSymbol)), "genes, and ",length(unique(gtf.info$Biotype)),"biotypes...\n")
)

head(gtf.info)
```

## gene lists by biotype
```{r}
genes = list()
for(biotype in unique(gtf.info$Biotype)){
  genes[[biotype]] <- gtf.info$GENEID[gtf.info$Biotype==biotype] %>%
    ens2gene(
      biomart.info = gtf.info,
      ens.colname = "GENEID",
      gene.colname = "GeneSymbol"
      ) %>%
    unique()
}

names(genes)
```

## Gene biotype spatial plots
```{r warning=FALSE}
visListPlot(
  skm.list,
  sample.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
        stringr::str_remove_all(pattern ="yPAP_")%>%
        stringr::str_remove_all(pattern ="ctrl_")%>%
        stringr::str_remove_all(pattern ="_SkM"),
  # assay="STARsolo_collapsed",
  reduction="space",
  # slot = 'counts',
  pt.size=0.6,
  legend.position = "bottom",
  font.size = 12,
  axis.title.angle.y=0,
  nrow = 1,
  combine = T,
  verbose=F,
  # colormap="inferno",colormap.direction = -1,
  colormap = mckolors$RdYlBu%>%rev(),
  features = c(
    # "nFeature_STARsolo","nCount_STARsolo","percent.mt",
    "pct.protein_coding",
    # "pct.lncRNA",
    "pct.rRNA",
    # "pct.Mt_rRNA", 
    "pct.Mt_tRNA",
    # "pct.ribo.pp",
    "pct.snoRNA",
    "pct.miRNA",
    "pct.snRNA"
  ) 
)%>% suppressMessages()
```
# Read in STARsolo/miRbase miRNA outputs
```{r}
require("R.utils")
NCORES=20
for(i in 1:length(skm.list)){
   # tmp.seu <- Read10X(
   #    data.dir = paste0(meta_skm$data.dir.STARsolo[i], "/mirbase/Solo.out/Gene/raw/")
   #  ) %>%
   #  CreateSeuratObject() %>%
   #  subset( # select cells
   #    cells=Cells(skm.list[[i]])
   #  )
  tmp.seu <- ReadMtx(
    mtx=paste0(meta_skm$data.dir.STARsolo[i],"/mirbase/Solo.out/Gene/raw/matrix.mtx"),
    cells=paste0(meta_skm$data.dir.STARsolo[i],"/mirbase/Solo.out/Gene/raw/barcodes.tsv"),
    features=paste0(meta_skm$data.dir.STARsolo[i],"/mirbase/Solo.out/Gene/raw/features.tsv"),
    feature.column=1,
    mtx.transpose = F
  ) %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(skm.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  print(dim(tmp.seu))
  
  # Add zeroes for spots missing miRNA reads
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  zero.cells <- Cells(skm.list[[i]])[!Cells(skm.list[[i]])%in%Cells(tmp.seu)]
  if(length(zero.cells)>0){
    zero.mat <- matrix(
      data=0,
      ncol = length(zero.cells),
      nrow = nrow(tmp.mat)
    )
    colnames(zero.mat) <- zero.cells
    rownames(zero.mat) <- rownames(zero.mat)
    
    tmp.mat <- cbind(tmp.mat,zero.mat)
    
    
    tmp.mat <- tmp.mat[,Cells(skm.list[[i]])] # make sure cells/spots are in the right order
    
  }
  
  # add kallisto counts to vis_list
  skm.list[[i]][["STAR_miRNA"]] <- CreateAssayObject(
    counts = tmp.mat
    # min.cells = 1
  )
}

rm(tmp.mat)
rm(tmp.seu)
gc()
```
Normalize miRNA counts
```{r}
skm.list <- lapply(
  skm.list,
  FUN=function(SEU) NormalizeData(
    SEU,
    assay = "STAR_miRNA"
  ) 
)
```


# Read in kallisto/miRbase miRNA outputs
```{r}
require("R.utils")
NCORES=20
for(i in 1:length(skm.list)){
  
  tmp.seu <- ReadMtx(
    mtx=paste0(meta_skm$data.dir.kallisto[i],"/kb_miRNA/counts_unfiltered/matrix.mtx"),
    cells=paste0(meta_skm$data.dir.kallisto[i],"/kb_miRNA/counts_unfiltered/barcodes.tsv"),
    features=paste0(meta_skm$data.dir.kallisto[i],"/kb_miRNA/counts_unfiltered/genes.tsv"),
    feature.column=1,
    mtx.transpose = F
  ) %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(skm.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  print(dim(tmp.seu))
  
  # Add zeroes for spots missing miRNA reads
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  zero.cells <- Cells(skm.list[[i]])[!Cells(skm.list[[i]])%in%Cells(tmp.seu)]
  if(length(zero.cells)>0){
    zero.mat <- matrix(
      data=0,
      ncol = length(zero.cells),
      nrow = nrow(tmp.mat)
    )
    colnames(zero.mat) <- zero.cells
    rownames(zero.mat) <- rownames(zero.mat)
    
    tmp.mat <- cbind(tmp.mat,zero.mat)
    
    
    tmp.mat <- tmp.mat[,Cells(skm.list[[i]])] # make sure cells/spots are in the right order
    
  }
  
  # add kallisto counts to vis_list
  skm.list[[i]][["miRNA"]] <- CreateAssayObject(
    counts = tmp.mat
    # min.cells = 1
  )
}

rm(tmp.mat)
rm(tmp.seu)
gc()
```
Normalize miRNA counts
```{r}
skm.list <- lapply(
  skm.list,
  FUN=function(SEU) NormalizeData(
    SEU,
    assay = "miRNA"
  ) 
)
```

# Add miRge3 outputs
```{r}
require("R.utils")
NCORES=8
for(i in 1:length(skm.list)){ #length(skm.list)
  
  #TODO
  # tmp.seu <- ReadMtx(
  #   mtx=paste0(meta_skm$data.dir.kallisto[i],"/mirge3_matrix/",meta_skm$sample[i],"/matrix.mtx.gz"),
  #   cells=paste0(meta_skm$data.dir.kallisto[i],"/mirge3_matrix/",meta_skm$sample[i],"/barcodes.tsv.gz"),
  #   features=paste0(meta_skm$data.dir.kallisto[i],"/mirge3_matrix/",meta_skm$sample[i],"/features.tsv.gz"),
  #   feature.column=1,
  #   mtx.transpose = F
  # ) %>%
  #   CreateSeuratObject() %>%
  #   subset( # select cells
  #     cells=Cells(skm.list[[i]])
  #   )
  
  # mirge.list[1:3] <- lapply(
  #   mirge.list[1:3],
  #   FUN=function(X){ 
  #     rownames(X)<-rownames(mirge.list[[5]])
  #     return(X)
  #   }
  # )
  tmp.seu <- mirge.list[[i]] %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(skm.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  # Add zeroes for spots missing miRNA reads
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  zero.cells <- Cells(skm.list[[i]])[!Cells(skm.list[[i]])%in%Cells(tmp.seu)]
  if(length(zero.cells)>0){
    zero.mat <- matrix(
      data=0,
      ncol = length(zero.cells),
      nrow = nrow(tmp.mat)
    )
    colnames(zero.mat) <- zero.cells
    rownames(zero.mat) <- rownames(zero.mat)
    
    tmp.mat <- cbind(tmp.mat,zero.mat)
    
    
    tmp.mat <- tmp.mat[,Cells(skm.list[[i]])] # make sure cells/spots are in the right order
    
  }
  
  # Log-normalize the miR counts according to the kallisto UMI counts
  tmp.norm.mat <- log1p(tmp.mat/skm.list[[i]]$nCount_kallisto*10000)
    
  # add counts to skm.list
  skm.list[[i]][["mirge3"]] <- CreateAssayObject(
    counts = tmp.mat
    # data = tmp.norm.mat
  )
  
  # add log-norm data
  skm.list[[i]][["mirge3"]]@data <- tmp.norm.mat
}

rm(tmp.mat)
rm(tmp.seu)
gc()
```

# Print highly expressed miRNAs for each object
```{r}
# mirnas <- gtf.info$GENEID[gtf.info$Biotype=="miRNA"]
top.mirnas <- lapply(
  skm.list,
  FUN = function(VIS) grepGenes(
    VIS,
    assay="STAR_miRNA",
    pattern = "",
    filter.pattern = c("miR-7", "miR-6"),
    verbose = F
  )
)

lapply(
  top.mirnas[4:7],
  FUN = function(X) X %>%
    head(n=20) %>%
    print()
)
```

# Print highly expressed tRNAs for each object
```{r}
top.mirs <- lapply(
  skm.list,
  FUN = function(VIS) grepGenes(
    VIS,
    assay="kallisto_collapsed",
    pattern = genes$miR,
    filter.pattern = c("hg","ENS")
  ) 
)

lapply(
  top.mirs,
  FUN = function(X) X %>%
    head(n=20) %>%
    print()
)
```

# Clustering analysis
Multi-resolution clustering 
```{r}
resolutions = 0.8#c(0.4,0.6,0.8,1,1.2)
skm.list <- skm.list %>% 
  lapply(
    FUN=function(VIS){
      for(RES in resolutions){
        VIS = FindClusters(
          object=VIS,
          graph="kallisto_collapsed_snn",
          resolution=RES,
          verbose=F
        )
      }
      return(VIS)
    }
  )
```

Plot 
```{r}
i=1
DimPlot(
    skm.list[[3]],
    reduction="space",
    pt.size=2,
    group.by = "kallisto_collapsed_snn_res.0.8",
    cols = mckolors$colblind_8
  )+
    coord_fixed(1/1.6)+
    scTheme$space
# wrap_plots(
#   FeaturePlot(
#     skm.list[[i]],
#     reduction="space",
#     pt.size=2,combine = F,
#     features=c("Myh1","Myh4","Ctss")
#   )+NoLegend()#&scTheme$space#&scale_color_viridis()
# )
```

Add cluster labels...
```{r}
tmp.labels <- list(
  CTRLD2=c(
    "injury_zone",
    "myofiber",
    "injury_border",
    "injury_zone",
    "myofiber"
  ),
  CTRLD5=c(
    "injury_zone",
    "myofiber",
    "injury_border",
    "myofiber",
    "injury_zone",
    "injury_border"
  ),
  CTRLD7=c(
    "injury_border",
    "injury_zone",
    "injury_border",
    "myofiber",
    "injury_zone",
    "myofiber",
    "injury_border"
  ),
  yPAPD0=c(
    "myofiber",
    "myofiber",
    "myofiber",
    "myofiber",
    "myofiber",
    "myofiber"
  ),
  yPAPD2=c(
    "injury_zone",
    "injury_zone",
    "myofiber",
    "injury_border",
    "myofiber"
  ),
  yPAPD5=c(
    "myofiber",
    "injury_zone",
    "injury_border",
    "injury_zone"
  ),
  yPAPD7=c(
    "myofiber",
    "myofiber",
    "myofiber",
    "injury_border",
    "myofiber",
    "myofiber"
  )
)

for(i in 1:length(skm.list)){
  skm.list[[i]]$kallisto_collapsed_res.0.8 <- as.numeric(skm.list[[i]]$kallisto_collapsed_snn_res.0.8)
  skm.list[[i]] <- AddCellTypeIdents(
    skm.list[[i]],
    old.name="kallisto_collapsed_res.0.8",
    new.name = "injury.zones",
    new.idents = tmp.labels[[i]]
  )
  skm.list[[i]]$injury.zones <- factor(
    skm.list[[i]]$injury.zones,
    levels=c(
       "injury_zone" ,"injury_border","myofiber"
    )
  )
}
```

PLot...
```{r}
lapply(
  skm.list[c(1:3,5:7)],
  FUN=function(VIS) DimPlot(
    VIS,
    reduction = "space",
    pt.size=0.4,
    cols = list(
      injury_zone=mckolors$primary[1],
      injury_border=mckolors$primary[2],
      myofiber=mckolors$primary[3]
    ),
    group.by = c(
      "injury.zones"
      # "kallisto_collapsed_snn_res.1"
    )
  )+
    ggtitle(VIS$sample[1])+
    theme(
      plot.title=element_text(color="black"),
      plot.margin = unit(rep(0,4), units="in")
      )
) %>% 
  lapply(
    FUN=function(VIS) VIS +
      scTheme$umap +
      theme(
        axis.title = element_blank(),
        axis.line =  element_blank(),
        plot.title=element_blank()
      )
  ) %>% 
  wrap_plots(
    guides="collect"
  )&coord_fixed(ratio=1/1.6)
```


# Spliced vs. unspliced analysis
Add spliced & unspliced count matrices (from kallisto la_manno workflow) to the Seurat objects as new assays
```{r}
require("R.utils")
NCORES=12
for(i in 1:length(skm.list)){
  
  # Spliced ----
  tmp.seu <- ReadMtx(
    mtx=paste0(meta_skm$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/spliced.mtx"),
    cells=paste0(meta_skm$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/spliced.barcodes.txt"),
    features=paste0(meta_skm$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/spliced.genes.txt"),
    feature.column=1,
    mtx.transpose = T
  ) %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(skm.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  # convert ensembl IDs to MGI gene IDs
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  rownames(tmp.mat) <- ens2gene(
    ens=rownames(tmp.mat),
    biomart.info = gtf.info,
    ens.colname = "GENEID", # Ensembl ID, yes periods
    # ens.colname = "GeneID", # ensembl IDs, no periods
    gene.colname = "GeneSymbol",
    force.unique = T,
    ncores=NCORES,
    verbose=T
  )
  
  
  # Log-normalize counts according to the total kallisto UMI counts
  tmp.norm.mat <- log1p(tmp.mat/skm.list[[i]]$nCount_kallisto*10000)
  
  # add kallisto counts to vis_list
  skm.list[[i]][["spliced"]] <- CreateAssayObject(
    counts = tmp.mat,
    min.cells = 1
  )
  
  # add log-norm data
  skm.list[[i]][["spliced"]]@data <- tmp.norm.mat
  
  
  # Unspliced ----
  tmp.seu <- ReadMtx(
    mtx=paste0(meta_skm$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/unspliced.mtx"),
    cells=paste0(meta_skm$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/unspliced.barcodes.txt"),
    features=paste0(meta_skm$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/unspliced.genes.txt"),
    feature.column=1,
    mtx.transpose = T
  ) %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(skm.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  # convert ensembl IDs to MGI gene IDs
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  rownames(tmp.mat) <- ens2gene(
    ens=rownames(tmp.mat),
    biomart.info = gtf.info,
    ens.colname = "GENEID", # Ensembl ID, yes periods
    # ens.colname = "GeneID", # ensembl IDs, no periods
    gene.colname = "GeneSymbol",
    force.unique = T,
    ncores=NCORES,
    verbose=T
  )
  
  # Log-normalize counts according to the total kallisto UMI counts
  tmp.norm.mat <- log1p(tmp.mat/skm.list[[i]]$nCount_kallisto*10000)
  
  # add kallisto counts to vis_list
  skm.list[[i]][["unspliced"]] <- CreateAssayObject(
    counts = tmp.mat,
    min.cells = 1
  )
  
  # add log-norm data
  skm.list[[i]][["unspliced"]]@data <- tmp.norm.mat
}

rm(tmp.mat)
rm(tmp.seu)
gc()
```
Normalize the spliced/unspliced read counts for visualization
```{r}
# skm.list <- lapply(
#   skm.list,
#   FUN=function(SEU){
#     SEU <- NormalizeData(
#       SEU,
#       assay="spliced"
#     )
#     
#     SEU <- NormalizeData(
#       SEU,
#       assay="unspliced"
#     )
#     
#     return(SEU)
#   }
# )
# message("Done.")
```


### Look at unspliced biotype scores 
```{r paged.print=TRUE}
skm.list <- lapply(
  skm.list,
  FUN = function(SEU) seu_biotypes(
    SEU=SEU,
    biomart=gtf.info,
    gene.colname = "GeneSymbol",
    biotype.colname = "Biotype",
    prefix = "pct.unspliced.",
    assay = "unspliced"
  )
)%>%suppressMessages()
```

Plot maps for unspliced gene biotype
```{r}
suppressMessages(
  visListPlot(
    skm.list,
    sample.titles = stringr::str_remove_all(meta_skm$sample,pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="_SkM"),
    reduction="space",
    assay="kallisto_unspliced", 
    slot = 'data',
    pt.size=0.4,
    legend.position = "bottom",
    font.size = 6,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    # colormap = "plasma",
    colormap = mckolors$Spectral%>%rev(),
    features = c(
      "pct.unspliced.protein_coding", "pct.unspliced.lncRNA", "pct.snRNA"
    )
  )&theme(
    legend.text = element_text(size=6)
  )&coord_fixed(
    ratio=1/1.6
  )
)
```

### Visualize gene splicing ratios with visCoMap
```{r fig.height=8, fig.width=6}
tmp.feat = "Myog"
visCoMap(
  skm.list,
  sample.titles = meta_skm$sample,
  reduction="space",
  slot = c('counts','counts'),
  pt.size=0.6,
  legend.position = "bottom",
  font.size = 8,
  axis.title.angle.y=0,
  nrow = 1,
  combine = T,
  verbose=T,
  # colormap = mckolors$Spectral,
  # colormap.direction = 1,
  comap.fxn = function(X) X[1]/X[2],
  coex.name = "Unspliced/Total", alt.titles = paste(tmp.feat,c("Unspliced", "Total")),
  features = c(tmp.feat, tmp.feat),assay=c("unspliced","kallisto_collapsed"),
)&coord_fixed(ratio=1/1.6)
```
# Spatially variable features
Find spatially variable features
```{r}
library(parallel)
num.cores=9
cl <- makeCluster(num.cores)

skm.list <- mclapply(
  skm.list,
  FUN = function(SEU) FindSpatiallyVariableFeatures(
      SEU,
      assay = "kallisto_collapsed", 
      # features = VariableFeatures(SEU)[1:1000],
      selection.method = "markvariogram"
  ),
  mc.cores=num.cores
)

```

Pull out top features
```{r}
top.spatial <- lapply(
  skm.list,
  FUN=function(VIS) SpatiallyVariableFeatures(
    VIS, selection.method = "markvariogram"
  ) %>%
    head(n=50)
) 

tmp.spatial <- top.spatial %>%
  unlist() %>%
  unique()
# tmp.spatial = tmp.spatial[tmp.spati]
tmp.spatial
```

Plot some of the top features
```{r fig.height=6, fig.width=16, message=FALSE, warning=FALSE}
suppressMessages(
  visListPlot(
    skm.list,
    sample.titles = stringr::str_remove_all(meta_skm$sample, pattern = "Vis_") %>%
      stringr::str_remove_all(pattern ="_SkM"),
    reduction="space",
    slot = 'data',
    pt.size=0.25,
    legend.position = "bottom",
    font.size = 8,
    axis.title.angle.y=0,
    colormap = mckolors$cubs,
    nrow = 1,
    combine = T,
    verbose=F,
    features = c("Zc3h7a","ENSMUSG00002076173"),assay=c("unspliced","kallisto"),
    
    # features = tmp.spatial[tmp.spatial %in% genes.ncrna] %>%rev()%>% head(n=12)
    # features=grepGenes(skm.list[[7]],assay="kallisto_collapsed",pattern = "Mir",filter.pattern = "hg")[1:12]
    # features = grepGenes(skm.list[[6]],assay="kallisto_collapsed",pattern = genes$miRNA,filter.pattern = "ENS")[1:12]
    # features=grepGenes(skm.list[[6]],assay="kallisto_collapsed",pattern = c("Sno","SNO"))[1:12]
  )&theme(
    legend.text = element_text(size=6)
  )
)
```

# Save processed muscle samples
```{r}
save(
  skm.list,
  file = "/workdir/dwm269/totalRNA/spTotal/robjs/skm_list_v2.RData"
)
```

# Session Info
```{r}
sessionInfo()
```