# 10x Visium + yPAP QC, preprocessing, & preliminary analyses
## David W. McKellar

# Session setup
## Libs, setwd
```{r message=FALSE, warning=FALSE}
# analysis
library(Matrix, quietly = T)
library(dplyr, quietly = T)
library(Seurat, quietly = T)
library(future, quietly = T)
# library(cluster)
library(parallel, quietly = T)
library(data.table, quietly = T)

# plotting
library(ggplot2, quietly = T)
library(patchwork, quietly = T)
library(pals, quietly = T)
library(viridis, quietly = T)
library(shades, quietly = T)

# DWMutils
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seutils.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seuplots.R")
```

## Figure settings & colors
```{r}
# fonts, sizes, etc.
small.font = 6*2
big.font = 8*2
line.width = 0.5
pt.size=0.01
pt.stroke=0.3
label.size=2
```

## Plot themes 
```{r}
source("/home/dwm269/DWM_utils/plotting_utils/scThemes.R")
scTheme <- scThemes(
  small.font = small.font,
  big.font = big.font,
  line.width = line.width,
  pt.size=pt.size,
  pt.stroke=pt.stroke,
  label.size=label.size
)

print(names(scTheme))
```
```{r}
mckolors <- read.csv("/home/dwm269/DWM_utils/plotting_utils/McKolors_v1.csv") %>% 
  as.list() %>%
  lapply(
    # FUN=function(X) X[X!=""]
  )
mckolors
```
## Read in metadata, `meta_vis`
```{r}
meta_vis <- read.csv(
  file="/workdir/dwm269/totalRNA/spTotal/resources/meta_sheet.csv"
)

print(meta_vis)
```

## Parallelization settings ####
```{r}
plan("sequential")

# plan("multiprocess", workers = 6)
# options(future.globals.maxSize = 2000 * 1024^2)
```

# Alignment QC & comparison
## Qualimap data
```{r}
qualimap.list <- list()

for(i in 1:nrow(meta_vis)){
  qualimap.list[[i]] <- fread(
    paste0(meta_vis$data.dir.STARsolo[i],"/qualimap_out/raw_data_qualimapReport/coverage_profile_along_genes_(total).txt")
  )
  colnames(qualimap.list[[i]]) <- c("position", "coverage")
  qualimap.list[[i]]$sample <- rep(meta_vis$sample[i],nrow(qualimap.list[[i]]))
  # qualimap.list[[i]]$type <- rep(meta_vis$type[i],nrow(qualimap.list[[i]]))
  qualimap.list[[i]]$polyA <- rep(meta_vis$polyA[i],nrow(qualimap.list[[i]]))
  qualimap.list[[i]]$rnase_inhib <- rep(meta_vis$rnase_inhib[i],nrow(qualimap.list[[i]]))
}
qualimap.df <- do.call(rbind,qualimap.list)

print(colnames(qualimap.df))
```

Line plot showing coverage across the (normalized) body of all genes; colored by the sample pre-processing (contrl/no preprocesssing, yPAP+SUPERase, or yPAP+Protector)
```{r message=FALSE}
ggplot(
  qualimap.df,
  aes(
    x=position,
    y=coverage,
    group=sample,
    # linetype=type,
    color=rnase_inhib #polyA
  )
) + 
  geom_line(size=1)+
  ggsci::scale_color_npg(alpha=0.8)+
  theme_minimal()+
  theme(
    axis.line = element_line(color="black",size = line.width),
    axis.ticks = element_line(color="black",size = line.width),
    axis.text = element_text(color="black")
  )+
  labs(
    x="Normalized Gene Position",
    y="Coverage",
    color=""
  )+
  scale_y_continuous(labels = scales::scientific)

```

## STAR alignment summary
```{r}
sum.list <- list()

for(i in 1:nrow(meta_vis)){
  tmp <- fread(
    paste0(meta_vis$data.dir.STARsolo[i],"/Solo.out/Gene/Summary.csv")
  ) %>% as.matrix()
  rownames(tmp) <- tmp[,1]
  sum.list[[i]] <- t(tmp[,2])%>%as.data.frame()
  
  sum.list[[i]]$sample <- rep(meta_vis$sample[i],nrow(sum.list[[i]]))
  sum.list[[i]]$polyA <- rep(meta_vis$polyA[i],nrow(sum.list[[i]]))
  sum.list[[i]]$rnase_inhib <- rep(meta_vis$rnase_inhib[i],nrow(sum.list[[i]]))
  
  colnames(sum.list[[i]]) <- stringr::str_replace_all(colnames(sum.list[[i]]),pattern=" ", replacement="_")
  colnames(sum.list[[i]]) <- stringr::str_replace_all(colnames(sum.list[[i]]),pattern=":", replacement="")
  # colnames(sum.list[[i]]) <- stringr::str_replace_all(colnames(sum.list[[i]]),pattern="+", replacement="_")
}
sum.df <- do.call(rbind,sum.list)

colnames(sum.df) <- stringr::str_replace_all(colnames(sum.df),pattern="\\+",replacement="_")

print(colnames(sum.df))
```
```{r message=FALSE}
ggplot(
  sum.df,
  aes(
    x=as.numeric(Reads_Mapped_to_Genome_Unique_Multiple),
    y=as.numeric(Reads_Mapped_to_Gene_Unique_Multipe_Gene),
    size=as.numeric(Q30_Bases_in_RNA_read),
    group=sample,
    # linetype=type,
    color=rnase_inhib #polyA
  )
) + 
  geom_point()+
  ggsci::scale_color_npg(alpha=0.8)+
  theme_minimal()+
  theme(
    axis.line = element_line(color="black",size = line.width),
    axis.ticks = element_line(color="black",size = line.width),
    axis.text = element_text(color="black")
  )+
  labs(
    x="Reads mapped to genome",
    y="Reads mapped to annotated genes",
    size="Q30 bases\nin RNA reads",
    color=""
  )


#   scale_y_continuous(labels = scales::scientific)

```

# Load and pre-process each of the datasets
## Read in matrices and initialize seurat objects - spaceranger
```{r}
vis.list <- list()
for(i in 1:length(meta_vis$data.dir.spaceranger)){ 
  if(file.exists(paste0(meta_vis$data.dir.spaceranger[i], '/outs/filtered_feature_bc_matrix'))){ 
    cat("Reading #",i, ": ", meta_vis$data.dir.spaceranger[i], '...\n')
    vis.list[[i]] <- Seurat::Load10X_Spatial(
      data.dir = paste0(meta_vis$data.dir.spaceranger[i], '/outs'),
      filter.matrix = T
    )
    
    # Strip "-1" suffix from cell barcode (colnames)
    if(stringr::str_detect(Cells(vis.list[[i]]),pattern = "-1")){
      cat("     Stripping suffix from spot barcodes...\n")
      vis.list[[i]] <- RenameCells(
        object = vis.list[[i]],
        new.names = stringr::str_remove_all(Cells(vis.list[[i]]),pattern="-1")
      )
    }
    
    cat("    Done!\n")
  }else{
    cat("Data not found for # ", i, " (", meta_vis$data.dir.spaceranger[i], ")", "\n")
  }
}
```

## Read in matrices - STARsolo 
```{r}
mat.list <- list()
for(i in 1:length(meta_vis$data.dir.STARsolo)){ 
  if(file.exists(paste0(meta_vis$data.dir.STARsolo[i], "/Solo.out/Gene/raw/"))){ 
    cat("Reading #",i, ": ", meta_vis$data.dir.STARsolo[i], ' \n')
    mat.list[[i]] <- Read10X(
      data.dir = paste0(meta_vis$data.dir.STARsolo[i], "/Solo.out/Gene/raw/")
    )
  }else{
    cat("Data not found for # ", i, " (", meta_vis$data.dir.STARsolo[i], ")", "\n")
  }
}

cat(sum(unlist(lapply(mat.list, ncol))),"spots (total) loaded...\n")
```

# Add STARsolo counts as new assay
```{r warning=FALSE}
for(i in 1:length(vis.list)){
  cat(paste0("Adding STARsolo for ", meta_vis$sample[i]),"\n")
  tmp <- CreateAssayObject(
    counts=mat.list[[i]],
    min.cells = 3
  )
  print(dim(tmp))
  
  vis.list[[i]][["STARsolo"]] <- subset(
    tmp,
    cells = Cells(vis.list[[i]])
  )
  
  vis.list[[i]]@active.assay <- "STARsolo"
}

```

## Initialize seurat objects w/ STARsolo counts
```{r}
# mt.genes <- read.csv("/workdir/dwm269/totalRNA/10xTotal/resources/GENCODEm27_MTgenes.csv")$gene.name

for(i in 1:length(vis.list)){
  cat(' #####################################\n',
      '### Processing dataset number ', i, '###\n',
      '#####################################\n')
  
  # Add meta_vis data
  for(md in colnames(meta_vis)){
    vis.list[[i]][[md]] <- meta_vis[[md]][i]
  }
  
  # add %MT
  vis.list[[i]][["percent.mt"]]  <- PercentageFeatureSet(
    vis.list[[i]],
    # features = mt.genes
    pattern = "mt-"
  ) 
  
    # hemaglobin scores
  vis.list[[i]][["percent.Hb"]]  <- PercentageFeatureSet(
    vis.list[[i]], 
    features = c(
      rownames(vis.list[[i]])[grep(rownames(vis.list[[i]]),pattern="Hba")],
      rownames(vis.list[[i]])[grep(rownames(vis.list[[i]]),pattern="Hbb")]
    )
  )
  
  # ribosomal protein scores
  vis.list[[i]][["percent.Rp"]]  <- PercentageFeatureSet(
    vis.list[[i]], 
    features = c(
      rownames(vis.list[[i]])[grep(rownames(vis.list[[i]]),pattern="Rps")],
      rownames(vis.list[[i]])[grep(rownames(vis.list[[i]]),pattern="Rpl")]
    )
  )
  
  # Filter out low quality cells according to the metrics defined above
  
  vis.list[[i]] <- subset(
    vis.list[[i]],
    subset = #percent.mt < 40 & 
      nCount_STARsolo > 1000 
      # nFeature_RNA < 4000
  )
  
  # Only mito and floor filtering; trying to find doublets
}
# cat((sum(unlist(lapply(mat.list, ncol)))-sum(unlist(lapply(vis.list, ncol)))),"cells (total) removed...\n")
```

### Collapse multimappers, preprocess collapsed counts
```{r warning=FALSE}
vis.list <- lapply(
  vis.list,
  FUN = function(SEU) collapseMultimappers(
    SEU,
    assay="STARsolo",
    new.assay.name = "STARsolo_collapsed",
    verbose=T
    )
)

vis.list <- lapply(
  vis.list,
  FUN = function(SEU) seuPreProcess(SEU, assay="STARsolo_collapsed",verbose=F)
)

```

## Add kallisto counts
```{r}
library(SeuratDisk)
# convert h5ad to h5seurat
for(i in 1:nrow(meta_vis)){
  tmp.h5 <- paste0(meta_vis$data.dir.kallisto[i],"/kb_out/counts_unfiltered/adata")
  if(!file.exists(paste0(tmp.h5,".h5seurat"))){
    message(paste0("Converting h5seurat for",meta_vis$capture.area[i],"...\n"))
    Convert(
      paste0(tmp.h5,".h5ad"),
      dest = "h5seurat",
      overwrite = TRUE
    )
  }
  
  message(paste0("Loading & cleaning kallisto output for ",meta_vis$sample[i],"...\n"))
  
  # Load h5seurat and subset out counts for sample
  tmp.seu <- LoadH5Seurat(
    paste0(tmp.h5,".h5seurat"),
  ) %>% subset( # select cells
    cells=Cells(vis.list[[i]])
  )
  tmp.seu <- subset( # Select features
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )
  
  # collapse multimappers
  tmp.seu <- collapseMultimappers(
    tmp.seu,
    assay="RNA",
    new.assay.name = "RNA",
    verbose=F
  )
  
  # convert ensembl IDs to MGI gene IDs
  tmp.mat <- GetAssayData(tmp.seu,assay="RNA")
  rownames(tmp.mat) <- ens2gene(
    ens=rownames(tmp.mat),
    biomart.info = mouse.info,
    ens.colname = "ensembl_gene_id",
    gene.colname = "mgi_symbol",
    force.unique = T,
    ncores=24,
    verbose=F
  )
  
  # add kallisto counts to vis_list
  vis.list[[i]][["kallisto"]] <- CreateAssayObject(
    counts = tmp.mat,
    min.cells = 1
  )
  
  rm(tmp.seu)
  rm(tmp.mat)
}
gc()

# read mtx - kallisto doesn't give filtered mtx, so mismatch in feature counts... can't use
# for(i in 1:nrow(meta_vis)){
#   tmp.dir <- paste0(meta_vis$data.dir.kallisto[i],"/kb_out/counts_unfiltered/")
#   message(paste0("Loading mtx for ",meta_vis$sample[i]))
#   tmp <- Seurat::ReadMtx(
#     mtx = paste0(tmp.dir,"cells_x_genes.mtx"),
#     cells= paste0(tmp.dir,"cells_x_genes.barcodes.txt"),
#     features = paste0(tmp.dir,"cells_x_genes.genes.txt"),
#     feature.column=1
#   )
# }
```
### Preprocess collapsed counts (kallisto)
```{r warning=FALSE}
# collapse multimappers
vis.list <- lapply(
  vis.list,
  FUN = function(SEU) collapseMultimappers(
    SEU,
    assay="kallisto",
    new.assay.name = "kallisto_collapsed",
    verbose=T
    )
)

# preprocess
vis.list <- lapply(
  vis.list,
  FUN = function(SEU) seuPreProcess(
    SEU, 
    assay="kallisto_collapsed",
    verbose=F
  )
)
```

# Add spatial locations as a `reduction` (for easy plotting with DimPlot)
```{r}
meta_vis$h_flip<- c(
  -1,-1,1, -1,-1,-1,1,1,1, -1,1,-1,1,  -1,-1,1,1
)
meta_vis$v_flip<- c(
  -1, -1, -1, 1, -1, 1, -1, 1, -1,  1,-1,1,-1,  1,-1,1,1
)


vis.list <- mapply(
  FUN = function(SEU, h_flip, v_flip){
    
    tmp <- as.matrix(cbind(
      SEU@images$slice1@coordinates$row * h_flip,
      SEU@images$slice1@coordinates$col * v_flip
    ))
      
    colnames(tmp) <- paste0("space_", 1:2)
    rownames(tmp) <- colnames(SEU)
    
    SEU[["space"]] <- CreateDimReducObject(
      embeddings=as.matrix(tmp),
      assay="STARsolo_collapsed",
      key = "space_"
    )
    
    return(SEU)
  },
  vis.list,
  meta_vis$h_flip,
  meta_vis$v_flip
)
```

# Compute gene biotype info
## biomaRt
```{r}
if(!exists("mouse.info")){
  mouse = biomaRt::useMart(
    # biomart = "ENSEMBL_MART_MOUSE",
   biomart = "ENSEMBL_MART_ENSEMBL",
    dataset = "mmusculus_gene_ensembl"
    # host = "https://apr2020.archive.ensembl.org" ##for mm10 info
  )
  mouse.info <- biomaRt::getBM(
    attributes = c(
      "mgi_symbol", "gene_biotype","transcript_biotype", 
      # "description", "name_1006","definition_1006","external_gene_name",
      "ensembl_gene_id","ensembl_transcript_id","ensembl_peptide_id",
      "chromosome_name","transcript_length", "strand"
    ),
    uniqueRows = T,
    # host = "useast.ensembl.org",
    verbose = T,
    mart = mouse
  )
}

cat(
  paste("Using", length(unique(mouse.info$mgi_symbol)), "genes, and ",length(unique(mouse.info$gene_biotype)),"biotypes...\n")
)
```

### gene set scores 
#### STARsolo
biomaRt gene biotype scores 
```{r paged.print=TRUE}
vis.list <- lapply(
  vis.list,
  FUN = function(SEU) seu_biotypes(
    SEU=SEU,
    biomart=mouse.info,
    gene.colname = "mgi_symbol",
    biotype.colname = "gene_biotype",
    assay = "STARsolo_collapsed"
  )
)
```

```{r}
rRNA.genes <- lapply(
  vis.list,
  FUN = function(VIS) grepGenes(
    VIS,
    assay = "STARsolo_collapsed",
    pattern = c(
      "^Rn[0-9]{1,2}s",
      "n-R5",
      "ENSMUSG00000118642"
      )
  )
) %>% 
  unlist() %>%
  unique() %>%
  sort()
```

```{r}
vis.list <- lapply(
  vis.list,
  FUN = function(VIS) PercentageFeatureSet(
    VIS,
    assay = "STARsolo_collapsed",
    features = rRNA.genes[rRNA.genes %in% Features(VIS,assay = "STARsolo_collapsed")],
    col.name = "pct.rRNA"
  )
)
```


```{r}
visListPlot(
  vis.list,
  features = "pct.rRNA",
  axis.title.angle.y = 0
)
```

#### Kallisto
biomaRt gene biotype scores 
```{r paged.print=TRUE}
vis.list <- lapply(
  vis.list,
  FUN = function(SEU) seu_biotypes(
    SEU=SEU,
    biomart=mouse.info,
    gene.colname = "mgi_symbol",
    biotype.colname = "gene_biotype",
    assay = "kallisto_collapsed"
  )
)%>%suppressMessages()
```

# Save visium list(s)
```{r}
# All samples
save(
  vis.list,
  file="/workdir/dwm269/totalRNA/spTotal/robjs/vis_list_v2.RData"
)
```

## Plot gene biotype info
### merged - scatter plot
```{r}
if(!exists("vis.merged")){
  vis.merged <- merge(
    vis.list[[1]],
    vis.list[2:length(vis.list)],
    add.cell.ids = meta_vis$sample
  )
}

tmp.feat <- paste0("pct.",unique(mouse.info$gene_biotype))
tmp.feat <- tmp.feat[tmp.feat %in% colnames(vis.merged@meta.data)]

tmp.df <- vis.merged@meta.data[,c("sample","polyA","rnase_inhib",tmp.feat)]

tmp.df <- reshape2::melt(tmp.df,id.vars=c("sample","polyA","rnase_inhib"))
colnames(tmp.df) <- c("sample","polyA","rnase_inhib", "biotype", "percent.counts")

# head(tmp.df)

ggplot(
  tmp.df,
  aes(
    x=polyA,
    y=percent.counts,
    fill=biotype
  )
)+
  geom_col(
    position="fill"
  )+
  scTheme$bar+
  theme(
    legend.position="right"
  )+
  # coord_polar()+
  ggsci::scale_fill_futurama()

```
### Gene biotype spatial plots
```{r Biotype spatial plots, fig.height=10, fig.width=17, message=FALSE, warning=FALSE, paged.print=TRUE}
i = c(
  1:3, # SkM ctrl
  # 4,6,7,5,8,9, # SkM SUPERase
  12,13,10,11 # # SkM Protector
  )
visListPlot(
  vis.list[i],
  sample.titles = stringr::str_remove_all(meta_vis$sample[i],pattern = "Vis_") %>%
        stringr::str_remove_all(pattern ="yPAP_")%>%
        stringr::str_remove_all(pattern ="ctrl_")%>%
        stringr::str_remove_all(pattern ="_Heart")%>%
        stringr::str_remove_all(pattern ="_SkM"),
  # assay="STARsolo_collapsed",
  reduction="space",
  # slot = 'counts',
  pt.size=0.6,
  legend.position = "bottom",
  font.size = 12,
  axis.title.angle.y=0,
  nrow = 1,
  combine = T,
  verbose=F,
  # colormap=c("deepskyblue","firebrick"),
  colormap = mckolors$cubs,
  features = c(
    # "nFeature_STARsolo","nCount_STARsolo","percent.mt",
    "pct.protein_coding",
    # "pct.lncRNA",
    "pct.rRNA",
    # "pct.Mt_rRNA", 
    "pct.Mt_tRNA",
    # "pct.ribo.pp",
    "pct.snoRNA",
    "pct.miRNA",
    "pct.snRNA"
  ) 
)
```

```{r}
i = c(
  1:3, # SkM ctrl
  4,6,7,5,8,9, # SkM SUPERase
  12,13,10,11 # # SkM Protector
  )
visListPlot(
    vis.list[i],
    sample.titles = stringr::str_remove_all(meta_vis$sample[i],pattern = "Vis_") %>%
        stringr::str_remove_all(pattern ="yPAP_")%>%
        stringr::str_remove_all(pattern ="ctrl_")%>%
        stringr::str_remove_all(pattern ="_Heart"),
    reduction="space",
    # slot = 'counts',
    pt.size=1.2,
    legend.position = "bottom",
    font.size = 12,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    colormap = mckolors$cubs,
    features = "pct.miRNA"
)
```
# QC analysis
```{r}
vis.merged@meta.data[order(vis.merged$nCount_kallisto_collapsed),] %>% ggplot(
  aes(
    x=nCount_kallisto_collapsed,
    fill=rnase_inhib
  )
)+
  geom_histogram(
    # geom="point"
  )+
  facet_wrap(
    facets = c("tissue","rnase_inhib")
  )
  # scTheme$bar

```



# Plotting
## 
```{r fig.height=6, fig.width=10, message=FALSE, warning=FALSE, paged.print=TRUE}

VlnPlot(
  vis.merged,
  assay="STARsolo_collapsed",
  # slot = 'counts',
  group.by="sample",
  fill.by = 'type',
  pt.size=0,
  combine = F,
  features = c(
    "nFeature_STARsolo","nCount_STARsolo","percent.mt",
    "pct.protein_coding","percent.ribo.RNA","percent.ribo.pp","pct.snoRNA","pct.miRNA"
  )
) %>% lapply(
  FUN=function(SEU) SEU + 
    scTheme$vln + 
    NoLegend() +
    theme(
      plot.title = element_text(hjust=0.5),
      axis.line = element_line(color="black"),
      axis.ticks = element_line(color="black"),
      axis.text.x = element_blank(),
      axis.title.x = element_blank()
    )
)%>%
  wrap_plots(nrow=2)
```


# Session Info
```{r}
sessionInfo()
```

