# 10x Visium + yPAP QC, preprocessing, & preliminary analyses
## David W. McKellar

# Session setup
## Libs, setwd
```{r message=FALSE, warning=FALSE}
# setwd("/workdir/dwm269/totalRNA/spTotal/")

# analysis
library(Matrix, quietly = T)
library(dplyr, quietly = T)
library(Seurat, quietly = T)
library(future, quietly = T)
# library(cluster)
library(parallel, quietly = T)
library(data.table, quietly = T)

# plotting
library(ggplot2, quietly = T)
library(patchwork, quietly = T)
library(pals, quietly = T)
library(viridis, quietly = T)
library(shades, quietly = T)

# DWMutils
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seutils.R")
source("/home/dwm269/DWM_utils/sc_utils/seurat_helpers/seuplots.R")
```

## Figure settings & colors
```{r}
# fonts, sizes, etc.
small.font = 6*2
big.font = 8*2
line.width = 0.5
pt.size=0.01
pt.stroke=0.3
label.size=2
```

## Plot themes 
```{r}
source("/home/dwm269/DWM_utils/plotting_utils/scThemes.R")
scTheme <- scThemes(
  small.font = small.font,
  big.font = big.font,
  line.width = line.width,
  pt.size=pt.size,
  pt.stroke=pt.stroke,
  label.size=label.size
)

print(names(scTheme))
```
```{r}
mckolors <- read.csv("/home/dwm269/DWM_utils/plotting_utils/McKolors_v1.csv") %>% 
  as.list() %>%
  lapply(
    FUN=function(X) X[X!=""]
  )
mckolors
```
## Read in metadata, `meta_vis`
```{r}
meta_vis <- read.csv(
  file="/workdir/dwm269/totalRNA/spTotal/resources/meta_sheet_visium.csv"
)

print(meta_vis)
```

## Parallelization settings ####
```{r}
plan("sequential")

# plan("multiprocess", workers = 6)
# options(future.globals.maxSize = 2000 * 1024^2)
```

## Get gene biotype info from the GENCODE annotations
`gtf.info` is just a reformatted .tsv of the GENCODE M28 annotations .gtf
```{r}
gtf.info <- read.csv(
  file="/workdir/dwm269/totalRNA/spTotal/resources/gene_lists/GRCm39_GENCODEm28_gene_info_gtf.tsv",
  sep = "\t"
)

# Remove period/decimal suffixes from ensembl IDs
gtf.info$GeneID <- strsplit(
  x=gtf.info$GENEID,
  split = "\\."
)%>%
  lapply(
    FUN=function(X) X[[1]]
  )%>%
  unlist()

cat(
  paste("Using", length(unique(gtf.info$GeneSymbol)), "genes, and ",length(unique(gtf.info$Biotype)),"biotypes...\n")
)

head(gtf.info)
```


# Alignment QC & comparison
## Qualimap data
```{r}
qualimap.list <- list()

for(i in 1:nrow(meta_vis)){
  qualimap.list[[i]] <- fread(
    paste0(meta_vis$data.dir.STARsolo[i],"/qualimap_out/raw_data_qualimapReport/coverage_profile_along_genes_(total).txt")
  )
  colnames(qualimap.list[[i]]) <- c("position", "coverage")
  qualimap.list[[i]]$sample <- rep(meta_vis$sample[i],nrow(qualimap.list[[i]]))
  # qualimap.list[[i]]$type <- rep(meta_vis$type[i],nrow(qualimap.list[[i]]))
  qualimap.list[[i]]$polyA <- rep(meta_vis$polyA[i],nrow(qualimap.list[[i]]))
  qualimap.list[[i]]$rnase_inhib <- rep(meta_vis$rnase_inhib[i],nrow(qualimap.list[[i]]))
}
qualimap.df <- do.call(rbind,qualimap.list)

print(colnames(qualimap.df))
```

Line plot showing coverage across the (normalized) body of all genes; colored by the sample pre-processing (contrl/no preprocesssing, yPAP+SUPERase, or yPAP+Protector)
```{r message=FALSE}
ggplot(
  qualimap.df,
  aes(
    x=position,
    y=coverage,
    group=sample,
    # linetype=type,
    color=rnase_inhib #polyA
  )
) + 
  geom_line(size=1)+
  ggsci::scale_color_npg(alpha=0.8)+
  theme_minimal()+
  theme(
    axis.line = element_line(color="black",size = line.width),
    axis.ticks = element_line(color="black",size = line.width),
    axis.text = element_text(color="black")
  )+
  labs(
    x="Normalized Gene Position",
    y="Coverage",
    color=""
  )+
  scale_y_continuous(labels = scales::scientific)

```

## STAR alignment summary
```{r}
sum.list <- list()

for(i in 1:nrow(meta_vis)){
  tmp <- fread(
    paste0(meta_vis$data.dir.STARsolo[i],"/Solo.out/Gene/Summary.csv")
  ) %>% as.matrix()
  rownames(tmp) <- tmp[,1]
  sum.list[[i]] <- t(tmp[,2])%>%as.data.frame()
  
  sum.list[[i]]$sample <- rep(meta_vis$sample[i],nrow(sum.list[[i]]))
  sum.list[[i]]$polyA <- rep(meta_vis$polyA[i],nrow(sum.list[[i]]))
  sum.list[[i]]$rnase_inhib <- rep(meta_vis$rnase_inhib[i],nrow(sum.list[[i]]))
  
  colnames(sum.list[[i]]) <- stringr::str_replace_all(colnames(sum.list[[i]]),pattern=" ", replacement="_")
  colnames(sum.list[[i]]) <- stringr::str_replace_all(colnames(sum.list[[i]]),pattern=":", replacement="")
  # colnames(sum.list[[i]]) <- stringr::str_replace_all(colnames(sum.list[[i]]),pattern="+", replacement="_")
}
sum.df <- do.call(rbind,sum.list)

colnames(sum.df) <- stringr::str_replace_all(colnames(sum.df),pattern="\\+",replacement="_")

print(colnames(sum.df))
```
```{r message=FALSE}
ggplot(
  sum.df,
  aes(
    x=as.numeric(Reads_Mapped_to_Genome_Unique_Multiple),
    y=as.numeric(Reads_Mapped_to_Gene_Unique_Multipe_Gene),
    size=as.numeric(Q30_Bases_in_RNA_read),
    group=sample,
    # linetype=type,
    color=rnase_inhib #polyA
  )
) + 
  geom_point()+
  ggsci::scale_color_npg(alpha=0.8)+
  theme_minimal()+
  theme(
    axis.line = element_line(color="black",size = line.width),
    axis.ticks = element_line(color="black",size = line.width),
    axis.text = element_text(color="black")
  )+
  labs(
    x="Reads mapped to genome",
    y="Reads mapped to annotated genes",
    size="Q30 bases\nin RNA reads",
    color=""
  )


#   scale_y_continuous(labels = scales::scientific)

```

# Load and pre-process each of the datasets
## Read in matrices and initialize seurat objects - spaceranger
```{r}
vis.list <- list()
for(i in 1:length(meta_vis$data.dir.spaceranger)){ 
  if(file.exists(paste0(meta_vis$data.dir.spaceranger[i], '/outs/filtered_feature_bc_matrix'))){ 
    cat("Reading #",i, ": ", meta_vis$data.dir.spaceranger[i], '...\n')
    vis.list[[i]] <- Seurat::Load10X_Spatial(
      data.dir = paste0(meta_vis$data.dir.spaceranger[i], '/outs'),
      filter.matrix = T
    )
    
    # Strip "-1" suffix from cell barcode (colnames)
    if(stringr::str_detect(Cells(vis.list[[i]]),pattern = "-1")){
      cat("     Stripping suffix from spot barcodes...\n")
      vis.list[[i]] <- RenameCells(
        object = vis.list[[i]],
        new.names = stringr::str_remove_all(Cells(vis.list[[i]]),pattern="-1")
      )
    }
    
    cat("    Done!\n")
  }else{
    cat("Data not found for # ", i, " (", meta_vis$data.dir.spaceranger[i], ")", "\n")
  }
}
```

## Read in matrices - STARsolo 
```{r}
mat.list <- list()
for(i in 1:length(meta_vis$data.dir.STARsolo)){ 
  if(file.exists(paste0(meta_vis$data.dir.STARsolo[i], "/Solo.out/Gene/raw/"))){ 
    cat("Reading #",i, ": ", meta_vis$data.dir.STARsolo[i], ' \n')
    mat.list[[i]] <- Read10X(
      data.dir = paste0(meta_vis$data.dir.STARsolo[i], "/Solo.out/Gene/raw/")
    )
  }else{
    cat("Data not found for # ", i, " (", meta_vis$data.dir.STARsolo[i], ")", "\n")
  }
}

cat(sum(unlist(lapply(mat.list, ncol))),"spots (total) loaded...\n")
```

# Add STARsolo counts as new assay
```{r warning=FALSE}
for(i in 1:length(vis.list)){
  cat(paste0("Adding STARsolo for ", meta_vis$sample[i]),"\n")
  tmp <- CreateAssayObject(
    counts=mat.list[[i]],
    min.cells = 3
  )
  print(dim(tmp))
  
  vis.list[[i]][["STARsolo"]] <- subset(
    tmp,
    cells = Cells(vis.list[[i]])
  )
  
  vis.list[[i]]@active.assay <- "STARsolo"
}

```

## Initialize seurat objects w/ STARsolo counts
```{r}
# mt.genes <- read.csv("/workdir/dwm269/totalRNA/10xTotal/resources/GENCODEm27_MTgenes.csv")$gene.name

for(i in 1:length(vis.list)){
  cat(' #####################################\n',
      '### Processing dataset number ', i, '###\n',
      '#####################################\n')
  
  # Add meta_vis data
  for(md in colnames(meta_vis)){
    vis.list[[i]][[md]] <- meta_vis[[md]][i]
  }
  
  # add %MT
  vis.list[[i]][["percent.mt"]]  <- PercentageFeatureSet(
    vis.list[[i]],
    # features = mt.genes
    pattern = "mt-"
  ) 
  
    # hemaglobin scores
  vis.list[[i]][["percent.Hb"]]  <- PercentageFeatureSet(
    vis.list[[i]], 
    features = c(
      rownames(vis.list[[i]])[grep(rownames(vis.list[[i]]),pattern="Hba")],
      rownames(vis.list[[i]])[grep(rownames(vis.list[[i]]),pattern="Hbb")]
    )
  )
  
  # ribosomal protein scores
  vis.list[[i]][["percent.Rp"]]  <- PercentageFeatureSet(
    vis.list[[i]], 
    features = c(
      rownames(vis.list[[i]])[grep(rownames(vis.list[[i]]),pattern="Rps")],
      rownames(vis.list[[i]])[grep(rownames(vis.list[[i]]),pattern="Rpl")]
    )
  )
  
  # Filter out low quality cells according to the metrics defined above
  
  vis.list[[i]] <- subset(
    vis.list[[i]],
    subset = #percent.mt < 40 & 
      nCount_STARsolo > 1000 
      # nFeature_RNA < 4000
  )
  
  # Only mito and floor filtering; trying to find doublets
}
# cat((sum(unlist(lapply(mat.list, ncol)))-sum(unlist(lapply(vis.list, ncol)))),"cells (total) removed...\n")
```
## Read in spliced/unspliced matrices and look at total counts for each
```{r}
require("R.utils")
for(i in 1:length(vis.list)){
  # Load in spot barcodes, features, and all three matrices
  cbs <- fread(
    paste0(meta_vis$data.dir.STARsolo[i],"/Solo.out/Velocyto/raw/barcodes.tsv.gz"),
    header=F
  )$V1
  
  # Spliced
  spliced.mat <- Matrix::readMM(
    file = paste0(meta_vis$data.dir.STARsolo[i],"/Solo.out/Velocyto/raw/spliced.mtx.gz")
  )
  colnames(spliced.mat) <- cbs
  vis.list[[i]]$nCount_STARsolo_spliced <- spliced.mat[,Cells(vis.list[[i]])] %>% colSums()
  vis.list[[i]]$pct_STARsolo_spliced <- vis.list[[i]]$nCount_STARsolo_spliced/vis.list[[i]]$nCount_STARsolo*100
  
  # Unspliced
  unspliced.mat <- Matrix::readMM(
    file = paste0(meta_vis$data.dir.STARsolo[i],"/Solo.out/Velocyto/raw/unspliced.mtx.gz")
  )
  colnames(unspliced.mat) <- cbs
  vis.list[[i]]$nCount_STARsolo_unspliced <- unspliced.mat[,Cells(vis.list[[i]])] %>% colSums()
  vis.list[[i]]$pct_STARsolo_unspliced <- vis.list[[i]]$nCount_STARsolo_unspliced/vis.list[[i]]$nCount_STARsolo*100
  
  # Ambiguous
  ambiguous.mat <- Matrix::readMM(
    file = paste0(meta_vis$data.dir.STARsolo[i],"/Solo.out/Velocyto/raw/ambiguous.mtx.gz")
  )
  colnames(ambiguous.mat) <- cbs
  vis.list[[i]]$nCount_STARsolo_ambiguous <- ambiguous.mat[,Cells(vis.list[[i]])] %>% colSums()
  vis.list[[i]]$pct_STARsolo_ambiguous <- vis.list[[i]]$nCount_STARsolo_ambiguous/vis.list[[i]]$nCount_STARsolo*100
}

rm(cbs)
rm(spliced.mat)
rm(unspliced.mat)
rm(ambiguous.mat)
```


### Collapse multimappers, preprocess collapsed counts
```{r warning=FALSE}
vis.list <- lapply(
  vis.list,
  FUN = function(SEU) collapseMultimappers(
    SEU,
    assay="STARsolo",
    new.assay.name = "STARsolo_collapsed",
    verbose=T
    )
)

vis.list <- lapply(
  vis.list,
  FUN = function(SEU) seuPreProcess(SEU, assay="STARsolo_collapsed",verbose=F)
)

```

## Add kallisto counts
```{r}
for(i in 1:nrow(meta_vis)){
  
  tmp.seu <- ReadMtx(
    mtx=paste0(meta_vis$data.dir.kallisto[i],"/kb_standard/counts_unfiltered/output.mtx"),
    cells=paste0(meta_vis$data.dir.kallisto[i],"/kb_standard/counts_unfiltered/output.barcodes.txt"),
    features=paste0(meta_vis$data.dir.kallisto[i],"/kb_standard/counts_unfiltered/output.genes.txt"),
    # mtx=paste0(meta_vis$data.dir.kallisto[i],"/kb_standard/counts_unfiltered/matrix.mtx"),
    # cells=paste0(meta_vis$data.dir.kallisto[i],"/kb_standard/counts_unfiltered/barcodes.tsv"),
    # features=paste0(meta_vis$data.dir.kallisto[i],"/kb_standard/counts_unfiltered/genes.tsv"),
    feature.column=1,
    mtx.transpose = T
  ) %>%
    CreateSeuratObject() %>%
    subset( # select cells
      cells=Cells(vis.list[[i]])
    )
  
  # Select non-zero features
  tmp.seu <- subset( 
    tmp.seu,
    features = rownames(tmp.seu)[Matrix::rowSums(tmp.seu) > 0]
  )

  # convert ensembl IDs to MGI gene IDs
  tmp.mat <- GetAssayData(tmp.seu, assay="RNA")
  rownames(tmp.mat) <- ens2gene(
    ens=rownames(tmp.mat),
    biomart.info = gtf.info,
    ens.colname = "GENEID",
    # ens.colname = "GeneID", # ensembl IDs, no periods
    gene.colname = "GeneSymbol",
    force.unique = T,
    ncores=24,
    verbose=T
  )
  
  # add kallisto counts to vis_list
  vis.list[[i]][["kallisto"]] <- CreateAssayObject(
    counts = tmp.mat,
    min.cells = 1
  )
  
  cat("\n")
  
  rm(tmp.seu)
  rm(tmp.mat)
}
gc()

# read mtx - kallisto doesn't give filtered mtx, so mismatch in feature counts... can't use
# for(i in 1:nrow(meta_vis)){
#   tmp.dir <- paste0(meta_vis$data.dir.kallisto[i],"/kb_out/counts_unfiltered/")
#   message(paste0("Loading mtx for ",meta_vis$sample[i]))
#   tmp <- Seurat::ReadMtx(
#     mtx = paste0(tmp.dir,"cells_x_genes.mtx"),
#     cells= paste0(tmp.dir,"cells_x_genes.barcodes.txt"),
#     features = paste0(tmp.dir,"cells_x_genes.genes.txt"),
#     feature.column=1
#   )
# }
```
## Read in spliced/unspliced matrices and look at total counts for each
```{r}
require("R.utils")
for(i in 1:length(vis.list)){
  
  # Spliced
  # Load in spot barcodes, features, and all three matrices
  cbs <- fread(
    paste0(meta_vis$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/spliced.barcodes.txt"),
    header=F
  )$V1
  
  spliced.mat <- Matrix::readMM(
    file = paste0(meta_vis$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/spliced.mtx")
  ) %>%t()
  colnames(spliced.mat) <- cbs
  vis.list[[i]]$nCount_kallisto_spliced <- spliced.mat[,Cells(vis.list[[i]])] %>% colSums()
  vis.list[[i]]$pct_kallisto_spliced <- vis.list[[i]]$nCount_kallisto_spliced/vis.list[[i]]$nCount_kallisto
  
  # Unspliced
  cbs <- fread(
    paste0(meta_vis$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/unspliced.barcodes.txt"),
    header=F
  )$V1
  
  unspliced.mat <- Matrix::readMM(
    file = paste0(meta_vis$data.dir.kallisto[i],"/kb_lamanno/counts_unfiltered/unspliced.mtx")
  )%>%t()
  colnames(unspliced.mat) <- cbs
  vis.list[[i]]$nCount_kallisto_unspliced <- unspliced.mat[,Cells(vis.list[[i]])] %>% colSums()
  vis.list[[i]]$pct_kallisto_unspliced <- vis.list[[i]]$nCount_kallisto_unspliced/vis.list[[i]]$nCount_kallisto
  
}

rm(cbs)
rm(spliced.mat)
rm(unspliced.mat)
# rm(ambiguous.mat)
```
### Preprocess collapsed counts (kallisto)
```{r warning=FALSE}
# collapse multimappers
vis.list <- lapply(
  vis.list,
  FUN = function(SEU) collapseMultimappers(
    SEU,
    assay="kallisto",
    new.assay.name = "kallisto_collapsed",
    verbose=T
    )
)

# preprocess
vis.list <- lapply(
  vis.list,
  FUN = function(SEU) seuPreProcess(
    SEU, 
    assay="kallisto_collapsed",
    verbose=F
  )
)
```

# Add spatial locations as a `reduction` (for easy plotting with DimPlot)
```{r}
# Image rotation (T = rotate image 90 degrees)
meta_vis$rotate <- c(
  F,F,F,
  F,F,F,F,F,F,
  F,F,F,F,
  T,T,T,T
)

# Horizontal flip (-1 = flip across y-axis)
meta_vis$h_flip<- c(
  -1,-1,1, # Standard Visium, SkM samples
  -1,-1,-1,1,1,1, # spTotal + SUPERase (SkM)
  -1,1,-1,1, # spTotal + Protector (SkM)
  1,-1,1,-1 # All heart samples
)

# Vertical flip (-1 = flip across x-axis)
meta_vis$v_flip<- c(
  -1,-1,-1,
  1,-1,1,-1,1,-1,  
  1,-1,1,-1,  
  1,1,-1,-1
)

vis.list <- mapply(
  FUN = function(SEU, rotate, h_flip, v_flip){
    
    if(rotate){
      tmp <- as.matrix(cbind(
        SEU@images$slice1@coordinates$col * h_flip,
        SEU@images$slice1@coordinates$row * v_flip
      ))
    }else{
      tmp <- as.matrix(cbind(
        SEU@images$slice1@coordinates$row * h_flip,
        SEU@images$slice1@coordinates$col * v_flip
      ))
    }
    colnames(tmp) <- paste0("space_", 1:2)
    rownames(tmp) <- colnames(SEU)
    
    SEU[["space"]] <- CreateDimReducObject(
      embeddings=as.matrix(tmp),
      assay="STARsolo_collapsed",
      key = "space_"
    )
    
    return(SEU)
  },
  vis.list,
  meta_vis$rotate,
  meta_vis$h_flip,
  meta_vis$v_flip
)
```

# Compute gene biotype gene set scores 
#### STARsolo
biomaRt gene biotype scores 
```{r paged.print=TRUE}
vis.list <- lapply(
  vis.list,
  FUN = function(SEU) seu_biotypes(
    SEU=SEU,
    biomart=gtf.info,
    gene.colname = "GeneSymbol",
    biotype.colname = "Biotype",
    prefix = "Ss.",
    assay = "STARsolo_collapsed",
    verbose = F
  )
)
```

```{r}
rRNA.genes <- lapply(
  vis.list,
  FUN = function(VIS) grepGenes(
    VIS,
    assay = "STARsolo_collapsed",
    pattern = c(
      "^Rn[0-9]{1,2}s",
      "n-R5",
      "ENSMUSG00000118642"
      )
  )
) %>% 
  unlist() %>%
  unique() %>%
  sort()
```

```{r}
vis.list <- lapply(
  vis.list,
  FUN = function(VIS) PercentageFeatureSet(
    VIS,
    assay = "STARsolo_collapsed",
    features = rRNA.genes[rRNA.genes %in% Features(VIS,assay = "STARsolo_collapsed")],
    col.name = "pct.rRNA"
  )
)
```


```{r}
visListPlot(
  vis.list,
  features = "pct.rRNA",
  axis.title.angle.y = 0
)&coord_fixed(ratio=1.6)
```

#### Kallisto
gene biotype scores 
```{r paged.print=TRUE}
vis.list <- lapply(
  vis.list,
  FUN = function(SEU) seu_biotypes(
    SEU=SEU,
    biomart=gtf.info,
    gene.colname = "GeneSymbol",
    biotype.colname = "Biotype",
    prefix = "kal.",
    assay = "kallisto_collapsed"
  )
  # FUN = function(SEU) seu_biotypes(
  #   SEU=SEU,
  #   biomart=mouse.info,
  #   gene.colname = "mgi_symbol",
  #   biotype.colname = "gene_biotype",
  #   assay = "kallisto_collapsed"
  # )
)%>%suppressMessages()
```

# Save visium list(s)
```{r}
# All samples
save(
  vis.list,
  file="/workdir/dwm269/totalRNA/spTotal/robjs/vis_list_v4.RData"
)
```
## Export cell barcode lists
```{r}
for(i in 1:nrow(meta_vis)){
  write.table(
    x = Cells(vis.list[[i]]),
    file = paste0("/workdir/dwm269/totalRNA/spTotal/resources/cb_lists/",meta_vis$sample[i],".txt"),
    quote=F,
    row.names = F,col.names = F
  )
}

```

# Plot gene biotype info
### merged - scatter plot
```{r}
if(!exists("vis.merged")){
  vis.merged <- merge(
    vis.list[[1]],
    vis.list[2:length(vis.list)],
    add.cell.ids = meta_vis$sample
  )
}

tmp.feat <- paste0("pct.",unique(gtf.info$Biotype))
tmp.feat <- tmp.feat[tmp.feat %in% colnames(vis.merged@meta.data)]
tmp.feat <- tmp.feat[!grepl(tmp.feat, pattern = "IG_")]
tmp.feat <- tmp.feat[!grepl(tmp.feat, pattern = "TR_")]

tmp.df <- vis.merged@meta.data[,c("sample","polyA","rnase_inhib",tmp.feat)]

tmp.df <- reshape2::melt(tmp.df,id.vars=c("sample","polyA","rnase_inhib"))
colnames(tmp.df) <- c("sample","polyA","rnase_inhib", "biotype", "percent.counts")

suppressMessages(
  ggplot(
    tmp.df,
    aes(
      x=sample,
      y=percent.counts,
      group=sample,
      fill=biotype
    )
  )+
    geom_col(
      position="fill"
    )+
    scTheme$bar+
    theme(
      legend.position="right",
      axis.text.x=element_text(angle=45,hjust=1)
    )+
    scale_fill_manual(
      values=c(
        # mckolors$colblind_8,
        mckolors$polychrome[6:32]
      )
    )
  # coord_polar()+
  # ggsci::scale_fill_futurama()
)
```
### Gene biotype spatial plots
```{r Biotype spatial plots, fig.height=10, fig.width=17, message=FALSE, warning=FALSE, paged.print=TRUE}
i = c(
  # 1:3, # SkM ctrl
  # 4,6,7,5,8,9, # SkM SUPERase
  # 12,13,10,11 # # SkM Protector
  14:17
  )
visListPlot(
  vis.list[i],
  sample.titles = stringr::str_remove_all(meta_vis$sample[i],pattern = "Vis_") %>%
        stringr::str_remove_all(pattern ="yPAP_")%>%
        stringr::str_remove_all(pattern ="ctrl_")%>%
        stringr::str_remove_all(pattern ="_Heart")%>%
        stringr::str_remove_all(pattern ="_SkM"),
  # assay="STARsolo_collapsed",
  reduction="space",
  # slot = 'counts',
  pt.size=0.4,
  legend.position = "bottom",
  font.size = 12,
  axis.title.angle.y=0,
  nrow = 1,
  combine = T,
  verbose=F,
  colormap="plasma",
  # colormap = mckolors$RdYlBu %>% rev(),
  features = c(
    "pct.protein_coding",
    # "pct.lncRNA",
    "pct.rRNA",
    # "pct.Mt_rRNA", 
    "pct.Mt_tRNA",
    # "pct.ribo.pp",
    "pct.snoRNA",
    "pct.miRNA",
    "pct.snRNA"
  )
)&coord_fixed(ratio=1.6)&scTheme$space&theme(legend.text = element_text(size=6))
```

```{r}
i = c(
  # 1:3, # SkM ctrl
  # 4,6,7,5,8,9, # SkM SUPERase
  # 12,13,10,11 # # SkM Protector
  14:17
  )
visListPlot(
    vis.list[i],
    sample.titles = stringr::str_remove_all(meta_vis$sample[i],pattern = "Vis_") %>%
        stringr::str_remove_all(pattern ="yPAP_")%>%
        stringr::str_remove_all(pattern ="ctrl_")%>%
        stringr::str_remove_all(pattern ="_Heart"),
    reduction="space",
    # slot = 'counts',
    pt.size=1.2,
    legend.position = "bottom",
    font.size = 12,
    axis.title.angle.y=0,
    nrow = 1,
    combine = T,
    verbose=F,
    colormap = mckolors$RdYlBu %>% rev(),
    features = "pct.miRNA"
)&coord_fixed(ratio=1.6)
```
# QC analysis
```{r}
vis.merged@meta.data[order(vis.merged$nCount_kallisto_collapsed),] %>% ggplot(
  aes(
    x=nCount_kallisto_collapsed,
    fill=rnase_inhib
  )
)+
  geom_histogram(
    # geom="point"
  )+
  facet_wrap(
    facets = c("tissue","rnase_inhib")
  )
  # scTheme$bar

```



# Plotting
## 
```{r fig.height=6, fig.width=10, message=FALSE, warning=FALSE, paged.print=TRUE}

VlnPlot(
  vis.merged,
  assay="STARsolo_collapsed",
  # slot = 'counts',
  group.by="sample",
  fill.by = 'type',
  pt.size=0,
  combine = F,
  features = c(
    "nFeature_STARsolo","nCount_STARsolo","percent.mt",
    "pct.protein_coding","percent.ribo.RNA","percent.ribo.pp","pct.snoRNA","pct.miRNA"
  )
) %>% lapply(
  FUN=function(SEU) SEU + 
    scTheme$vln + 
    NoLegend() +
    theme(
      plot.title = element_text(hjust=0.5),
      axis.line = element_line(color="black"),
      axis.ticks = element_line(color="black"),
      axis.text.x = element_blank(),
      axis.title.x = element_blank()
    )
)%>%
  wrap_plots(nrow=2)
```


# Session Info
```{r}
sessionInfo()
```

